---
title: "WIP_death_explore"
author: "Cassandra Wattenburger"
date: "11/18/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(results = "show")
knitr::opts_chunk$set(message = FALSE)
```

```{r}
# Clear working directory, load in packages, generate package info
rm(list=ls())

library("tidyverse")
library("lmerTest")
library("emmeans")

sessionInfo()
```

# Import data

Files produced in 04_SFA2_deathestimates.Rmd
* death estimates
* normalized abundance data for time series with estimated death

```{r}
# Growth
death <- readRDS("../data_intermediate/SFA2_death_estimates_pap.rds")

# Normalized abundances
norm <- readRDS("../data_intermediate/SFA2_norm_deestimated.rds")

# Taxonomic information
tax <- readRDS("../data_intermediate/SFA2_taxonomy.rds")
```

# Overview

### Number of dying taxa

```{r}
# Number of taxa growing
num_died <- death %>% 
  group_by(DOC, Inoculant) %>% 
  summarize(num_died = n()) %>% 
  ungroup()

num_died

# treatment
num_died %>% 
  group_by(DOC) %>% 
  summarize(mean_num = mean(num_died), sd_num = sd(num_died)) %>% 
  ungroup()

num_died %>% 
  ggplot(aes(x=DOC, y=num_died)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="Number of death estimated taxa") +
  theme_test()
```

### Phylogenetic membership

```{r}
# Merge death estimates with taxonomy
death_tax <- inner_join(death, tax) 

# Graph phylogenetic membership
death_phylo_plot <- death_tax %>% 
  group_by(Phylum, DOC, Inoculant) %>% 
  summarize(count = n()) %>% 
  ungroup() %>% 
  arrange(count) %>% 
  ggplot(aes(x=count, y=reorder(Phylum, count), color=DOC)) +
  geom_boxplot() +
  labs(title="Growth", y="", x="Number of ASVs") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=16),
        axis.text = element_text(size=14),
        panel.grid.major.y = element_line(color = "gray",
                                          size = 0.5,
                                          linetype = 2))
death_phylo_plot
```

# Death parameters

* specific death rate (k)
* start day
* end day
* length of growth
* start abundance
* end abundance
* change in abundance

Other:
* 16S copy number
* genome size

# Sample-level analysis

Mean of all ASVs in each replicate.

```{r}
# Summarize parameters
death_sample <- death %>% 
  group_by(DOC, Inoculant) %>% 
  summarize(k = mean(k),
            start_day = mean(start_day),
            end_day = mean(end_day),
            length = mean(end_day - start_day),
            start_abund = mean(start_abund/n16S), # adjust by copy number
            end_abund = mean(end_abund/n16S), # adjust by copy number
            mean_change_abund = mean(change_abund/n16S), # adjust by copy number
            total_change_abund = sum(change_abund/n16S), # adjust by copy number
            n16S = mean(n16S),
            genome_size = mean(genome_size)) %>%  
  ungroup()
```

### Plots

```{r}
# Death k
death_sample %>% 
  ggplot(aes(x=DOC, y=k)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="specific death rate (k)") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=16),
        axis.text = element_text(size=14))
  
# Lag time
death_sample %>% 
  ggplot(aes(x=DOC, y=start_day)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="lag (days)") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=16),
        axis.text = element_text(size=14))


# Mean change abundance
death_sample %>% 
  ggplot(aes(x=DOC, y=mean_change_abund)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="Mean change in abundance") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=14),
        axis.text = element_text(size=14))

# Total change abundance
death_sample %>% 
  ggplot(aes(x=DOC, y=total_change_abund)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="Total change in abundance") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=14),
        axis.text = element_text(size=14))

# Length of death
death_sample %>%
  ggplot(aes(x=DOC, y=length)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="Length of death (days)") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=14),
        axis.text = element_text(size=14))

# rrn copies
death_sample %>%
  ggplot(aes(x=DOC, y=n16S)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="rrn copy number") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=14),
        axis.text = element_text(size=14))

# genome size
death_sample %>%
  ggplot(aes(x=DOC, y=genome_size)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="genome size") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=14),
        axis.text = element_text(size=14))
```

### Statistics

Specific death rate:

```{r}
# Linear model
k_lm <- lm(k ~ DOC, data=death_sample)
hist(resid(k_lm))
plot(predict(k_lm), resid(k_lm)) # heteroskedastic

# Welch's t-test
t.test(k ~ DOC, data=death_sample, var.equal=FALSE)
```

Lag time:

```{r}
# Linear model
lag_lm <- lm(start_day ~ DOC, data=death_sample)
hist(resid(lag_lm))
plot(predict(lag_lm), resid(lag_lm)) # heteroskedastic

# Welch's t-test
t.test(start_day ~ DOC, data=death_sample, var.equal=FALSE)
```

Length of death:

```{r}
# Linear model
length_lm <- lm(length ~ DOC, data=death_sample)
hist(resid(length_lm)) # non-normal
plot(predict(length_lm), resid(length_lm))

# Kruskal-Wallis test
kruskal.test(length ~ DOC, data=death_sample)
```

Mean change in abundance:

```{r}
# Linear model
change_lm <- lm(mean_change_abund ~ DOC, data=death_sample)
hist(resid(change_lm))
plot(predict(change_lm), resid(change_lm)) # heteroskedastic

# Welch's t-test
t.test(mean_change_abund ~ DOC, data=death_sample, var.equal=FALSE)
```

Total change in abundance:

```{r}
# Linear model
total_change_lm <- lm(total_change_abund ~ DOC, data=death_sample)
hist(resid(total_change_lm))
plot(predict(total_change_lm), resid(total_change_lm))
summary(total_change_lm)
```

rrn copies:

```{r}
# Linear model
rrn_lm <- lm(n16S ~ DOC, data=death_sample)
hist(resid(rrn_lm))
plot(predict(rrn_lm), resid(rrn_lm)) # heteroskedastic

# Welch's t-test
t.test(n16S ~ DOC, data=death_sample, var.equal=FALSE)
```

Genome size:

```{r}
# Linear model
genome_lm <- lm(genome_size ~ DOC, data=death_sample)
hist(resid(genome_lm)) # non-normal
plot(predict(genome_lm), resid(genome_lm)) # slightly heteroskedastic

# Kruskal Wallis test
kruskal.test(genome_size ~ DOC, data=death_sample)
```

### Relationship to DOC concentration

Import DOC measurements:

```{r}
# DOC measurements
doc <- readRDS("../data_DOC/SFA2/SFA2_doc.rds") %>% 
  select(everything(), Sample=sample_id) %>% 
  mutate(Sample = as.numeric(Sample))

# Metadata
meta <- read_tsv(file = "~/SFA/data_amplicon/SFA2_full/SFA2_full_metadata.tsv") %>% 
  select(Sample, Type, Inoculant, DOC, Replicate, Day)

# Merge and average technical replicates
doc_meta <- inner_join(doc, meta) %>% 
  group_by(DOC, Inoculant) %>% 
  summarize(mean_doc = mean(mean)) %>% 
  ungroup() %>% 
  mutate(Inoculant = as.character(Inoculant))

# Join with death estimates
doc_death <- inner_join(doc_meta, death_sample)
```

```{r}
# k
doc_death %>% 
  ggplot(aes(x=mean_doc, y=k)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# start day
doc_death %>% 
  ggplot(aes(x=mean_doc, y=start_day)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# start abundance
doc_death %>% 
  ggplot(aes(x=mean_doc, y=start_abund)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# end abundance
doc_death %>% 
  ggplot(aes(x=mean_doc, y=end_abund)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# mean change abundance
doc_death %>% 
  ggplot(aes(x=mean_doc, y=mean_change_abund)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# total change abundance
doc_death %>% 
  ggplot(aes(x=mean_doc, y=total_change_abund)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# genome size
doc_death %>% 
  ggplot(aes(x=mean_doc, y=genome_size)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# rrn
doc_death %>% 
  ggplot(aes(x=mean_doc, y=n16S)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()
```

# ASV-level analysis

Mean of all replicates for each ASV.

```{r}
death_asv <- death %>% 
  group_by(DOC, ASV) %>% 
  summarize(k = mean(k),
            start_pt = mean(start_pt),
            end_pt = mean(end_pt),
            start_day = mean(start_day),
            end_day = mean(end_day),
            length = mean(end_day - start_day),
            start_abund = mean(start_abund/n16S), # adjust for copy number
            end_abund = mean(end_abund/n16S), # adjust for copy number
            change_abund = mean(change_abund/n16S), # adjust for copy number
            n16S = mean(n16S),
            genome_size = mean(genome_size)) %>% 
  ungroup() %>% 
  inner_join(tax)
```

### Distributions

Specific death rates:

```{r}
# Distribution of death_asv rates
death_asv %>%
  ggplot(aes(x=k, color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="Specific death rate", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

Change in abundance:

```{r}
# Distribution of change in abundance during death_asv
death_asv %>%
  ggplot(aes(x=log(abs(change_abund)), color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="ln abs change in abundance during death", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

Start abundance:

```{r}
death_asv %>%
  ggplot(aes(x=log(start_abund), color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="ln Starting relational abundance", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

End abundance:

```{r}
death_asv %>%
  ggplot(aes(x=log(end_abund), color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="ln Ending relational abundance", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

Length of death:

```{r}
death_asv %>%
  ggplot(aes(x=length, color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="length of death_asv (days)", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

rrn copy number:

```{r}
death_asv %>%
  ggplot(aes(x=n16S, color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="rrn copies", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

Genome size:

```{r}
death_asv %>%
  ggplot(aes(x=genome_size, color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="genome size", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

# Mixed model approach

Likely losing too much power by taking averages. I'm going to use all of the data available in a mixed-model approach.

### Plots

```{r}
death %>% ggplot(aes(x=DOC, y=k, color=Inoculant)) +
  geom_boxplot() +
  geom_jitter() +
  theme_test()

death %>% ggplot(aes(x=DOC, y=start_day, color=Inoculant)) +
  geom_boxplot() +
  geom_jitter() +
  theme_test()

death %>% ggplot(aes(x=DOC, y=abs(change_abund), color=Inoculant)) +
  geom_boxplot() +
  geom_jitter() +
  theme_test()

death %>% ggplot(aes(x=DOC, y=log(abs(change_abund)), color=Inoculant)) +
  geom_boxplot() +
  geom_jitter() +
  theme_test()
```

### Statistics

Treating Inoculant as a random variable, like a blocking effect.

Specific death rate:

```{r}
# Linear mixed effects
k_lmer <- lmer(log(abs(k)) ~ DOC + (1|Inoculant), data=death)
hist(resid(k_lmer)) # log transform correct normality
plot(predict(k_lmer), resid(k_lmer))
summary(k_lmer)
```

Change in abundance:

```{r}
# Linear mixed effects
chgabund_lmer <- lmer(log(abs(change_abund)) ~ DOC + (1|Inoculant), data=death)
hist(resid(chgabund_lmer)) # log transform correct normality
plot(predict(chgabund_lmer), resid(chgabund_lmer))
summary(chgabund_lmer)
```

# Window heatmap

I want to know if the time frame of growth might be a mechanism determining DOC.

Reformat data:

```{r}
# Mark each day where death occurs for each taxon
norm_heat <- data.frame()
for (l in unique(death$label)) {
  death_sub <- filter(death, label==l)
  start <- death_sub$start_day
  end <- death_sub$end_day
  norm_sub <- filter(norm, label==l)
  norm_sub2 <- mutate(norm_sub, window = if_else(Day <= end & Day >= start, 1, 0))
  norm_heat <- bind_rows(norm_heat, norm_sub2)
}

# Summarize each Inoculant
norm_heat <- norm_heat %>% 
  group_by(Inoculant, DOC, Day) %>% 
  summarize(value = sum(window)) %>% 
  ungroup()
```

Plot:

Here, value refers to the number of taxa that were actively growing at that time point.

```{r}
norm_heat %>% 
  ggplot(aes(x=Day, y=Inoculant)) +
  geom_tile(aes(fill=value)) +  
  scale_fill_gradient(low="white", high="red") +
  facet_wrap(~DOC, nrow = 2) +
  theme_test()
```