{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Project: SFA2\n",
    "\n",
    "Cassandra Wattenburger, 11/01/21\n",
    "\n",
    "### Notes:\n",
    "* Sequenced data for all SFA2 samples including redos that did not sequence well originally\n",
    "* Original script written by Roli Wilhelm, heavily modified by Cassandra Wattenburger\n",
    "* QIIME2 v2021.4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline information\n",
    "\n",
    "### Pipeline to process raw sequences with DADA2 ###\n",
    "* Prep for import into QIIME2 (modify sequence IDs and combine two index files)\n",
    "* Import into QIIME2\n",
    "* Demultiplex\n",
    "* Denoise and merge with DADA2\n",
    "* Prepare ASV tables and representative sequences *(Note: sample names starting with a digit will break this step)*\n",
    "* Classify sequences\n",
    "* Construct phylogenetic tree\n",
    "* Export from QIIME2\n",
    "\n",
    "*100% Appropriated from the \"Atacama Desert Tutorial\" for QIIME2*\n",
    "\n",
    "### Pipeline can handle both 16S rRNA gene and ITS sequences ###\n",
    "* Tested on 515f and 806r\n",
    "* Tested on ITS1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Before you start\n",
    "\n",
    "### Commands to install dependencies ####\n",
    "##### || QIIME2 and biopython ||\n",
    "QIIME2 is still actively in development with frequent new releases. Check for the most up-to-date version and use that.\n",
    "\n",
    "Install QIIME2: \n",
    "* <https://docs.qiime2.org/2017.11/install/native/#install-qiime-2-within-a-conda-environment>, follow the instructions to install QIIME2 in Linux (64-bit).\n",
    "\n",
    "Activate the QIIME2 environment:\n",
    "* source activate [qiime2-pipeline-name]\n",
    "\n",
    "After you install and activate the QIIME2 environment, you must also install biopython for the barcode concatenation step to work. To install biopython make sure the QIIME2 environment is activated and run:\n",
    "* conda install -c anaconda biopython\n",
    "\n",
    "Install cutadapt to the QIIME2 environment as well. Cutadapt removes primers from the sequences.\n",
    "* conda install -c bioconda cutadapt\n",
    "\n",
    "##### || Copyrighter rrn database ||\n",
    "The script will automatically install the curated GreenGenes rrn attribute database: https://github.com/fangly/AmpliCopyrighter\n",
    "\n",
    "#### Citations\n",
    "* Caporaso, J. G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F. D., Costello, E. K., *et al.* (2010). QIIME allows analysis of high-throughput community sequencing data. Nature methods, 7(5), 335-336.\n",
    "\n",
    "* Angly, F. E., Dennis, P. G., Skarshewski, A., Vanwonterghem, I., Hugenholtz, P., & Tyson, G. W. (2014). CopyRighter: a rapid tool for improving the accuracy of microbial community profiles through lineage-specific gene copy number correction. Microbiome, 2(1), 11.\n",
    "\n",
    "### Using jupyter notebook screens ###\n",
    "\n",
    "With the QIIME2 environment activated, open your jupyter notebook screen in the directory containing this script:\n",
    "* jupyter-n [screen-name] [port #]\n",
    "\n",
    "See [these instructions](https://github.com/buckleylab/Buckley_lab_protocols/blob/master/Using_the_server/getting_started_on_server.md#make-jupyter-notebook-screens-command) for how to set up and use this command on the server.\n",
    "\n",
    "### Directory and data organization ###\n",
    "\n",
    "This pipeline assumes that you've organized your data in a certain way:\n",
    "* Each library of raw data is contained in a separate directory\n",
    "* Each library has a separate working directory within a larger project directory\n",
    "* Tree construction assumes all 16S libraries processed together will be analyzed together, so one tree is made based on all 16S libraries and is placed in a separate tree directory within the project directory\n",
    "\n",
    "### Troubleshooting tip ###\n",
    "Replace os.system with print, and copy/paste the output into the command line to view the error message."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1: User Input\n",
    "\n",
    "Metadata requirements:\n",
    "* Must be located in each library directory\n",
    "* Must be .tsv format \n",
    "* First column is named \"SampleID\" with sample names as rows\n",
    "* Contains another column named \"BarcodeSequence\" with the relevant barcode seqeunces as rows (rev. comp. reverse barcode sequence concatenated with forward barcode sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, re, numpy as np\n",
    "\n",
    "# Prepare an object with the name of the library and all related file paths\n",
    "# datasets = [['library name prefix', 'processed data directory path', 'raw data directory', 'read1 file name', 'read2 file name', 'index1 file name', 'index2 file name', 'metadata file name', 'domain of life (bacteria or fungi)'], ...]\n",
    "project = \"/home/cassi/SFAgrowthrate/data_amplicon/SFA2_full\" # this can be the same as the library directory if you only have one library to process\n",
    "libraries = [['SFA2_rep1', \n",
    "             '/home/cassi/SFAgrowthrate/data_amplicon/SFA2_full/rep1', \n",
    "             '/home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep1',\n",
    "             '145626_JTC3D_SFA2_rep1_S1_R1_001.fastq.gz', \n",
    "             '145626_JTC3D_SFA2_rep1_S1_R2_001.fastq.gz', \n",
    "             '145626_JTC3D_SFA2_rep1_S1_I1_001.fastq.gz', \n",
    "             '145626_JTC3D_SFA2_rep1_S1_I2_001.fastq.gz', \n",
    "             'SFA2_rep1_metadata.tsv',\n",
    "             'bacteria'],\n",
    "            ['SFA2_rep2', \n",
    "             '/home/cassi/SFAgrowthrate/data_amplicon/SFA2_full/rep2', \n",
    "             '/home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep2',\n",
    "             '145627_JWKRR_SFA2_rep2_S1_R1_001.fastq.gz', \n",
    "             '145627_JWKRR_SFA2_rep2_S1_R2_001.fastq.gz', \n",
    "             '145627_JWKRR_SFA2_rep2_S1_I1_001.fastq.gz', \n",
    "             '145627_JWKRR_SFA2_rep2_S1_I2_001.fastq.gz', \n",
    "             'SFA2_rep2_metadata.tsv',\n",
    "             'bacteria'],\n",
    "            ['SFA2_rep3', \n",
    "             '/home/cassi/SFAgrowthrate/data_amplicon/SFA2_full/rep3', \n",
    "             '/home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep3',\n",
    "             '145628_JWVWD_SFA2_rep3_S1_R1_001.fastq.gz', \n",
    "             '145628_JWVWD_SFA2_rep3_S1_R2_001.fastq.gz', \n",
    "             '145628_JWVWD_SFA2_rep3_S1_I1_001.fastq.gz', \n",
    "             '145628_JWVWD_SFA2_rep3_S1_I2_001.fastq.gz', \n",
    "             'SFA2_rep3_metadata.tsv',\n",
    "             'bacteria'],\n",
    "            ['SFA2_redos', \n",
    "             '/home/cassi/SFAgrowthrate/data_amplicon/SFA2_full/redos', \n",
    "             '/home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/redos',\n",
    "             '147155_G9HWY_SFA2_redos_S1_R1_001.fastq.gz', \n",
    "             '147155_G9HWY_SFA2_redos_S1_R2_001.fastq.gz', \n",
    "             '147155_G9HWY_SFA2_redos_S1_I1_001.fastq.gz', \n",
    "             '147155_G9HWY_SFA2_redos_S1_I2_001.fastq.gz', \n",
    "             'SFA2_redos_metadata.tsv',\n",
    "             'bacteria']]\n",
    "\n",
    "# Set # of processors\n",
    "processors = 10\n",
    "\n",
    "# Which bacterial database will you use? Silva or GreenGenes\n",
    "db = \"Silva\"\n",
    "\n",
    "# Phylogenetic tree (non-fungal data)\n",
    "treename = \"SFA2_full\" # name prefix for the tree file\n",
    "\n",
    "## Enter minimum support for keeping QIIME classification\n",
    "# Note: Classifications that do not meet this criteria will be retained, labeled 'putative'\n",
    "min_support = 0.9"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 2: Truncate sequence identifiers\n",
    "\n",
    "This step removes a portion at the end of the sequence ID that is incompatible with QIIME2. It will also create a modified directory in your raw data directory to house the modified data. The original raw data will not be modified, only the copies.\n",
    "\n",
    "Slow step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "for library in libraries:\n",
    "    raw = library[2]\n",
    "    read1 = library[3]\n",
    "    read2 = library[4]\n",
    "    index1 = library[5]\n",
    "    index2 = library[6]\n",
    "    \n",
    "    # Create a directory to place modified sequence data\n",
    "    if not os.path.isdir(os.path.join(raw, \"modified\")):\n",
    "        !mkdir $raw/modified\n",
    "    \n",
    "    # Copy/paste all raw sequence data into modified directory\n",
    "    !cp $raw/*.fastq.gz $raw/modified\n",
    "    \n",
    "    # Decompress all files\n",
    "    !unpigz $raw/modified/*.fastq.gz\n",
    "    \n",
    "    # Decompressed file names\n",
    "    read1decomp = re.sub(\".fastq.gz\", \".fastq\", read1)\n",
    "    read2decomp = re.sub(\".fastq.gz\", \".fastq\", read2)\n",
    "    index1decomp = re.sub(\".fastq.gz\", \".fastq\", index1)\n",
    "    index2decomp = re.sub(\".fastq.gz\", \".fastq\", index2)\n",
    "    \n",
    "    # Remove problematic part of sequence IDs\n",
    "    !sed 's/\\ [0-9]:[YN]:[0-9]:[0-9]$//g' $raw/modified/$read1decomp > $raw/modified/read1_mod.fastq\n",
    "    !sed 's/\\ [0-9]:[YN]:[0-9]:[0-9]$//g' $raw/modified/$read2decomp > $raw/modified/read2_mod.fastq\n",
    "    !sed 's/\\ [0-9]:[YN]:[0-9]:[0-9]$//g' $raw/modified/$index1decomp > $raw/modified/index1_mod.fastq\n",
    "    !sed 's/\\ [0-9]:[YN]:[0-9]:[0-9]$//g' $raw/modified/$index2decomp > $raw/modified/index2_mod.fastq\n",
    "    \n",
    "    # Delete file copies\n",
    "    !rm $raw/modified/$read1decomp\n",
    "    !rm $raw/modified/$read2decomp\n",
    "    !rm $raw/modified/$index1decomp\n",
    "    !rm $raw/modified/$index2decomp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 3: Remove primer from sequences\n",
    "\n",
    "There may still be portions of the primers left in the read sequences that need to be removed. Use cutadapt to remove those portions. Read1 will have the reverse complement of the reverse primer and read2 will have the reverse complement of the forward primer in some sequences on the 3' end.\n",
    "\n",
    "You will get a warning that the adapter is preceded by \"A\" or \"G\" extremely often. These are the link sequences in the primer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 3.4 with Python 3.8.8\n",
      "Command line parameters: -a ATTAGAWACCCBDGTAGTCC -o /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep1/modified/read1_noprimer.fastq /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep1/modified/read1_mod.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "[----->8     ] 00:02:53    14,552,895 reads  @     11.9 µs/read;   5.03 M reads/minute\n",
      "Finished in 173.50 s (12 µs/read; 5.03 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              14,552,895\n",
      "Reads with adapters:                 4,573,597 (31.4%)\n",
      "Reads written (passing filters):    14,552,895 (100.0%)\n",
      "\n",
      "Total basepairs processed: 3,652,776,645 bp\n",
      "Total written (filtered):  3,530,148,124 bp (96.6%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: ATTAGAWACCCBDGTAGTCC; Type: regular 3'; Length: 20; Trimmed: 4573597 times\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20 bp: 2\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 0.6%\n",
      "  C: 0.8%\n",
      "  G: 97.3%\n",
      "  T: 1.3%\n",
      "  none/other: 0.0%\n",
      "WARNING:\n",
      "    The adapter is preceded by \"G\" extremely often.\n",
      "    The provided adapter sequence could be incomplete at its 5' end.\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t50684\t227389.0\t0\t50684\n",
      "4\t13101\t56847.2\t0\t13101\n",
      "5\t3304\t14211.8\t0\t3304\n",
      "6\t1648\t3553.0\t0\t1648\n",
      "7\t399\t888.2\t0\t399\n",
      "8\t376\t222.1\t0\t376\n",
      "9\t423\t55.5\t0\t213 210\n",
      "10\t122\t13.9\t1\t106 16\n",
      "11\t131\t3.5\t1\t110 21\n",
      "12\t162\t0.9\t1\t137 25\n",
      "13\t94\t0.2\t1\t81 13\n",
      "14\t121\t0.1\t1\t108 13\n",
      "15\t160\t0.0\t1\t142 18\n",
      "16\t100\t0.0\t1\t92 8\n",
      "17\t112\t0.0\t1\t99 13\n",
      "18\t98\t0.0\t1\t82 12 4\n",
      "19\t365\t0.0\t1\t342 22 1\n",
      "20\t126\t0.0\t2\t104 15 7\n",
      "21\t147\t0.0\t2\t131 12 4\n",
      "22\t6510\t0.0\t2\t5419 597 494\n",
      "23\t840626\t0.0\t2\t749771 60093 30762\n",
      "24\t113501\t0.0\t2\t102893 7170 3438\n",
      "25\t8038\t0.0\t2\t7069 628 341\n",
      "26\t10651\t0.0\t2\t9359 802 490\n",
      "27\t1701284\t0.0\t2\t1521798 119604 59882\n",
      "28\t1110116\t0.0\t2\t997231 75151 37734\n",
      "29\t47916\t0.0\t2\t42071 3753 2092\n",
      "30\t594611\t0.0\t2\t528065 44473 22073\n",
      "31\t3712\t0.0\t2\t3062 419 231\n",
      "32\t756\t0.0\t2\t658 71 27\n",
      "33\t92\t0.0\t2\t75 11 6\n",
      "34\t45\t0.0\t2\t35 7 3\n",
      "35\t71\t0.0\t2\t66 3 2\n",
      "36\t246\t0.0\t2\t69 4 173\n",
      "37\t44\t0.0\t2\t32 2 10\n",
      "38\t12\t0.0\t2\t4 1 7\n",
      "39\t12\t0.0\t2\t9 1 2\n",
      "40\t22\t0.0\t2\t21 1\n",
      "41\t23\t0.0\t2\t20 3\n",
      "42\t97\t0.0\t2\t85 9 3\n",
      "43\t5\t0.0\t2\t5\n",
      "44\t26\t0.0\t2\t23 3\n",
      "45\t90\t0.0\t2\t84 3 3\n",
      "46\t73\t0.0\t2\t41 20 12\n",
      "47\t60822\t0.0\t2\t56567 2937 1318\n",
      "48\t342\t0.0\t2\t287 33 22\n",
      "49\t93\t0.0\t2\t74 6 13\n",
      "50\t20\t0.0\t2\t8 1 11\n",
      "51\t2\t0.0\t2\t1 0 1\n",
      "52\t3\t0.0\t2\t3\n",
      "53\t15\t0.0\t2\t3 1 11\n",
      "54\t1\t0.0\t2\t0 0 1\n",
      "55\t10\t0.0\t2\t7 0 3\n",
      "56\t3\t0.0\t2\t1 1 1\n",
      "57\t32\t0.0\t2\t12 11 9\n",
      "58\t10\t0.0\t2\t8 0 2\n",
      "59\t44\t0.0\t2\t3 1 40\n",
      "60\t1\t0.0\t2\t1\n",
      "61\t29\t0.0\t2\t19 9 1\n",
      "62\t3\t0.0\t2\t0 0 3\n",
      "63\t50\t0.0\t2\t45 3 2\n",
      "64\t9\t0.0\t2\t8 0 1\n",
      "65\t3\t0.0\t2\t0 0 3\n",
      "66\t2\t0.0\t2\t2\n",
      "68\t3\t0.0\t2\t2 0 1\n",
      "69\t1\t0.0\t2\t1\n",
      "70\t1\t0.0\t2\t0 1\n",
      "71\t1\t0.0\t2\t1\n",
      "73\t10\t0.0\t2\t6 2 2\n",
      "74\t85\t0.0\t2\t4 76 5\n",
      "76\t4\t0.0\t2\t2 1 1\n",
      "77\t8\t0.0\t2\t0 7 1\n",
      "78\t1\t0.0\t2\t0 1\n",
      "79\t2\t0.0\t2\t2\n",
      "81\t16\t0.0\t2\t0 11 5\n",
      "82\t2\t0.0\t2\t1 0 1\n",
      "83\t15\t0.0\t2\t7 5 3\n",
      "84\t6\t0.0\t2\t1 4 1\n",
      "85\t2\t0.0\t2\t2\n",
      "86\t1\t0.0\t2\t0 0 1\n",
      "87\t9\t0.0\t2\t4 3 2\n",
      "89\t2\t0.0\t2\t2\n",
      "92\t1\t0.0\t2\t0 1\n",
      "93\t4\t0.0\t2\t0 3 1\n",
      "94\t1\t0.0\t2\t0 0 1\n",
      "95\t1\t0.0\t2\t0 0 1\n",
      "96\t1\t0.0\t2\t0 0 1\n",
      "97\t3\t0.0\t2\t0 2 1\n",
      "98\t1\t0.0\t2\t1\n",
      "99\t7\t0.0\t2\t2 3 2\n",
      "102\t4\t0.0\t2\t0 1 3\n",
      "105\t3\t0.0\t2\t0 0 3\n",
      "106\t7\t0.0\t2\t3 2 2\n",
      "108\t5\t0.0\t2\t0 5\n",
      "109\t10\t0.0\t2\t0 10\n",
      "110\t1\t0.0\t2\t0 0 1\n",
      "113\t3\t0.0\t2\t0 2 1\n",
      "114\t6\t0.0\t2\t1 2 3\n",
      "115\t5\t0.0\t2\t3 1 1\n",
      "116\t3\t0.0\t2\t2 1\n",
      "118\t6\t0.0\t2\t2 1 3\n",
      "120\t1\t0.0\t2\t1\n",
      "121\t2\t0.0\t2\t0 0 2\n",
      "124\t2\t0.0\t2\t0 0 2\n",
      "126\t4\t0.0\t2\t2 1 1\n",
      "127\t1\t0.0\t2\t0 1\n",
      "129\t1\t0.0\t2\t0 1\n",
      "130\t5\t0.0\t2\t3 2\n",
      "132\t6\t0.0\t2\t1 2 3\n",
      "133\t1\t0.0\t2\t0 0 1\n",
      "134\t1\t0.0\t2\t1\n",
      "136\t2\t0.0\t2\t1 1\n",
      "138\t2\t0.0\t2\t1 1\n",
      "139\t4\t0.0\t2\t0 0 4\n",
      "140\t11\t0.0\t2\t0 1 10\n",
      "141\t2\t0.0\t2\t0 0 2\n",
      "142\t3\t0.0\t2\t0 0 3\n",
      "144\t2\t0.0\t2\t2\n",
      "145\t5\t0.0\t2\t4 0 1\n",
      "146\t3\t0.0\t2\t2 0 1\n",
      "148\t2\t0.0\t2\t0 0 2\n",
      "150\t6\t0.0\t2\t6\n",
      "152\t5\t0.0\t2\t0 3 2\n",
      "153\t1\t0.0\t2\t1\n",
      "154\t9\t0.0\t2\t0 9\n",
      "155\t3\t0.0\t2\t0 0 3\n",
      "157\t2\t0.0\t2\t2\n",
      "158\t1\t0.0\t2\t0 0 1\n",
      "159\t3\t0.0\t2\t3\n",
      "160\t2\t0.0\t2\t1 1\n",
      "161\t3\t0.0\t2\t0 0 3\n",
      "163\t1\t0.0\t2\t0 0 1\n",
      "167\t1\t0.0\t2\t1\n",
      "168\t1\t0.0\t2\t0 0 1\n",
      "169\t1\t0.0\t2\t0 0 1\n",
      "172\t9\t0.0\t2\t8 1\n",
      "174\t1\t0.0\t2\t0 0 1\n",
      "175\t26\t0.0\t2\t26\n",
      "177\t1\t0.0\t2\t1\n",
      "178\t2\t0.0\t2\t0 1 1\n",
      "179\t3\t0.0\t2\t2 1\n",
      "180\t1\t0.0\t2\t0 1\n",
      "181\t4\t0.0\t2\t0 3 1\n",
      "182\t27\t0.0\t2\t1 26\n",
      "183\t1\t0.0\t2\t0 1\n",
      "184\t8\t0.0\t2\t1 2 5\n",
      "185\t6\t0.0\t2\t3 0 3\n",
      "187\t2\t0.0\t2\t2\n",
      "188\t1\t0.0\t2\t1\n",
      "189\t2\t0.0\t2\t0 0 2\n",
      "190\t1\t0.0\t2\t1\n",
      "191\t1\t0.0\t2\t1\n",
      "192\t1\t0.0\t2\t0 1\n",
      "193\t8\t0.0\t2\t0 0 8\n",
      "194\t8\t0.0\t2\t0 7 1\n",
      "197\t2\t0.0\t2\t0 2\n",
      "200\t2\t0.0\t2\t0 1 1\n",
      "201\t2\t0.0\t2\t0 0 2\n",
      "203\t2\t0.0\t2\t0 0 2\n",
      "205\t1\t0.0\t2\t0 1\n",
      "206\t41\t0.0\t2\t0 0 41\n",
      "207\t1\t0.0\t2\t0 0 1\n",
      "208\t8\t0.0\t2\t6 1 1\n",
      "209\t18\t0.0\t2\t1 13 4\n",
      "210\t4\t0.0\t2\t1 3\n",
      "211\t339\t0.0\t2\t286 40 13\n",
      "212\t55\t0.0\t2\t52 1 2\n",
      "213\t3\t0.0\t2\t1 2\n",
      "214\t1\t0.0\t2\t1\n",
      "215\t1\t0.0\t2\t1\n",
      "216\t346\t0.0\t2\t5 319 22\n",
      "217\t10\t0.0\t2\t1 3 6\n",
      "218\t12\t0.0\t2\t1 0 11\n",
      "219\t17\t0.0\t2\t10 5 2\n",
      "220\t1\t0.0\t2\t1\n",
      "222\t10\t0.0\t2\t2 0 8\n",
      "223\t12\t0.0\t2\t10 0 2\n",
      "225\t5\t0.0\t2\t2 2 1\n",
      "226\t3\t0.0\t2\t3\n",
      "229\t1\t0.0\t2\t0 1\n",
      "230\t5\t0.0\t2\t1 4\n",
      "231\t16\t0.0\t2\t1 0 15\n",
      "232\t6\t0.0\t2\t4 0 2\n",
      "233\t2\t0.0\t2\t1 0 1\n",
      "234\t8\t0.0\t2\t1 5 2\n",
      "235\t1\t0.0\t2\t0 0 1\n",
      "236\t123\t0.0\t2\t88 26 9\n",
      "237\t7\t0.0\t2\t0 7\n",
      "238\t5\t0.0\t2\t2 0 3\n",
      "239\t3\t0.0\t2\t0 0 3\n",
      "240\t5\t0.0\t2\t1 0 4\n",
      "241\t10\t0.0\t2\t7 0 3\n",
      "242\t8\t0.0\t2\t1 0 7\n",
      "243\t25\t0.0\t2\t10 1 14\n",
      "244\t9\t0.0\t2\t2 1 6\n",
      "245\t10\t0.0\t2\t0 0 10\n",
      "246\t14\t0.0\t2\t0 1 13\n",
      "247\t8\t0.0\t2\t1 1 6\n",
      "248\t24\t0.0\t2\t0 0 24\n",
      "249\t7\t0.0\t2\t0 0 7\n",
      "250\t16\t0.0\t2\t2 0 14\n",
      "251\t189\t0.0\t2\t9 149 31\n",
      "\n",
      "\n",
      "WARNING:\n",
      "    One or more of your adapter sequences may be incomplete.\n",
      "    Please see the detailed output above.\n",
      "This is cutadapt 3.4 with Python 3.8.8\n",
      "Command line parameters: -a TTACCGCGGCKGCTGGCAC -o /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep1/modified/read2_noprimer.fastq /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep1/modified/read2_mod.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "[--->8       ] 00:02:13    14,552,895 reads  @      9.2 µs/read;   6.52 M reads/minute\n",
      "Finished in 133.92 s (9 µs/read; 6.52 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              14,552,895\n",
      "Reads with adapters:                 4,121,437 (28.3%)\n",
      "Reads written (passing filters):    14,552,895 (100.0%)\n",
      "\n",
      "Total basepairs processed: 3,652,776,645 bp\n",
      "Total written (filtered):  3,548,920,073 bp (97.2%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: TTACCGCGGCKGCTGGCAC; Type: regular 3'; Length: 19; Trimmed: 4121437 times\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 94.5%\n",
      "  C: 3.8%\n",
      "  G: 0.3%\n",
      "  T: 1.3%\n",
      "  none/other: 0.0%\n",
      "WARNING:\n",
      "    The adapter is preceded by \"A\" extremely often.\n",
      "    The provided adapter sequence could be incomplete at its 5' end.\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t42039\t227389.0\t0\t42039\n",
      "4\t185847\t56847.2\t0\t185847\n",
      "5\t132522\t14211.8\t0\t132522\n",
      "6\t599\t3553.0\t0\t599\n",
      "7\t627\t888.2\t0\t627\n",
      "8\t147\t222.1\t0\t147\n",
      "9\t166\t55.5\t0\t154 12\n",
      "10\t125\t13.9\t1\t72 53\n",
      "11\t119\t3.5\t1\t82 37\n",
      "12\t133\t0.9\t1\t107 26\n",
      "13\t75\t0.2\t1\t54 21\n",
      "14\t91\t0.1\t1\t65 26\n",
      "15\t132\t0.0\t1\t92 40\n",
      "16\t75\t0.0\t1\t66 9\n",
      "17\t94\t0.0\t1\t72 22\n",
      "18\t72\t0.0\t1\t20 52\n",
      "19\t267\t0.0\t1\t201 66\n",
      "20\t97\t0.0\t1\t70 27\n",
      "21\t118\t0.0\t1\t87 31\n",
      "22\t4802\t0.0\t1\t3860 942\n",
      "23\t645913\t0.0\t1\t522016 123897\n",
      "24\t84848\t0.0\t1\t8343 76505\n",
      "25\t6006\t0.0\t1\t4359 1647\n",
      "26\t8814\t0.0\t1\t7007 1807\n",
      "27\t1509121\t0.0\t1\t1308428 200693\n",
      "28\t908489\t0.0\t1\t756592 151897\n",
      "29\t37551\t0.0\t1\t30689 6862\n",
      "30\t493667\t0.0\t1\t420349 73318\n",
      "31\t2891\t0.0\t1\t2328 563\n",
      "32\t583\t0.0\t1\t471 112\n",
      "33\t74\t0.0\t1\t55 19\n",
      "34\t36\t0.0\t1\t31 5\n",
      "35\t57\t0.0\t1\t43 14\n",
      "36\t57\t0.0\t1\t42 15\n",
      "37\t29\t0.0\t1\t24 5\n",
      "38\t4\t0.0\t1\t2 2\n",
      "39\t7\t0.0\t1\t7\n",
      "40\t17\t0.0\t1\t11 6\n",
      "41\t16\t0.0\t1\t14 2\n",
      "42\t76\t0.0\t1\t56 20\n",
      "43\t5\t0.0\t1\t4 1\n",
      "44\t22\t0.0\t1\t21 1\n",
      "45\t75\t0.0\t1\t60 15\n",
      "46\t56\t0.0\t1\t17 39\n",
      "47\t53626\t0.0\t1\t47470 6156\n",
      "48\t272\t0.0\t1\t228 44\n",
      "49\t71\t0.0\t1\t59 12\n",
      "50\t9\t0.0\t1\t8 1\n",
      "51\t1\t0.0\t1\t1\n",
      "52\t2\t0.0\t1\t2\n",
      "53\t4\t0.0\t1\t2 2\n",
      "55\t7\t0.0\t1\t7\n",
      "56\t3\t0.0\t1\t1 2\n",
      "57\t36\t0.0\t1\t3 33\n",
      "58\t8\t0.0\t1\t7 1\n",
      "59\t4\t0.0\t1\t3 1\n",
      "60\t1\t0.0\t1\t1\n",
      "61\t20\t0.0\t1\t20\n",
      "63\t40\t0.0\t1\t35 5\n",
      "64\t5\t0.0\t1\t4 1\n",
      "66\t1\t0.0\t1\t1\n",
      "68\t2\t0.0\t1\t2\n",
      "69\t1\t0.0\t1\t0 1\n",
      "71\t2\t0.0\t1\t1 1\n",
      "73\t7\t0.0\t1\t6 1\n",
      "74\t21\t0.0\t1\t13 8\n",
      "76\t2\t0.0\t1\t2\n",
      "77\t1\t0.0\t1\t0 1\n",
      "79\t2\t0.0\t1\t1 1\n",
      "83\t11\t0.0\t1\t4 7\n",
      "84\t6\t0.0\t1\t0 6\n",
      "85\t2\t0.0\t1\t2\n",
      "87\t9\t0.0\t1\t0 9\n",
      "97\t4\t0.0\t1\t0 4\n",
      "98\t1\t0.0\t1\t1\n",
      "99\t3\t0.0\t1\t3\n",
      "102\t1\t0.0\t1\t0 1\n",
      "106\t2\t0.0\t1\t0 2\n",
      "114\t5\t0.0\t1\t1 4\n",
      "115\t5\t0.0\t1\t2 3\n",
      "116\t2\t0.0\t1\t1 1\n",
      "118\t3\t0.0\t1\t0 3\n",
      "125\t1\t0.0\t1\t1\n",
      "126\t2\t0.0\t1\t1 1\n",
      "130\t3\t0.0\t1\t3\n",
      "134\t1\t0.0\t1\t1\n",
      "138\t2\t0.0\t1\t1 1\n",
      "144\t2\t0.0\t1\t1 1\n",
      "145\t4\t0.0\t1\t3 1\n",
      "146\t2\t0.0\t1\t2\n",
      "150\t6\t0.0\t1\t6\n",
      "153\t1\t0.0\t1\t1\n",
      "157\t2\t0.0\t1\t2\n",
      "159\t3\t0.0\t1\t3\n",
      "160\t1\t0.0\t1\t1\n",
      "172\t8\t0.0\t1\t7 1\n",
      "173\t1\t0.0\t1\t0 1\n",
      "175\t26\t0.0\t1\t25 1\n",
      "179\t2\t0.0\t1\t1 1\n",
      "182\t15\t0.0\t1\t12 3\n",
      "185\t3\t0.0\t1\t2 1\n",
      "187\t1\t0.0\t1\t1\n",
      "188\t1\t0.0\t1\t1\n",
      "194\t1\t0.0\t1\t0 1\n",
      "195\t1\t0.0\t1\t1\n",
      "210\t3\t0.0\t1\t1 2\n",
      "211\t43\t0.0\t1\t39 4\n",
      "212\t5\t0.0\t1\t3 2\n",
      "215\t1\t0.0\t1\t0 1\n",
      "216\t215\t0.0\t1\t174 41\n",
      "217\t1\t0.0\t1\t1\n",
      "219\t13\t0.0\t1\t11 2\n",
      "225\t2\t0.0\t1\t2\n",
      "226\t1\t0.0\t1\t0 1\n",
      "229\t1\t0.0\t1\t0 1\n",
      "230\t2\t0.0\t1\t0 2\n",
      "231\t1\t0.0\t1\t1\n",
      "232\t2\t0.0\t1\t2\n",
      "233\t3\t0.0\t1\t0 3\n",
      "234\t4\t0.0\t1\t3 1\n",
      "236\t128\t0.0\t1\t1 127\n",
      "238\t3\t0.0\t1\t0 3\n",
      "240\t1\t0.0\t1\t1\n",
      "241\t7\t0.0\t1\t2 5\n",
      "242\t1\t0.0\t1\t1\n",
      "243\t7\t0.0\t1\t4 3\n",
      "244\t3\t0.0\t1\t2 1\n",
      "246\t1\t0.0\t1\t0 1\n",
      "247\t2\t0.0\t1\t1 1\n",
      "248\t1\t0.0\t1\t0 1\n",
      "250\t4\t0.0\t1\t0 4\n",
      "251\t130\t0.0\t1\t7 123\n",
      "\n",
      "\n",
      "WARNING:\n",
      "    One or more of your adapter sequences may be incomplete.\n",
      "    Please see the detailed output above.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 3.4 with Python 3.8.8\n",
      "Command line parameters: -a ATTAGAWACCCBDGTAGTCC -o /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep2/modified/read1_noprimer.fastq /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep2/modified/read1_mod.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "[ 8=---------] 00:03:11    16,008,099 reads  @     12.0 µs/read;   5.00 M reads/minute\n",
      "Finished in 191.92 s (12 µs/read; 5.00 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              16,008,099\n",
      "Reads with adapters:                 4,878,402 (30.5%)\n",
      "Reads written (passing filters):    16,008,099 (100.0%)\n",
      "\n",
      "Total basepairs processed: 4,018,032,849 bp\n",
      "Total written (filtered):  3,888,064,035 bp (96.8%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: ATTAGAWACCCBDGTAGTCC; Type: regular 3'; Length: 20; Trimmed: 4878402 times\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20 bp: 2\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 0.7%\n",
      "  C: 0.8%\n",
      "  G: 97.3%\n",
      "  T: 1.2%\n",
      "  none/other: 0.0%\n",
      "WARNING:\n",
      "    The adapter is preceded by \"G\" extremely often.\n",
      "    The provided adapter sequence could be incomplete at its 5' end.\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t47675\t250126.5\t0\t47675\n",
      "4\t12861\t62531.6\t0\t12861\n",
      "5\t3016\t15632.9\t0\t3016\n",
      "6\t1743\t3908.2\t0\t1743\n",
      "7\t470\t977.1\t0\t470\n",
      "8\t411\t244.3\t0\t411\n",
      "9\t392\t61.1\t0\t191 201\n",
      "10\t210\t15.3\t1\t173 37\n",
      "11\t269\t3.8\t1\t230 39\n",
      "12\t259\t1.0\t1\t227 32\n",
      "13\t143\t0.2\t1\t129 14\n",
      "14\t102\t0.1\t1\t89 13\n",
      "15\t147\t0.0\t1\t123 24\n",
      "16\t151\t0.0\t1\t129 22\n",
      "17\t95\t0.0\t1\t79 16\n",
      "18\t142\t0.0\t1\t118 22 2\n",
      "19\t4462\t0.0\t1\t3912 519 31\n",
      "20\t196\t0.0\t2\t159 27 10\n",
      "21\t149\t0.0\t2\t129 14 6\n",
      "22\t7793\t0.0\t2\t6072 844 877\n",
      "23\t1197356\t0.0\t2\t1039081 101332 56943\n",
      "24\t127996\t0.0\t2\t113536 9561 4899\n",
      "25\t10966\t0.0\t2\t9389 972 605\n",
      "26\t11907\t0.0\t2\t9982 1147 778\n",
      "27\t1594494\t0.0\t2\t1381204 136539 76751\n",
      "28\t987678\t0.0\t2\t861239 81429 45010\n",
      "29\t63200\t0.0\t2\t54038 5784 3378\n",
      "30\t724710\t0.0\t2\t627618 62271 34821\n",
      "31\t3602\t0.0\t2\t2870 466 266\n",
      "32\t678\t0.0\t2\t574 79 25\n",
      "33\t72\t0.0\t2\t56 9 7\n",
      "34\t44\t0.0\t2\t38 2 4\n",
      "35\t72\t0.0\t2\t55 8 9\n",
      "36\t284\t0.0\t2\t89 8 187\n",
      "37\t79\t0.0\t2\t65 10 4\n",
      "38\t88\t0.0\t2\t55 6 27\n",
      "39\t24\t0.0\t2\t23 1\n",
      "40\t14\t0.0\t2\t8 3 3\n",
      "41\t17\t0.0\t2\t13 3 1\n",
      "42\t85\t0.0\t2\t67 11 7\n",
      "43\t12\t0.0\t2\t10 2\n",
      "44\t8\t0.0\t2\t6 1 1\n",
      "45\t61\t0.0\t2\t50 7 4\n",
      "46\t101\t0.0\t2\t54 26 21\n",
      "47\t70882\t0.0\t2\t64337 4394 2151\n",
      "48\t836\t0.0\t2\t735 67 34\n",
      "49\t44\t0.0\t2\t27 1 16\n",
      "50\t20\t0.0\t2\t7 1 12\n",
      "51\t5\t0.0\t2\t3 0 2\n",
      "52\t4\t0.0\t2\t4\n",
      "53\t20\t0.0\t2\t2 1 17\n",
      "54\t7\t0.0\t2\t3 3 1\n",
      "56\t9\t0.0\t2\t5 3 1\n",
      "57\t35\t0.0\t2\t12 9 14\n",
      "58\t3\t0.0\t2\t3\n",
      "59\t20\t0.0\t2\t2 1 17\n",
      "60\t3\t0.0\t2\t2 0 1\n",
      "61\t21\t0.0\t2\t11 7 3\n",
      "63\t57\t0.0\t2\t52 2 3\n",
      "64\t17\t0.0\t2\t17\n",
      "66\t10\t0.0\t2\t8 0 2\n",
      "67\t1\t0.0\t2\t1\n",
      "68\t11\t0.0\t2\t9 1 1\n",
      "69\t2\t0.0\t2\t2\n",
      "70\t1\t0.0\t2\t1\n",
      "71\t14\t0.0\t2\t4 1 9\n",
      "73\t7\t0.0\t2\t5 1 1\n",
      "74\t185\t0.0\t2\t8 167 10\n",
      "75\t5\t0.0\t2\t2 3\n",
      "77\t12\t0.0\t2\t0 11 1\n",
      "79\t3\t0.0\t2\t2 0 1\n",
      "81\t7\t0.0\t2\t1 5 1\n",
      "82\t1\t0.0\t2\t0 0 1\n",
      "83\t9\t0.0\t2\t5 0 4\n",
      "84\t10\t0.0\t2\t4 2 4\n",
      "85\t3\t0.0\t2\t3\n",
      "86\t2\t0.0\t2\t2\n",
      "87\t16\t0.0\t2\t5 5 6\n",
      "88\t2\t0.0\t2\t1 0 1\n",
      "92\t2\t0.0\t2\t0 0 2\n",
      "94\t3\t0.0\t2\t0 3\n",
      "96\t7\t0.0\t2\t1 2 4\n",
      "97\t4\t0.0\t2\t1 1 2\n",
      "98\t3\t0.0\t2\t1 2\n",
      "99\t2\t0.0\t2\t0 2\n",
      "104\t1\t0.0\t2\t0 0 1\n",
      "105\t3\t0.0\t2\t0 0 3\n",
      "106\t8\t0.0\t2\t7 1\n",
      "107\t3\t0.0\t2\t0 1 2\n",
      "108\t2\t0.0\t2\t0 2\n",
      "109\t5\t0.0\t2\t0 4 1\n",
      "110\t1\t0.0\t2\t0 1\n",
      "111\t1\t0.0\t2\t0 1\n",
      "112\t2\t0.0\t2\t1 1\n",
      "114\t4\t0.0\t2\t0 1 3\n",
      "115\t6\t0.0\t2\t3 2 1\n",
      "116\t1\t0.0\t2\t0 0 1\n",
      "118\t12\t0.0\t2\t3 4 5\n",
      "120\t2\t0.0\t2\t0 0 2\n",
      "121\t1\t0.0\t2\t0 0 1\n",
      "123\t1\t0.0\t2\t0 0 1\n",
      "124\t1\t0.0\t2\t0 0 1\n",
      "125\t1\t0.0\t2\t1\n",
      "126\t2\t0.0\t2\t2\n",
      "127\t1\t0.0\t2\t0 1\n",
      "128\t1\t0.0\t2\t0 0 1\n",
      "129\t2\t0.0\t2\t0 2\n",
      "130\t3\t0.0\t2\t0 3\n",
      "131\t1\t0.0\t2\t0 0 1\n",
      "132\t5\t0.0\t2\t0 5\n",
      "134\t1\t0.0\t2\t0 1\n",
      "139\t2\t0.0\t2\t0 0 2\n",
      "140\t14\t0.0\t2\t1 0 13\n",
      "141\t1\t0.0\t2\t0 0 1\n",
      "142\t6\t0.0\t2\t2 1 3\n",
      "145\t1\t0.0\t2\t1\n",
      "146\t1\t0.0\t2\t0 0 1\n",
      "147\t6\t0.0\t2\t2 1 3\n",
      "149\t1\t0.0\t2\t1\n",
      "150\t2\t0.0\t2\t2\n",
      "151\t1\t0.0\t2\t0 0 1\n",
      "152\t3\t0.0\t2\t0 0 3\n",
      "153\t2\t0.0\t2\t2\n",
      "154\t1\t0.0\t2\t0 1\n",
      "157\t1\t0.0\t2\t0 0 1\n",
      "158\t3\t0.0\t2\t3\n",
      "159\t2\t0.0\t2\t2\n",
      "161\t2\t0.0\t2\t0 0 2\n",
      "162\t1\t0.0\t2\t1\n",
      "163\t3\t0.0\t2\t3\n",
      "164\t1\t0.0\t2\t0 0 1\n",
      "166\t1\t0.0\t2\t1\n",
      "172\t17\t0.0\t2\t15 2\n",
      "174\t2\t0.0\t2\t0 0 2\n",
      "175\t29\t0.0\t2\t28 1\n",
      "177\t7\t0.0\t2\t5 2\n",
      "178\t1\t0.0\t2\t1\n",
      "181\t8\t0.0\t2\t4 3 1\n",
      "182\t24\t0.0\t2\t3 18 3\n",
      "184\t11\t0.0\t2\t1 6 4\n",
      "185\t10\t0.0\t2\t7 0 3\n",
      "186\t3\t0.0\t2\t0 0 3\n",
      "187\t1\t0.0\t2\t1\n",
      "191\t4\t0.0\t2\t2 2\n",
      "193\t2\t0.0\t2\t1 0 1\n",
      "194\t15\t0.0\t2\t0 13 2\n",
      "196\t1\t0.0\t2\t0 0 1\n",
      "197\t9\t0.0\t2\t0 7 2\n",
      "199\t2\t0.0\t2\t0 1 1\n",
      "200\t2\t0.0\t2\t1 1\n",
      "201\t2\t0.0\t2\t1 0 1\n",
      "202\t3\t0.0\t2\t0 0 3\n",
      "203\t2\t0.0\t2\t0 0 2\n",
      "205\t2\t0.0\t2\t0 1 1\n",
      "206\t59\t0.0\t2\t0 0 59\n",
      "207\t1\t0.0\t2\t1\n",
      "208\t5\t0.0\t2\t3 2\n",
      "209\t3\t0.0\t2\t0 1 2\n",
      "210\t7\t0.0\t2\t0 7\n",
      "211\t418\t0.0\t2\t356 43 19\n",
      "212\t47\t0.0\t2\t44 3\n",
      "213\t3\t0.0\t2\t2 0 1\n",
      "214\t2\t0.0\t2\t1 0 1\n",
      "215\t5\t0.0\t2\t5\n",
      "216\t433\t0.0\t2\t2 390 41\n",
      "217\t6\t0.0\t2\t3 1 2\n",
      "218\t15\t0.0\t2\t0 3 12\n",
      "219\t13\t0.0\t2\t10 1 2\n",
      "220\t2\t0.0\t2\t2\n",
      "222\t9\t0.0\t2\t3 0 6\n",
      "223\t31\t0.0\t2\t24 2 5\n",
      "224\t7\t0.0\t2\t1 3 3\n",
      "225\t2\t0.0\t2\t0 2\n",
      "226\t8\t0.0\t2\t6 0 2\n",
      "227\t2\t0.0\t2\t0 0 2\n",
      "228\t1\t0.0\t2\t0 0 1\n",
      "229\t1\t0.0\t2\t0 0 1\n",
      "230\t6\t0.0\t2\t3 2 1\n",
      "231\t5\t0.0\t2\t1 0 4\n",
      "232\t9\t0.0\t2\t4 1 4\n",
      "233\t1\t0.0\t2\t0 1\n",
      "235\t2\t0.0\t2\t0 0 2\n",
      "236\t128\t0.0\t2\t87 31 10\n",
      "237\t9\t0.0\t2\t3 5 1\n",
      "238\t6\t0.0\t2\t1 1 4\n",
      "239\t2\t0.0\t2\t1 0 1\n",
      "240\t7\t0.0\t2\t3 1 3\n",
      "241\t8\t0.0\t2\t5 0 3\n",
      "242\t7\t0.0\t2\t3 0 4\n",
      "243\t20\t0.0\t2\t12 1 7\n",
      "244\t10\t0.0\t2\t0 1 9\n",
      "245\t20\t0.0\t2\t2 0 18\n",
      "246\t6\t0.0\t2\t0 0 6\n",
      "247\t2\t0.0\t2\t0 0 2\n",
      "248\t13\t0.0\t2\t1 0 12\n",
      "249\t5\t0.0\t2\t0 0 5\n",
      "250\t14\t0.0\t2\t4 0 10\n",
      "251\t220\t0.0\t2\t2 131 87\n",
      "\n",
      "\n",
      "WARNING:\n",
      "    One or more of your adapter sequences may be incomplete.\n",
      "    Please see the detailed output above.\n",
      "This is cutadapt 3.4 with Python 3.8.8\n",
      "Command line parameters: -a TTACCGCGGCKGCTGGCAC -o /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep2/modified/read2_noprimer.fastq /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep2/modified/read2_mod.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "[       8<---] 00:02:29    16,008,099 reads  @      9.3 µs/read;   6.43 M reads/minute\n",
      "Finished in 149.48 s (9 µs/read; 6.43 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              16,008,099\n",
      "Reads with adapters:                 4,465,664 (27.9%)\n",
      "Reads written (passing filters):    16,008,099 (100.0%)\n",
      "\n",
      "Total basepairs processed: 4,018,032,849 bp\n",
      "Total written (filtered):  3,906,216,221 bp (97.2%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: TTACCGCGGCKGCTGGCAC; Type: regular 3'; Length: 19; Trimmed: 4465664 times\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 94.9%\n",
      "  C: 3.6%\n",
      "  G: 0.3%\n",
      "  T: 1.2%\n",
      "  none/other: 0.0%\n",
      "WARNING:\n",
      "    The adapter is preceded by \"A\" extremely often.\n",
      "    The provided adapter sequence could be incomplete at its 5' end.\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t43389\t250126.5\t0\t43389\n",
      "4\t216240\t62531.6\t0\t216240\n",
      "5\t121583\t15632.9\t0\t121583\n",
      "6\t800\t3908.2\t0\t800\n",
      "7\t638\t977.1\t0\t638\n",
      "8\t174\t244.3\t0\t174\n",
      "9\t161\t61.1\t0\t154 7\n",
      "10\t168\t15.3\t1\t118 50\n",
      "11\t247\t3.8\t1\t178 69\n",
      "12\t223\t1.0\t1\t176 47\n",
      "13\t125\t0.2\t1\t94 31\n",
      "14\t87\t0.1\t1\t62 25\n",
      "15\t131\t0.0\t1\t108 23\n",
      "16\t140\t0.0\t1\t111 29\n",
      "17\t67\t0.0\t1\t50 17\n",
      "18\t114\t0.0\t1\t52 62\n",
      "19\t3415\t0.0\t1\t2531 884\n",
      "20\t148\t0.0\t1\t114 34\n",
      "21\t134\t0.0\t1\t113 21\n",
      "22\t5927\t0.0\t1\t4916 1011\n",
      "23\t954760\t0.0\t1\t800424 154336\n",
      "24\t100383\t0.0\t1\t10241 90142\n",
      "25\t8476\t0.0\t1\t6628 1848\n",
      "26\t10095\t0.0\t1\t8442 1653\n",
      "27\t1444225\t0.0\t1\t1258242 185983\n",
      "28\t831257\t0.0\t1\t708488 122769\n",
      "29\t52511\t0.0\t1\t44280 8231\n",
      "30\t602113\t0.0\t1\t517306 84807\n",
      "31\t2812\t0.0\t1\t2282 530\n",
      "32\t522\t0.0\t1\t429 93\n",
      "33\t53\t0.0\t1\t43 10\n",
      "34\t32\t0.0\t1\t28 4\n",
      "35\t55\t0.0\t1\t45 10\n",
      "36\t83\t0.0\t1\t70 13\n",
      "37\t61\t0.0\t1\t49 12\n",
      "38\t53\t0.0\t1\t46 7\n",
      "39\t21\t0.0\t1\t17 4\n",
      "40\t10\t0.0\t1\t6 4\n",
      "41\t12\t0.0\t1\t12\n",
      "42\t66\t0.0\t1\t60 6\n",
      "43\t11\t0.0\t1\t9 2\n",
      "44\t5\t0.0\t1\t2 3\n",
      "45\t44\t0.0\t1\t35 9\n",
      "46\t76\t0.0\t1\t28 48\n",
      "47\t62107\t0.0\t1\t55158 6949\n",
      "48\t703\t0.0\t1\t599 104\n",
      "49\t27\t0.0\t1\t27\n",
      "50\t9\t0.0\t1\t6 3\n",
      "51\t3\t0.0\t1\t3\n",
      "52\t4\t0.0\t1\t3 1\n",
      "53\t2\t0.0\t1\t2\n",
      "54\t4\t0.0\t1\t4\n",
      "55\t1\t0.0\t1\t0 1\n",
      "56\t11\t0.0\t1\t4 7\n",
      "57\t33\t0.0\t1\t3 30\n",
      "58\t3\t0.0\t1\t3\n",
      "59\t3\t0.0\t1\t3\n",
      "60\t2\t0.0\t1\t0 2\n",
      "61\t12\t0.0\t1\t11 1\n",
      "63\t51\t0.0\t1\t43 8\n",
      "64\t12\t0.0\t1\t10 2\n",
      "66\t8\t0.0\t1\t7 1\n",
      "67\t1\t0.0\t1\t1\n",
      "68\t9\t0.0\t1\t9\n",
      "69\t2\t0.0\t1\t2\n",
      "71\t3\t0.0\t1\t2 1\n",
      "73\t4\t0.0\t1\t3 1\n",
      "74\t93\t0.0\t1\t62 31\n",
      "75\t3\t0.0\t1\t1 2\n",
      "77\t4\t0.0\t1\t0 4\n",
      "79\t3\t0.0\t1\t2 1\n",
      "81\t1\t0.0\t1\t1\n",
      "83\t13\t0.0\t1\t5 8\n",
      "84\t16\t0.0\t1\t2 14\n",
      "85\t3\t0.0\t1\t3\n",
      "86\t2\t0.0\t1\t1 1\n",
      "87\t19\t0.0\t1\t0 19\n",
      "88\t2\t0.0\t1\t0 2\n",
      "94\t2\t0.0\t1\t0 2\n",
      "96\t1\t0.0\t1\t1\n",
      "97\t4\t0.0\t1\t0 4\n",
      "98\t2\t0.0\t1\t0 2\n",
      "102\t1\t0.0\t1\t0 1\n",
      "106\t8\t0.0\t1\t0 8\n",
      "109\t1\t0.0\t1\t0 1\n",
      "110\t1\t0.0\t1\t0 1\n",
      "112\t1\t0.0\t1\t0 1\n",
      "114\t3\t0.0\t1\t0 3\n",
      "115\t3\t0.0\t1\t1 2\n",
      "118\t9\t0.0\t1\t0 9\n",
      "125\t1\t0.0\t1\t0 1\n",
      "126\t1\t0.0\t1\t1\n",
      "129\t1\t0.0\t1\t0 1\n",
      "132\t1\t0.0\t1\t1\n",
      "133\t1\t0.0\t1\t1\n",
      "140\t3\t0.0\t1\t1 2\n",
      "142\t2\t0.0\t1\t2\n",
      "145\t1\t0.0\t1\t1\n",
      "147\t3\t0.0\t1\t1 2\n",
      "149\t1\t0.0\t1\t1\n",
      "150\t1\t0.0\t1\t0 1\n",
      "153\t1\t0.0\t1\t1\n",
      "158\t3\t0.0\t1\t3\n",
      "159\t2\t0.0\t1\t2\n",
      "162\t1\t0.0\t1\t0 1\n",
      "163\t10\t0.0\t1\t3 7\n",
      "166\t3\t0.0\t1\t0 3\n",
      "171\t1\t0.0\t1\t0 1\n",
      "172\t13\t0.0\t1\t8 5\n",
      "174\t2\t0.0\t1\t0 2\n",
      "175\t28\t0.0\t1\t24 4\n",
      "177\t1\t0.0\t1\t1\n",
      "178\t1\t0.0\t1\t1\n",
      "181\t5\t0.0\t1\t0 5\n",
      "182\t9\t0.0\t1\t2 7\n",
      "184\t1\t0.0\t1\t0 1\n",
      "185\t7\t0.0\t1\t7\n",
      "186\t1\t0.0\t1\t1\n",
      "187\t1\t0.0\t1\t1\n",
      "193\t1\t0.0\t1\t1\n",
      "194\t1\t0.0\t1\t0 1\n",
      "195\t4\t0.0\t1\t2 2\n",
      "200\t1\t0.0\t1\t1\n",
      "201\t1\t0.0\t1\t1\n",
      "205\t1\t0.0\t1\t1\n",
      "207\t1\t0.0\t1\t1\n",
      "209\t3\t0.0\t1\t0 3\n",
      "210\t1\t0.0\t1\t1\n",
      "211\t69\t0.0\t1\t63 6\n",
      "212\t4\t0.0\t1\t2 2\n",
      "213\t1\t0.0\t1\t1\n",
      "215\t2\t0.0\t1\t0 2\n",
      "216\t288\t0.0\t1\t249 39\n",
      "217\t2\t0.0\t1\t1 1\n",
      "218\t1\t0.0\t1\t0 1\n",
      "219\t11\t0.0\t1\t8 3\n",
      "222\t1\t0.0\t1\t0 1\n",
      "223\t3\t0.0\t1\t1 2\n",
      "226\t4\t0.0\t1\t3 1\n",
      "228\t1\t0.0\t1\t0 1\n",
      "230\t2\t0.0\t1\t1 1\n",
      "231\t1\t0.0\t1\t1\n",
      "232\t4\t0.0\t1\t3 1\n",
      "233\t2\t0.0\t1\t1 1\n",
      "235\t1\t0.0\t1\t0 1\n",
      "236\t145\t0.0\t1\t1 144\n",
      "237\t5\t0.0\t1\t3 2\n",
      "238\t1\t0.0\t1\t1\n",
      "239\t1\t0.0\t1\t0 1\n",
      "240\t1\t0.0\t1\t0 1\n",
      "241\t5\t0.0\t1\t2 3\n",
      "242\t3\t0.0\t1\t2 1\n",
      "243\t11\t0.0\t1\t4 7\n",
      "244\t1\t0.0\t1\t0 1\n",
      "245\t2\t0.0\t1\t1 1\n",
      "248\t1\t0.0\t1\t1\n",
      "250\t8\t0.0\t1\t3 5\n",
      "251\t107\t0.0\t1\t0 107\n",
      "\n",
      "\n",
      "WARNING:\n",
      "    One or more of your adapter sequences may be incomplete.\n",
      "    Please see the detailed output above.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 3.4 with Python 3.8.8\n",
      "Command line parameters: -a ATTAGAWACCCBDGTAGTCC -o /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep3/modified/read1_noprimer.fastq /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep3/modified/read1_mod.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "[>8          ] 00:03:51    19,366,781 reads  @     12.0 µs/read;   5.02 M reads/minute\n",
      "Finished in 231.63 s (12 µs/read; 5.02 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              19,366,781\n",
      "Reads with adapters:                 3,130,107 (16.2%)\n",
      "Reads written (passing filters):    19,366,781 (100.0%)\n",
      "\n",
      "Total basepairs processed: 4,861,062,031 bp\n",
      "Total written (filtered):  4,778,918,963 bp (98.3%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: ATTAGAWACCCBDGTAGTCC; Type: regular 3'; Length: 20; Trimmed: 3130107 times\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20 bp: 2\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 2.2%\n",
      "  C: 1.0%\n",
      "  G: 95.5%\n",
      "  T: 1.3%\n",
      "  none/other: 0.0%\n",
      "WARNING:\n",
      "    The adapter is preceded by \"G\" extremely often.\n",
      "    The provided adapter sequence could be incomplete at its 5' end.\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t52241\t302606.0\t0\t52241\n",
      "4\t14143\t75651.5\t0\t14143\n",
      "5\t4415\t18912.9\t0\t4415\n",
      "6\t1087\t4728.2\t0\t1087\n",
      "7\t431\t1182.1\t0\t431\n",
      "8\t200\t295.5\t0\t200\n",
      "9\t226\t73.9\t0\t68 158\n",
      "10\t165\t18.5\t1\t66 99\n",
      "11\t154\t4.6\t1\t56 98\n",
      "12\t94\t1.2\t1\t50 44\n",
      "13\t96\t0.3\t1\t54 42\n",
      "14\t78\t0.1\t1\t40 38\n",
      "15\t54\t0.0\t1\t28 26\n",
      "16\t54\t0.0\t1\t21 33\n",
      "17\t65\t0.0\t1\t42 23\n",
      "18\t41\t0.0\t1\t16 24 1\n",
      "19\t40\t0.0\t1\t18 17 5\n",
      "20\t76\t0.0\t2\t34 25 17\n",
      "21\t130\t0.0\t2\t30 24 76\n",
      "22\t6380\t0.0\t2\t2885 1449 2046\n",
      "23\t807560\t0.0\t2\t418689 218259 170612\n",
      "24\t88648\t0.0\t2\t50854 21304 16490\n",
      "25\t4955\t0.0\t2\t2573 1233 1149\n",
      "26\t6693\t0.0\t2\t2851 1773 2069\n",
      "27\t1086093\t0.0\t2\t593511 274887 217695\n",
      "28\t601426\t0.0\t2\t327360 150730 123336\n",
      "29\t35965\t0.0\t2\t19258 8717 7990\n",
      "30\t374953\t0.0\t2\t217856 87604 69493\n",
      "31\t2145\t0.0\t2\t1075 477 593\n",
      "32\t395\t0.0\t2\t228 83 84\n",
      "33\t29\t0.0\t2\t13 6 10\n",
      "34\t14\t0.0\t2\t7 4 3\n",
      "35\t22\t0.0\t2\t10 6 6\n",
      "36\t110\t0.0\t2\t31 20 59\n",
      "37\t19\t0.0\t2\t4 6 9\n",
      "38\t37\t0.0\t2\t14 12 11\n",
      "39\t19\t0.0\t2\t9 6 4\n",
      "40\t16\t0.0\t2\t11 2 3\n",
      "41\t10\t0.0\t2\t7 1 2\n",
      "42\t41\t0.0\t2\t22 16 3\n",
      "43\t9\t0.0\t2\t6 0 3\n",
      "44\t14\t0.0\t2\t7 4 3\n",
      "45\t36\t0.0\t2\t17 9 10\n",
      "46\t30\t0.0\t2\t15 4 11\n",
      "47\t38361\t0.0\t2\t25941 7380 5040\n",
      "48\t401\t0.0\t2\t250 102 49\n",
      "49\t37\t0.0\t2\t23 8 6\n",
      "50\t9\t0.0\t2\t0 3 6\n",
      "51\t2\t0.0\t2\t2\n",
      "52\t19\t0.0\t2\t12 1 6\n",
      "53\t4\t0.0\t2\t1 0 3\n",
      "54\t1\t0.0\t2\t0 0 1\n",
      "55\t1\t0.0\t2\t0 1\n",
      "56\t6\t0.0\t2\t2 0 4\n",
      "57\t18\t0.0\t2\t1 5 12\n",
      "58\t3\t0.0\t2\t2 0 1\n",
      "59\t6\t0.0\t2\t0 1 5\n",
      "61\t38\t0.0\t2\t18 11 9\n",
      "63\t22\t0.0\t2\t17 2 3\n",
      "64\t13\t0.0\t2\t7 4 2\n",
      "65\t4\t0.0\t2\t1 1 2\n",
      "66\t3\t0.0\t2\t1 1 1\n",
      "68\t3\t0.0\t2\t2 0 1\n",
      "69\t1\t0.0\t2\t0 0 1\n",
      "70\t1\t0.0\t2\t0 1\n",
      "71\t1\t0.0\t2\t0 1\n",
      "72\t1\t0.0\t2\t1\n",
      "73\t14\t0.0\t2\t12 0 2\n",
      "74\t69\t0.0\t2\t1 51 17\n",
      "77\t11\t0.0\t2\t2 6 3\n",
      "81\t21\t0.0\t2\t0 14 7\n",
      "82\t2\t0.0\t2\t0 0 2\n",
      "83\t4\t0.0\t2\t1 1 2\n",
      "84\t8\t0.0\t2\t1 1 6\n",
      "85\t4\t0.0\t2\t2 2\n",
      "86\t1\t0.0\t2\t0 1\n",
      "87\t6\t0.0\t2\t0 0 6\n",
      "93\t1\t0.0\t2\t0 1\n",
      "96\t2\t0.0\t2\t0 0 2\n",
      "97\t1\t0.0\t2\t0 1\n",
      "98\t4\t0.0\t2\t4\n",
      "99\t3\t0.0\t2\t0 1 2\n",
      "105\t2\t0.0\t2\t0 0 2\n",
      "106\t4\t0.0\t2\t2 2\n",
      "107\t2\t0.0\t2\t0 1 1\n",
      "108\t2\t0.0\t2\t0 1 1\n",
      "111\t1\t0.0\t2\t1\n",
      "112\t1\t0.0\t2\t0 0 1\n",
      "114\t1\t0.0\t2\t1\n",
      "115\t5\t0.0\t2\t2 3\n",
      "116\t2\t0.0\t2\t1 0 1\n",
      "117\t1\t0.0\t2\t0 0 1\n",
      "118\t2\t0.0\t2\t0 1 1\n",
      "120\t1\t0.0\t2\t0 1\n",
      "121\t1\t0.0\t2\t0 0 1\n",
      "122\t3\t0.0\t2\t0 1 2\n",
      "123\t1\t0.0\t2\t0 1\n",
      "125\t1\t0.0\t2\t1\n",
      "126\t1\t0.0\t2\t0 0 1\n",
      "128\t1\t0.0\t2\t1\n",
      "129\t4\t0.0\t2\t1 1 2\n",
      "133\t1\t0.0\t2\t0 1\n",
      "135\t3\t0.0\t2\t3\n",
      "137\t1\t0.0\t2\t0 0 1\n",
      "138\t1\t0.0\t2\t1\n",
      "139\t3\t0.0\t2\t0 0 3\n",
      "140\t3\t0.0\t2\t1 0 2\n",
      "141\t1\t0.0\t2\t0 0 1\n",
      "142\t1\t0.0\t2\t0 1\n",
      "144\t2\t0.0\t2\t1 1\n",
      "145\t7\t0.0\t2\t6 1\n",
      "146\t1\t0.0\t2\t0 0 1\n",
      "147\t6\t0.0\t2\t4 2\n",
      "149\t1\t0.0\t2\t0 1\n",
      "151\t2\t0.0\t2\t0 0 2\n",
      "152\t1\t0.0\t2\t0 0 1\n",
      "153\t1\t0.0\t2\t1\n",
      "154\t5\t0.0\t2\t1 4\n",
      "155\t1\t0.0\t2\t1\n",
      "156\t2\t0.0\t2\t0 0 2\n",
      "159\t1\t0.0\t2\t1\n",
      "167\t2\t0.0\t2\t1 1\n",
      "171\t4\t0.0\t2\t3 1\n",
      "172\t10\t0.0\t2\t7 1 2\n",
      "175\t61\t0.0\t2\t54 3 4\n",
      "179\t3\t0.0\t2\t3\n",
      "181\t6\t0.0\t2\t1 4 1\n",
      "182\t10\t0.0\t2\t1 8 1\n",
      "183\t4\t0.0\t2\t0 2 2\n",
      "184\t6\t0.0\t2\t0 3 3\n",
      "185\t3\t0.0\t2\t3\n",
      "187\t1\t0.0\t2\t1\n",
      "191\t2\t0.0\t2\t2\n",
      "192\t1\t0.0\t2\t0 0 1\n",
      "193\t2\t0.0\t2\t0 0 2\n",
      "194\t5\t0.0\t2\t0 3 2\n",
      "196\t3\t0.0\t2\t2 0 1\n",
      "197\t10\t0.0\t2\t0 9 1\n",
      "201\t4\t0.0\t2\t2 1 1\n",
      "204\t2\t0.0\t2\t0 2\n",
      "205\t18\t0.0\t2\t0 2 16\n",
      "206\t7\t0.0\t2\t0 0 7\n",
      "208\t3\t0.0\t2\t2 1\n",
      "209\t6\t0.0\t2\t1 5\n",
      "210\t8\t0.0\t2\t3 4 1\n",
      "211\t353\t0.0\t2\t288 50 15\n",
      "212\t82\t0.0\t2\t74 7 1\n",
      "213\t10\t0.0\t2\t10\n",
      "214\t3\t0.0\t2\t1 1 1\n",
      "215\t3\t0.0\t2\t2 0 1\n",
      "216\t360\t0.0\t2\t0 316 44\n",
      "217\t8\t0.0\t2\t2 5 1\n",
      "218\t13\t0.0\t2\t0 1 12\n",
      "219\t12\t0.0\t2\t8 2 2\n",
      "220\t1\t0.0\t2\t1\n",
      "221\t1\t0.0\t2\t1\n",
      "222\t9\t0.0\t2\t3 3 3\n",
      "223\t4\t0.0\t2\t0 0 4\n",
      "224\t3\t0.0\t2\t2 0 1\n",
      "226\t3\t0.0\t2\t3\n",
      "227\t4\t0.0\t2\t3 1\n",
      "228\t1\t0.0\t2\t1\n",
      "229\t2\t0.0\t2\t0 2\n",
      "230\t4\t0.0\t2\t3 1\n",
      "231\t4\t0.0\t2\t0 0 4\n",
      "232\t1\t0.0\t2\t1\n",
      "234\t1\t0.0\t2\t0 0 1\n",
      "235\t3\t0.0\t2\t0 0 3\n",
      "236\t122\t0.0\t2\t84 25 13\n",
      "237\t2\t0.0\t2\t1 1\n",
      "238\t1\t0.0\t2\t0 0 1\n",
      "239\t4\t0.0\t2\t1 0 3\n",
      "240\t8\t0.0\t2\t2 0 6\n",
      "241\t7\t0.0\t2\t3 1 3\n",
      "242\t5\t0.0\t2\t0 3 2\n",
      "243\t11\t0.0\t2\t7 1 3\n",
      "244\t5\t0.0\t2\t0 0 5\n",
      "245\t5\t0.0\t2\t1 0 4\n",
      "246\t4\t0.0\t2\t0 0 4\n",
      "247\t7\t0.0\t2\t3 1 3\n",
      "248\t27\t0.0\t2\t1 0 26\n",
      "249\t14\t0.0\t2\t2 2 10\n",
      "250\t30\t0.0\t2\t18 4 8\n",
      "251\t194\t0.0\t2\t2 160 32\n",
      "\n",
      "\n",
      "WARNING:\n",
      "    One or more of your adapter sequences may be incomplete.\n",
      "    Please see the detailed output above.\n",
      "This is cutadapt 3.4 with Python 3.8.8\n",
      "Command line parameters: -a TTACCGCGGCKGCTGGCAC -o /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep3/modified/read2_noprimer.fastq /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/rep3/modified/read2_mod.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "[------=8    ] 00:02:52    19,366,781 reads  @      8.9 µs/read;   6.74 M reads/minute\n",
      "Finished in 172.53 s (9 µs/read; 6.73 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              19,366,781\n",
      "Reads with adapters:                 1,240,018 (6.4%)\n",
      "Reads written (passing filters):    19,366,781 (100.0%)\n",
      "\n",
      "Total basepairs processed: 4,861,062,031 bp\n",
      "Total written (filtered):  4,846,948,332 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: TTACCGCGGCKGCTGGCAC; Type: regular 3'; Length: 19; Trimmed: 1240018 times\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 74.9%\n",
      "  C: 10.3%\n",
      "  G: 2.9%\n",
      "  T: 11.9%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t77909\t302606.0\t0\t77909\n",
      "4\t459576\t75651.5\t0\t459576\n",
      "5\t320221\t18912.9\t0\t320221\n",
      "6\t356\t4728.2\t0\t356\n",
      "7\t337\t1182.1\t0\t337\n",
      "8\t23\t295.5\t0\t23\n",
      "9\t57\t73.9\t0\t6 51\n",
      "10\t105\t18.5\t1\t5 100\n",
      "11\t41\t4.6\t1\t1 40\n",
      "12\t18\t1.2\t1\t1 17\n",
      "13\t14\t0.3\t1\t1 13\n",
      "14\t2\t0.1\t1\t0 2\n",
      "15\t4\t0.0\t1\t2 2\n",
      "16\t4\t0.0\t1\t0 4\n",
      "17\t8\t0.0\t1\t1 7\n",
      "18\t1\t0.0\t1\t0 1\n",
      "19\t1\t0.0\t1\t0 1\n",
      "20\t3\t0.0\t1\t0 3\n",
      "21\t4\t0.0\t1\t1 3\n",
      "22\t409\t0.0\t1\t102 307\n",
      "23\t78415\t0.0\t1\t21044 57371\n",
      "24\t3125\t0.0\t1\t255 2870\n",
      "25\t552\t0.0\t1\t165 387\n",
      "26\t903\t0.0\t1\t258 645\n",
      "27\t171054\t0.0\t1\t53283 117771\n",
      "28\t65069\t0.0\t1\t16685 48384\n",
      "29\t6890\t0.0\t1\t2446 4444\n",
      "30\t49154\t0.0\t1\t14901 34253\n",
      "31\t341\t0.0\t1\t88 253\n",
      "32\t63\t0.0\t1\t22 41\n",
      "33\t1\t0.0\t1\t1\n",
      "34\t1\t0.0\t1\t0 1\n",
      "35\t1\t0.0\t1\t0 1\n",
      "36\t4\t0.0\t1\t0 4\n",
      "37\t1\t0.0\t1\t0 1\n",
      "38\t1\t0.0\t1\t1\n",
      "40\t1\t0.0\t1\t0 1\n",
      "41\t1\t0.0\t1\t1\n",
      "42\t2\t0.0\t1\t0 2\n",
      "44\t3\t0.0\t1\t0 3\n",
      "45\t2\t0.0\t1\t1 1\n",
      "46\t2\t0.0\t1\t1 1\n",
      "47\t4479\t0.0\t1\t2006 2473\n",
      "48\t48\t0.0\t1\t15 33\n",
      "49\t8\t0.0\t1\t2 6\n",
      "50\t1\t0.0\t1\t1\n",
      "52\t1\t0.0\t1\t1\n",
      "56\t2\t0.0\t1\t0 2\n",
      "57\t11\t0.0\t1\t0 11\n",
      "61\t17\t0.0\t1\t2 15\n",
      "63\t3\t0.0\t1\t2 1\n",
      "64\t1\t0.0\t1\t0 1\n",
      "68\t1\t0.0\t1\t0 1\n",
      "73\t3\t0.0\t1\t1 2\n",
      "74\t13\t0.0\t1\t6 7\n",
      "77\t1\t0.0\t1\t0 1\n",
      "83\t3\t0.0\t1\t0 3\n",
      "84\t4\t0.0\t1\t0 4\n",
      "85\t2\t0.0\t1\t2\n",
      "87\t4\t0.0\t1\t0 4\n",
      "97\t2\t0.0\t1\t1 1\n",
      "98\t2\t0.0\t1\t2\n",
      "106\t2\t0.0\t1\t0 2\n",
      "108\t1\t0.0\t1\t0 1\n",
      "111\t1\t0.0\t1\t1\n",
      "114\t1\t0.0\t1\t1\n",
      "115\t2\t0.0\t1\t0 2\n",
      "116\t1\t0.0\t1\t1\n",
      "118\t2\t0.0\t1\t0 2\n",
      "128\t1\t0.0\t1\t0 1\n",
      "135\t3\t0.0\t1\t2 1\n",
      "138\t1\t0.0\t1\t0 1\n",
      "140\t1\t0.0\t1\t0 1\n",
      "144\t1\t0.0\t1\t0 1\n",
      "145\t2\t0.0\t1\t0 2\n",
      "147\t6\t0.0\t1\t5 1\n",
      "149\t1\t0.0\t1\t0 1\n",
      "150\t1\t0.0\t1\t0 1\n",
      "154\t1\t0.0\t1\t1\n",
      "156\t1\t0.0\t1\t0 1\n",
      "159\t1\t0.0\t1\t1\n",
      "163\t2\t0.0\t1\t0 2\n",
      "171\t4\t0.0\t1\t4\n",
      "172\t8\t0.0\t1\t7 1\n",
      "173\t2\t0.0\t1\t0 2\n",
      "175\t47\t0.0\t1\t34 13\n",
      "179\t1\t0.0\t1\t0 1\n",
      "181\t1\t0.0\t1\t0 1\n",
      "182\t8\t0.0\t1\t1 7\n",
      "185\t2\t0.0\t1\t2\n",
      "187\t1\t0.0\t1\t1\n",
      "196\t1\t0.0\t1\t1\n",
      "201\t3\t0.0\t1\t3\n",
      "205\t1\t0.0\t1\t0 1\n",
      "208\t1\t0.0\t1\t1\n",
      "210\t2\t0.0\t1\t0 2\n",
      "211\t43\t0.0\t1\t28 15\n",
      "212\t1\t0.0\t1\t0 1\n",
      "213\t3\t0.0\t1\t0 3\n",
      "215\t1\t0.0\t1\t0 1\n",
      "216\t223\t0.0\t1\t161 62\n",
      "217\t4\t0.0\t1\t4\n",
      "219\t8\t0.0\t1\t6 2\n",
      "221\t1\t0.0\t1\t0 1\n",
      "222\t3\t0.0\t1\t1 2\n",
      "224\t2\t0.0\t1\t1 1\n",
      "230\t3\t0.0\t1\t0 3\n",
      "232\t1\t0.0\t1\t0 1\n",
      "233\t1\t0.0\t1\t0 1\n",
      "234\t4\t0.0\t1\t0 4\n",
      "236\t126\t0.0\t1\t1 125\n",
      "237\t1\t0.0\t1\t0 1\n",
      "239\t3\t0.0\t1\t0 3\n",
      "240\t3\t0.0\t1\t1 2\n",
      "241\t5\t0.0\t1\t2 3\n",
      "243\t1\t0.0\t1\t1\n",
      "245\t1\t0.0\t1\t1\n",
      "247\t6\t0.0\t1\t3 3\n",
      "248\t1\t0.0\t1\t1\n",
      "249\t2\t0.0\t1\t0 2\n",
      "250\t24\t0.0\t1\t11 13\n",
      "251\t148\t0.0\t1\t3 145\n",
      "This is cutadapt 3.4 with Python 3.8.8\n",
      "Command line parameters: -a ATTAGAWACCCBDGTAGTCC -o /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/redos/modified/read1_noprimer.fastq /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/redos/modified/read1_mod.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "[          8=] 00:01:09     5,773,974 reads  @     12.0 µs/read;   5.00 M reads/minute\n",
      "Finished in 69.27 s (12 µs/read; 5.00 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               5,773,974\n",
      "Reads with adapters:                 1,961,941 (34.0%)\n",
      "Reads written (passing filters):     5,773,974 (100.0%)\n",
      "\n",
      "Total basepairs processed: 1,449,267,474 bp\n",
      "Total written (filtered):  1,397,604,775 bp (96.4%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: ATTAGAWACCCBDGTAGTCC; Type: regular 3'; Length: 20; Trimmed: 1961941 times\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1; 20 bp: 2\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 1.0%\n",
      "  C: 0.5%\n",
      "  G: 97.4%\n",
      "  T: 1.1%\n",
      "  none/other: 0.0%\n",
      "WARNING:\n",
      "    The adapter is preceded by \"G\" extremely often.\n",
      "    The provided adapter sequence could be incomplete at its 5' end.\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t12253\t90218.3\t0\t12253\n",
      "4\t3279\t22554.6\t0\t3279\n",
      "5\t993\t5638.6\t0\t993\n",
      "6\t444\t1409.7\t0\t444\n",
      "7\t156\t352.4\t0\t156\n",
      "8\t190\t88.1\t0\t190\n",
      "9\t132\t22.0\t0\t86 46\n",
      "10\t55\t5.5\t1\t42 13\n",
      "11\t104\t1.4\t1\t84 20\n",
      "12\t67\t0.3\t1\t52 15\n",
      "13\t83\t0.1\t1\t67 16\n",
      "14\t50\t0.0\t1\t42 8\n",
      "15\t65\t0.0\t1\t47 18\n",
      "16\t24\t0.0\t1\t17 7\n",
      "17\t118\t0.0\t1\t90 28\n",
      "18\t109\t0.0\t1\t89 20\n",
      "19\t112\t0.0\t1\t94 18\n",
      "20\t69\t0.0\t2\t52 7 10\n",
      "21\t59\t0.0\t2\t47 6 6\n",
      "22\t2119\t0.0\t2\t1244 357 518\n",
      "23\t482287\t0.0\t2\t385785 60515 35987\n",
      "24\t22106\t0.0\t2\t17911 2664 1531\n",
      "25\t2089\t0.0\t2\t1631 286 172\n",
      "26\t3104\t0.0\t2\t2137 507 460\n",
      "27\t1072908\t0.0\t2\t872730 127247 72931\n",
      "28\t182054\t0.0\t2\t147405 21599 13050\n",
      "29\t8667\t0.0\t2\t6795 1163 709\n",
      "30\t155826\t0.0\t2\t125761 19004 11061\n",
      "31\t802\t0.0\t2\t564 153 85\n",
      "32\t127\t0.0\t2\t95 22 10\n",
      "33\t53\t0.0\t2\t32 11 10\n",
      "34\t23\t0.0\t2\t17 5 1\n",
      "35\t20\t0.0\t2\t16 2 2\n",
      "36\t78\t0.0\t2\t54 7 17\n",
      "37\t24\t0.0\t2\t22 1 1\n",
      "38\t7\t0.0\t2\t7\n",
      "40\t18\t0.0\t2\t16 2\n",
      "41\t21\t0.0\t2\t12 5 4\n",
      "42\t32\t0.0\t2\t19 10 3\n",
      "43\t3\t0.0\t2\t3\n",
      "44\t1\t0.0\t2\t1\n",
      "45\t1\t0.0\t2\t0 1\n",
      "46\t4\t0.0\t2\t4\n",
      "47\t9774\t0.0\t2\t8559 825 390\n",
      "48\t58\t0.0\t2\t49 4 5\n",
      "49\t9\t0.0\t2\t5 2 2\n",
      "50\t5\t0.0\t2\t4 0 1\n",
      "52\t1\t0.0\t2\t1\n",
      "53\t5\t0.0\t2\t0 0 5\n",
      "55\t1\t0.0\t2\t0 0 1\n",
      "57\t7\t0.0\t2\t3 1 3\n",
      "61\t6\t0.0\t2\t4 1 1\n",
      "63\t4\t0.0\t2\t4\n",
      "64\t1\t0.0\t2\t1\n",
      "66\t1\t0.0\t2\t1\n",
      "70\t2\t0.0\t2\t2\n",
      "71\t2\t0.0\t2\t0 0 2\n",
      "73\t7\t0.0\t2\t5 0 2\n",
      "74\t3\t0.0\t2\t0 3\n",
      "75\t3\t0.0\t2\t0 3\n",
      "77\t3\t0.0\t2\t0 3\n",
      "79\t1\t0.0\t2\t0 1\n",
      "80\t2\t0.0\t2\t2\n",
      "81\t7\t0.0\t2\t0 6 1\n",
      "83\t6\t0.0\t2\t6\n",
      "84\t3\t0.0\t2\t1 0 2\n",
      "85\t1\t0.0\t2\t1\n",
      "86\t4\t0.0\t2\t0 4\n",
      "91\t1\t0.0\t2\t1\n",
      "92\t3\t0.0\t2\t1 2\n",
      "95\t2\t0.0\t2\t0 0 2\n",
      "97\t1\t0.0\t2\t1\n",
      "100\t2\t0.0\t2\t1 0 1\n",
      "105\t3\t0.0\t2\t1 0 2\n",
      "107\t2\t0.0\t2\t0 2\n",
      "115\t2\t0.0\t2\t1 0 1\n",
      "116\t5\t0.0\t2\t3 2\n",
      "120\t2\t0.0\t2\t1 1\n",
      "129\t1\t0.0\t2\t1\n",
      "133\t2\t0.0\t2\t0 0 2\n",
      "135\t1\t0.0\t2\t0 0 1\n",
      "140\t1\t0.0\t2\t0 0 1\n",
      "142\t2\t0.0\t2\t1 0 1\n",
      "144\t1\t0.0\t2\t0 0 1\n",
      "145\t1\t0.0\t2\t1\n",
      "146\t1\t0.0\t2\t0 0 1\n",
      "150\t2\t0.0\t2\t2\n",
      "151\t1\t0.0\t2\t1\n",
      "159\t4\t0.0\t2\t3 0 1\n",
      "160\t1\t0.0\t2\t0 1\n",
      "161\t1\t0.0\t2\t0 0 1\n",
      "171\t1\t0.0\t2\t0 0 1\n",
      "172\t3\t0.0\t2\t3\n",
      "174\t1\t0.0\t2\t1\n",
      "175\t1\t0.0\t2\t1\n",
      "176\t1\t0.0\t2\t0 0 1\n",
      "177\t3\t0.0\t2\t3\n",
      "181\t1\t0.0\t2\t0 1\n",
      "184\t1\t0.0\t2\t0 1\n",
      "185\t2\t0.0\t2\t1 1\n",
      "188\t1\t0.0\t2\t0 1\n",
      "190\t2\t0.0\t2\t2\n",
      "192\t6\t0.0\t2\t6\n",
      "193\t3\t0.0\t2\t0 2 1\n",
      "194\t1\t0.0\t2\t0 1\n",
      "200\t1\t0.0\t2\t0 0 1\n",
      "201\t1\t0.0\t2\t0 0 1\n",
      "203\t2\t0.0\t2\t0 0 2\n",
      "206\t5\t0.0\t2\t0 0 5\n",
      "207\t1\t0.0\t2\t1\n",
      "209\t10\t0.0\t2\t2 7 1\n",
      "210\t2\t0.0\t2\t0 2\n",
      "211\t416\t0.0\t2\t336 66 14\n",
      "212\t61\t0.0\t2\t54 6 1\n",
      "213\t10\t0.0\t2\t10\n",
      "215\t3\t0.0\t2\t2 0 1\n",
      "216\t132\t0.0\t2\t1 117 14\n",
      "217\t2\t0.0\t2\t1 1\n",
      "218\t2\t0.0\t2\t0 0 2\n",
      "219\t7\t0.0\t2\t5 2\n",
      "221\t1\t0.0\t2\t0 1\n",
      "222\t1\t0.0\t2\t0 0 1\n",
      "223\t96\t0.0\t2\t3 18 75\n",
      "224\t41\t0.0\t2\t3 4 34\n",
      "225\t3\t0.0\t2\t2 0 1\n",
      "226\t5\t0.0\t2\t5\n",
      "227\t2\t0.0\t2\t0 0 2\n",
      "229\t3\t0.0\t2\t0 0 3\n",
      "232\t17\t0.0\t2\t8 3 6\n",
      "233\t4\t0.0\t2\t3 1\n",
      "235\t4\t0.0\t2\t1 0 3\n",
      "236\t14\t0.0\t2\t5 4 5\n",
      "237\t11\t0.0\t2\t1 6 4\n",
      "238\t3\t0.0\t2\t0 0 3\n",
      "239\t8\t0.0\t2\t2 0 6\n",
      "240\t4\t0.0\t2\t0 0 4\n",
      "241\t3\t0.0\t2\t1 0 2\n",
      "242\t9\t0.0\t2\t0 0 9\n",
      "243\t29\t0.0\t2\t2 1 26\n",
      "244\t18\t0.0\t2\t5 0 13\n",
      "245\t17\t0.0\t2\t1 0 16\n",
      "246\t7\t0.0\t2\t0 0 7\n",
      "247\t5\t0.0\t2\t0 0 5\n",
      "248\t54\t0.0\t2\t0 0 54\n",
      "249\t28\t0.0\t2\t3 1 24\n",
      "250\t24\t0.0\t2\t0 1 23\n",
      "251\t154\t0.0\t2\t2 126 26\n",
      "\n",
      "\n",
      "WARNING:\n",
      "    One or more of your adapter sequences may be incomplete.\n",
      "    Please see the detailed output above.\n",
      "This is cutadapt 3.4 with Python 3.8.8\n",
      "Command line parameters: -a TTACCGCGGCKGCTGGCAC -o /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/redos/modified/read2_noprimer.fastq /home/backup_files/raw_reads/SFA2.cassi.2021/SFA2_full/redos/modified/read2_mod.fastq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "[   8=-------] 00:00:52     5,773,974 reads  @      9.1 µs/read;   6.57 M reads/minute\n",
      "Finished in 52.75 s (9 µs/read; 6.57 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:               5,773,974\n",
      "Reads with adapters:                 1,795,098 (31.1%)\n",
      "Reads written (passing filters):     5,773,974 (100.0%)\n",
      "\n",
      "Total basepairs processed: 1,449,267,474 bp\n",
      "Total written (filtered):  1,408,252,460 bp (97.2%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: TTACCGCGGCKGCTGGCAC; Type: regular 3'; Length: 19; Trimmed: 1795098 times\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-19 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 95.1%\n",
      "  C: 3.6%\n",
      "  G: 0.2%\n",
      "  T: 1.1%\n",
      "  none/other: 0.0%\n",
      "WARNING:\n",
      "    The adapter is preceded by \"A\" extremely often.\n",
      "    The provided adapter sequence could be incomplete at its 5' end.\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t12525\t90218.3\t0\t12525\n",
      "4\t248596\t22554.6\t0\t248596\n",
      "5\t37077\t5638.6\t0\t37077\n",
      "6\t176\t1409.7\t0\t176\n",
      "7\t183\t352.4\t0\t183\n",
      "8\t77\t88.1\t0\t77\n",
      "9\t69\t22.0\t0\t62 7\n",
      "10\t40\t5.5\t1\t28 12\n",
      "11\t73\t1.4\t1\t44 29\n",
      "12\t47\t0.3\t1\t31 16\n",
      "13\t60\t0.1\t1\t38 22\n",
      "14\t25\t0.0\t1\t13 12\n",
      "15\t47\t0.0\t1\t34 13\n",
      "16\t18\t0.0\t1\t12 6\n",
      "17\t82\t0.0\t1\t51 31\n",
      "18\t72\t0.0\t1\t14 58\n",
      "19\t74\t0.0\t1\t43 31\n",
      "20\t44\t0.0\t1\t28 16\n",
      "21\t41\t0.0\t1\t26 15\n",
      "22\t1163\t0.0\t1\t807 356\n",
      "23\t316178\t0.0\t1\t218778 97400\n",
      "24\t13565\t0.0\t1\t2463 11102\n",
      "25\t1212\t0.0\t1\t738 474\n",
      "26\t2190\t0.0\t1\t1597 593\n",
      "27\t902134\t0.0\t1\t710340 191794\n",
      "28\t130018\t0.0\t1\t97012 33006\n",
      "29\t6018\t0.0\t1\t4352 1666\n",
      "30\t114048\t0.0\t1\t88113 25935\n",
      "31\t529\t0.0\t1\t378 151\n",
      "32\t86\t0.0\t1\t55 31\n",
      "33\t30\t0.0\t1\t20 10\n",
      "34\t16\t0.0\t1\t12 4\n",
      "35\t16\t0.0\t1\t9 7\n",
      "36\t43\t0.0\t1\t29 14\n",
      "37\t19\t0.0\t1\t15 4\n",
      "38\t3\t0.0\t1\t2 1\n",
      "40\t11\t0.0\t1\t7 4\n",
      "41\t13\t0.0\t1\t10 3\n",
      "42\t18\t0.0\t1\t12 6\n",
      "43\t3\t0.0\t1\t2 1\n",
      "44\t1\t0.0\t1\t0 1\n",
      "46\t4\t0.0\t1\t3 1\n",
      "47\t8035\t0.0\t1\t6579 1456\n",
      "48\t38\t0.0\t1\t29 9\n",
      "49\t6\t0.0\t1\t5 1\n",
      "50\t4\t0.0\t1\t4\n",
      "52\t1\t0.0\t1\t1\n",
      "56\t2\t0.0\t1\t0 2\n",
      "57\t14\t0.0\t1\t2 12\n",
      "61\t4\t0.0\t1\t2 2\n",
      "63\t4\t0.0\t1\t4\n",
      "64\t1\t0.0\t1\t0 1\n",
      "66\t1\t0.0\t1\t1\n",
      "70\t2\t0.0\t1\t0 2\n",
      "79\t1\t0.0\t1\t0 1\n",
      "80\t2\t0.0\t1\t2\n",
      "83\t5\t0.0\t1\t4 1\n",
      "84\t9\t0.0\t1\t0 9\n",
      "85\t1\t0.0\t1\t1\n",
      "87\t1\t0.0\t1\t0 1\n",
      "91\t1\t0.0\t1\t1\n",
      "92\t2\t0.0\t1\t1 1\n",
      "97\t2\t0.0\t1\t0 2\n",
      "100\t1\t0.0\t1\t1\n",
      "105\t1\t0.0\t1\t1\n",
      "107\t2\t0.0\t1\t2\n",
      "114\t1\t0.0\t1\t0 1\n",
      "115\t3\t0.0\t1\t0 3\n",
      "116\t4\t0.0\t1\t4\n",
      "118\t1\t0.0\t1\t0 1\n",
      "129\t1\t0.0\t1\t1\n",
      "142\t1\t0.0\t1\t1\n",
      "144\t1\t0.0\t1\t0 1\n",
      "145\t1\t0.0\t1\t1\n",
      "150\t1\t0.0\t1\t0 1\n",
      "151\t2\t0.0\t1\t1 1\n",
      "159\t3\t0.0\t1\t2 1\n",
      "164\t1\t0.0\t1\t0 1\n",
      "170\t1\t0.0\t1\t0 1\n",
      "172\t3\t0.0\t1\t1 2\n",
      "174\t1\t0.0\t1\t1\n",
      "175\t1\t0.0\t1\t0 1\n",
      "176\t2\t0.0\t1\t0 2\n",
      "177\t2\t0.0\t1\t1 1\n",
      "185\t1\t0.0\t1\t1\n",
      "190\t2\t0.0\t1\t2\n",
      "191\t1\t0.0\t1\t0 1\n",
      "192\t6\t0.0\t1\t5 1\n",
      "195\t1\t0.0\t1\t0 1\n",
      "207\t1\t0.0\t1\t1\n",
      "208\t2\t0.0\t1\t0 2\n",
      "209\t11\t0.0\t1\t3 8\n",
      "211\t30\t0.0\t1\t27 3\n",
      "212\t3\t0.0\t1\t1 2\n",
      "215\t7\t0.0\t1\t1 6\n",
      "216\t56\t0.0\t1\t40 16\n",
      "218\t1\t0.0\t1\t1\n",
      "219\t5\t0.0\t1\t3 2\n",
      "220\t1\t0.0\t1\t0 1\n",
      "225\t2\t0.0\t1\t2\n",
      "226\t1\t0.0\t1\t0 1\n",
      "228\t6\t0.0\t1\t2 4\n",
      "232\t8\t0.0\t1\t8\n",
      "233\t3\t0.0\t1\t3\n",
      "235\t1\t0.0\t1\t1\n",
      "236\t12\t0.0\t1\t0 12\n",
      "237\t1\t0.0\t1\t1\n",
      "238\t1\t0.0\t1\t0 1\n",
      "239\t3\t0.0\t1\t2 1\n",
      "240\t2\t0.0\t1\t0 2\n",
      "241\t1\t0.0\t1\t1\n",
      "243\t3\t0.0\t1\t0 3\n",
      "244\t3\t0.0\t1\t2 1\n",
      "245\t1\t0.0\t1\t1\n",
      "249\t2\t0.0\t1\t2\n",
      "250\t1\t0.0\t1\t0 1\n",
      "251\t105\t0.0\t1\t2 103\n",
      "\n",
      "\n",
      "WARNING:\n",
      "    One or more of your adapter sequences may be incomplete.\n",
      "    Please see the detailed output above.\n"
     ]
    }
   ],
   "source": [
    "for library in libraries:\n",
    "    raw = library[2]\n",
    "    read1 = library[3]\n",
    "    read2 = library[4]\n",
    "    \n",
    "    !cutadapt -a ATTAGAWACCCBDGTAGTCC -o $raw/modified/read1_noprimer.fastq $raw/modified/read1_mod.fastq\n",
    "    !cutadapt -a TTACCGCGGCKGCTGGCAC -o $raw/modified/read2_noprimer.fastq $raw/modified/read2_mod.fastq\n",
    "    \n",
    "    # Delete unneeded intermediate files\n",
    "    !rm $raw/modified/read*_mod.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 4: Filter short reads\n",
    "\n",
    "Remove sequences with from primer-removed data with less than 100 bp from all files (too short).\n",
    "\n",
    "SLOW STEP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for library in libraries:\n",
    "    raw = library[2]\n",
    "    os.system(' '.join([\n",
    "        \"python /home/cassi/scripts/qiime2_filter_shortreads.py\",\n",
    "        raw+\"/modified/\"\n",
    "    ]))\n",
    "    \n",
    "    # Remove unneeded intermediate files\n",
    "    !rm $raw/modified/*_noprimer.fastq\n",
    "    !rm $raw/modified/index*_mod.fastq\n",
    "    \n",
    "    # Recompress modified read files\n",
    "    !pigz $raw/modified/read*_filtered.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 5: Concatenate barcodes\n",
    "\n",
    "This step calls a custom script, \"concatenate_barcodes_qiime2.py\". The script must be shuttled to the command line instead of run directly in jupyter notebooks because jupyter has memory issues that truncates the barcodes file without an error (I think).\n",
    "\n",
    "This script requires your index1 and index2 files to be named index1_mod.fastq and index2_mod.fastq and be located in a directory within the raw read directory called 'modified'. This should have been taken care of in earlier steps in this pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "for library in libraries:\n",
    "    raw = library[2]\n",
    "    \n",
    "    os.system(' '.join([\n",
    "        \"python /home/cassi/scripts/qiime2_concatenate_barcodes.py\",\n",
    "        raw+\"/modified\"]))\n",
    "  \n",
    "    # Recompress modified index files and newly created barcodes.fastq file\n",
    "    !pigz $raw/modified/*.fastq   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 6: Move raw data to library directories\n",
    "\n",
    "Creates intermediate directory in library directory. All subsequent files except for the final files will be placed there."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "for library in libraries:\n",
    "    proc = library[1]\n",
    "    raw = library[2]\n",
    "    \n",
    "    # Create output directory if it doesn't exist already\n",
    "    if not os.path.isdir(os.path.join(proc, \"intermediate\")):\n",
    "        !mkdir $proc/intermediate\n",
    "    \n",
    "    # Create a symbolic link to the read data\n",
    "    # QIIME2 import requires a directory containing files named: forward.fastq.gz, reverse.fastq.gz and barcodes.fastq.gz \n",
    "    !ln -s $raw/modified/read1_filtered.fastq.gz $proc/intermediate/forward.fastq.gz\n",
    "    !ln -s $raw/modified/read2_filtered.fastq.gz $proc/intermediate/reverse.fastq.gz\n",
    "    \n",
    "    # Move concatenated barcodes to project directory\n",
    "    !cp $raw/modified/barcodes.fastq.gz $proc/intermediate/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 7: Import into QIIME2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for library in libraries:\n",
    "    name = library[0]\n",
    "    proc = library[1]\n",
    "    \n",
    "    os.system(' '.join([\n",
    "        \"qiime tools import\",\n",
    "        \"--type EMPPairedEndSequences\",\n",
    "        \"--input-path \"+proc+\"/intermediate\",\n",
    "        \"--output-path \"+proc+\"/intermediate/\"+name+\".qza\"\n",
    "    ]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 8: Demultiplex\n",
    "\n",
    "The barcode you supply to QIIME is now a concatenation of your forward and reverse barcode. Your 'forward' barcode is actually the reverse complement of your reverse barcode and the 'reverse' is your forward barcode.\n",
    "\n",
    "Slow step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for library in libraries:\n",
    "    name = library[0]\n",
    "    proc = library[1]\n",
    "    metadata = library[7]\n",
    "    \n",
    "    os.system(' '.join([\n",
    "        \"qiime demux emp-paired\",\n",
    "        \"--m-barcodes-file \"+proc+\"/\"+metadata,\n",
    "        \"--m-barcodes-column BarcodeSequence\",\n",
    "        \"--p-no-golay-error-correction\",\n",
    "        \"--i-seqs \"+proc+\"/intermediate/\"+name+\".qza\",\n",
    "        \"--o-per-sample-sequences \"+proc+\"/intermediate/\"+name+\".demux\",\n",
    "        \"--o-error-correction-details \"+proc+\"/intermediate/\"+name+\".demux-details.qza\"\n",
    "    ]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 9: Visualize quality scores\n",
    "\n",
    "Drop output from below command into https://view.qiime2.org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "for library in libraries:\n",
    "    name = library[0]\n",
    "    proc = library[1]\n",
    "    \n",
    "    os.system(' '.join([\n",
    "        \"qiime demux summarize\",\n",
    "        \"--i-data \"+proc+\"/intermediate/\"+name+\".demux.qza\",\n",
    "        \"--o-visualization \"+proc+\"/intermediate/\"+name+\".demux.qzv\"\n",
    "    ]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 10: Trimming parameters | USER INPUT REQUIRED\n",
    "\n",
    "Based on the quality scores of the bp along the reads, choose trim and truncate values for the forward and reverse reads. Trim refers to the start of a sequence and truncate the total length (i.e. number of bases to remove from end).\n",
    "\n",
    "All trimming parameters must be the same for datasets that will be directly compared to one-another because ASVs are determined, in part, by sequence length."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Input your trimming parameters into a python dictionary for all libraries\n",
    "# trim_dict = {\"LibraryName1\":[trim_forward, truncate_forward, trim_reverse, truncate_reverse],\n",
    "#              \"LibraryName2\":[trim_forward, truncate_forward, trim_reverse, truncate_reverse],\n",
    "#               etc...}\n",
    "\n",
    "# The example in the Atacama Desert Tutorial trims 13 bp from the start of each read and does not remove any bases from the end of the 150 bp reads:\n",
    "#  --p-trim-left-f 13 \\  \n",
    "#  --p-trim-left-r 13 \\\n",
    "#  --p-trunc-len-f 150 \\\n",
    "#  --p-trunc-len-r 150\n",
    "\n",
    "trim_dict = {\"SFA2_rep1\":[10,200,10,200],\n",
    "             \"SFA2_rep2\":[10,200,10,200],\n",
    "             \"SFA2_rep3\":[10,200,10,200], # poor quality unfortunately\n",
    "             \"SFA2_redos\":[10,200,10,200]} "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 11: Trim, denoise and join (aka 'merge') reads using DADA2\n",
    "\n",
    "See the [QIIME2 dada2 denoise-paired documentation](https://docs.qiime2.org/2021.8/plugins/available/dada2/denoise-paired/) for the default parameters used.\n",
    "\n",
    "Slow step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for library in libraries:\n",
    "    name = library[0]\n",
    "    proc = library[1]\n",
    "    \n",
    "    os.system(' '.join([\n",
    "        \"qiime dada2 denoise-paired\",\n",
    "        \"--i-demultiplexed-seqs \"+proc+\"/intermediate/\"+name+\".demux.qza\",\n",
    "        \"--o-table \"+proc+\"/intermediate/\"+name+\".table.qza\",\n",
    "        \"--o-representative-sequences \"+proc+\"/intermediate/\"+name+\".rep-seqs.qza\",\n",
    "        \"--o-denoising-stats \"+proc+\"/intermediate/\"+name+\".denoising-stats.qza\",\n",
    "        \"--p-trim-left-f \"+str(trim_dict[name][0]),\n",
    "        \"--p-trim-left-r \"+str(trim_dict[name][2]),\n",
    "        \"--p-trunc-len-f \"+str(trim_dict[name][1]),\n",
    "        \"--p-trunc-len-r \"+str(trim_dict[name][3]),\n",
    "        \"--p-n-threads\",\n",
    "        str(processors)\n",
    "    ]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 12: Create summary of ASVs\n",
    "\n",
    "Drop outputs from below command into https://view.qiime2.org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for library in libraries:\n",
    "    name = library[0]\n",
    "    proc = library[1]\n",
    "    metadata = library[7]\n",
    "    \n",
    "    os.system(' '.join([\n",
    "        \"qiime feature-table summarize\",\n",
    "        \"--i-table \"+proc+\"/intermediate/\"+name+\".table.qza\",\n",
    "        \"--o-visualization \"+proc+\"/intermediate/\"+name+\".table.qzv\",\n",
    "        \"--m-sample-metadata-file \"+proc+\"/\"+metadata\n",
    "    ]))\n",
    "\n",
    "    os.system(' '.join([\n",
    "        \"qiime feature-table tabulate-seqs\",\n",
    "        \"--i-data \"+proc+\"/intermediate/\"+name+\".rep-seqs.qza\",\n",
    "        \"--o-visualization \"+proc+\"/intermediate/\"+name+\".rep-seqs.qzv\"\n",
    "    ])) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 13: Classify sequences\n",
    "\n",
    "Different QIIME2 versions can conflict with certain classifier database versions. This section will likely need to be updated. Download the latest classifiers here: https://docs.qiime2.org/2021.4/data-resources/\n",
    "\n",
    "* Using SILVA v138 pre-built classifier trained on scikit learn 0.24.1.\n",
    "\n",
    "View output in https://view.qiime2.org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use classifier for chosen database\n",
    "try:\n",
    "    if db == \"GreenGenes\":\n",
    "        classifier_db = \"/home/db/GreenGenes/qiime2_13.8.99_515.806_nb.classifier.qza\" # out of date\n",
    "    else:\n",
    "        classifier_db = \"~/databases/silva/silva-138-99-515-806-nb-classifier.qza\"\n",
    "except:\n",
    "        classifier_db = \"~/databases/silva/silva-138-99-515-806-nb-classifier.qza\"\n",
    "        \n",
    "for library in libraries:\n",
    "    name = library[0]\n",
    "    proc = library[1]\n",
    "    metadata = library[7]\n",
    "    domain = library[8]\n",
    "\n",
    "    # Classify\n",
    "    if domain == 'bacteria':\n",
    "        os.system(' '.join([\n",
    "            \"qiime feature-classifier classify-sklearn\",\n",
    "            \"--i-classifier\",\n",
    "            classifier_db,\n",
    "            \"--i-reads \"+proc+\"/intermediate/\"+name+\".rep-seqs.qza\",\n",
    "            \"--o-classification \"+proc+\"/intermediate/\"+name+\".taxonomy.qza\",\n",
    "            \"--p-n-jobs\",\n",
    "            str(processors)\n",
    "        ]))\n",
    "\n",
    "    if domain == 'fungi':\n",
    "        os.system(' '.join([\n",
    "            \"qiime feature-classifier classify-sklearn\",\n",
    "            \"--i-classifier /home/db/UNITE/qiime2_unite_ver7.99_20.11.2016_classifier.qza\", # out of date\n",
    "            \"--i-reads \"+proc+\"/intermediate/\"+name+\".rep-seqs.qza\",\n",
    "            \"--o-classification \"+proc+\"/intermediate/\"+name+\".taxonomy.qza\",\n",
    "            \"--p-n-jobs\",\n",
    "            str(processors)\n",
    "        ]))\n",
    "\n",
    "    # Output summary\n",
    "    os.system(' '.join([\n",
    "        \"qiime metadata tabulate\",\n",
    "        \"--m-input-file \"+proc+\"/intermediate/\"+name+\".taxonomy.qza\",\n",
    "        \"--o-visualization \"+proc+\"/intermediate/\"+name+\".taxonomy-summary.qzv\"\n",
    "    ])) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 14: Combine representative sequences for tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bacteria\n",
    "for library in libraries:\n",
    "    name = library[0]\n",
    "    proc = library[1]\n",
    "    domain = library[8]\n",
    "    \n",
    "    # Create representative sequences file\n",
    "    repseqsbac = []\n",
    "    \n",
    "    if domain == \"bacteria\":\n",
    "        repseqsbac.append(\"--i-data \" + os.path.join(proc, \"intermediate\", name+\".rep-seqs.qza\"))\n",
    "\n",
    "# Create tree directory\n",
    "if not os.path.isdir(os.path.join(project, \"tree\")):\n",
    "    !mkdir $project/tree\n",
    "        \n",
    "# Merge rep sequences from all bacterial libraries for tree\n",
    "os.system(' '.join([\n",
    "    \"qiime feature-table merge-seqs\",\n",
    "    \" \".join(repseqsbac),\n",
    "    \"--o-merged-data \"+project+\"/tree/\"+treename+\".rep-seqs-merged.qza\"\n",
    "]))\n",
    "\n",
    "# Fungi\n",
    "for library in libraries:\n",
    "    name = library[0]\n",
    "    proc = library[1]\n",
    "    domain = library[8]\n",
    "    \n",
    "    # Create representative sequences file\n",
    "    repseqsfung = []\n",
    "    \n",
    "    if domain == \"fungi\":\n",
    "        repseqsfung.append(\"--i-data \" + os.path.join(proc, \"intermediate\", name+\".rep-seqs.qza\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 15: Make phylogenetic tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "if domain == \"bacteria\":\n",
    "    # Generate alignment with MAFFT\n",
    "    os.system(' '.join([\n",
    "        \"qiime alignment mafft\",\n",
    "        \"--i-sequences \"+project+\"/tree/\"+treename+\".rep-seqs-merged.qza\",\n",
    "        \"--o-alignment \"+project+\"/tree/\"+treename+\".rep-seqs-merged-aligned.qza\",\n",
    "        \"--p-n-threads\",\n",
    "        str(processors)\n",
    "    ]))\n",
    "\n",
    "    # Mask hypervariable regions in alignment\n",
    "    os.system(' '.join([\n",
    "        \"qiime alignment mask\",\n",
    "        \"--i-alignment \"+project+\"/tree/\"+treename+\".rep-seqs-merged-aligned.qza\",\n",
    "        \"--o-masked-alignment \"+project+\"/tree/\"+treename+\".rep-seqs-merged-aligned-masked.qza\",\n",
    "    ]))\n",
    "\n",
    "    # Generate tree with FastTree\n",
    "    os.system(' '.join([\n",
    "        \"qiime phylogeny fasttree\",\n",
    "        \"--i-alignment \"+project+\"/tree/\"+treename+\".rep-seqs-merged-aligned-masked.qza\",\n",
    "        \"--o-tree \"+project+\"/tree/\"+treename+\".tree-unrooted.qza\",\n",
    "        \"--p-n-threads\",\n",
    "        str(processors)\n",
    "    ]))\n",
    "\n",
    "    # Root the tree\n",
    "    os.system(' '.join([\n",
    "        \"qiime phylogeny midpoint-root\",\n",
    "        \"--i-tree \"+project+\"/tree/\"+treename+\".tree-unrooted.qza\",\n",
    "        \"--o-rooted-tree \"+project+\"/tree/\"+treename+\".tree-rooted.qza\"\n",
    "    ]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 16: Reformat taxonomy\n",
    "\n",
    "Define function to tidy the taxonomy and make it compatible with phyloseq."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_taxonomy(tax_dirty, min_support):\n",
    "    output = open(re.sub(\".tsv\",\"-fixed.tsv\",tax_dirty), \"w\")\n",
    "    \n",
    "    full_rank_length = 7\n",
    "    output.write(\"\\t\".join([\"ASV\",\"Domain\",\"Phylum\",\"Class\",\"Order\",\"Family\",\"Genus\",\"Species\"])+\"\\n\")\n",
    "\n",
    "    with open(tax_dirty, \"r\") as f:\n",
    "        next(f)\n",
    "\n",
    "        for line in f:\n",
    "                line = line.strip()\n",
    "                line = line.split(\"\\t\")\n",
    "\n",
    "                read_id = line[0]\n",
    "                tax_string = line[1]\n",
    "                \n",
    "                ## Remove taxonomy prefixes and underscores (only coded for Silva classifications so far)\n",
    "                if db == \"Silva\":\n",
    "                    tax_string = re.sub(\"[a-z]__\", \"\", tax_string)\n",
    "                    tax_string = re.sub(\"_\", \" \", tax_string)\n",
    "\n",
    "                # Split full rank into ranks\n",
    "                full_rank = tax_string.split(\";\")\n",
    "\n",
    "                ## Identify the lowest classified taxonomic rank\n",
    "                # Account for cases when a taxonomic rank contains an empty space (common in GreenGenes output)\n",
    "                last_classified = full_rank[len(full_rank)-1]            \n",
    "\n",
    "                count = 1\n",
    "                while last_classified == \" \":\n",
    "                    last_classified = full_rank[len(full_rank)-count]\n",
    "                    count = count + 1\n",
    "\n",
    "                # Annotate the last classified as 'putative' if it does not meet the minimum support criteria\n",
    "                if float(line[2]) < float(min_support):\n",
    "                        full_rank[full_rank.index(last_classified)] = \"putative \"+last_classified\n",
    "                        last_classified = \"putative \"+last_classified\n",
    "\n",
    "                # Add in columns containing unclassified taxonomic information\n",
    "                for n in range(full_rank.index(last_classified)+1, full_rank_length, 1):               \n",
    "                    try:\n",
    "                        full_rank[n] = \"unclassified \"+last_classified\n",
    "                    except:\n",
    "                        full_rank.append(\"unclassified \"+last_classified)\n",
    "\n",
    "                # Write taxonomy to file\n",
    "                output.write(read_id+\"\\t\"+'\\t'.join(full_rank)+\"\\n\")\n",
    "    return()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 17: Export from QIIME2\n",
    "\n",
    "All final files will be placed in 'final' directory in library directories. Final tree will be in a 'tree' directory in the project directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "for library in libraries:\n",
    "    name = library[0]\n",
    "    proc = library[1]\n",
    "    metadata = library[7]\n",
    "\n",
    "    # Final output paths\n",
    "    fasta_final = proc+\"/final/\"+name+\".rep-seqs-final.fasta\"\n",
    "    tax_final= proc+\"/final/\"+name+\".taxonomy-final.tsv\"\n",
    "    count_final = proc+\"/final/\"+name+\".counts-final.biom\"\n",
    "    \n",
    "    # Make final data directories\n",
    "    if not os.path.isdir(os.path.join(proc, \"final\")):\n",
    "        !mkdir $proc/final\n",
    "        \n",
    "    # Export ASV table\n",
    "    os.system(' '.join([\n",
    "        \"qiime tools export\",\n",
    "        \"--input-path \"+proc+\"/intermediate/\"+name+\".table.qza\",\n",
    "        \"--output-path \"+proc+\"/intermediate/\"\n",
    "    ]))\n",
    "\n",
    "    # Export taxonomic classifications\n",
    "    os.system(' '.join([\n",
    "        \"qiime tools export\",\n",
    "        \"--input-path \"+proc+\"/intermediate/\"+name+\".taxonomy.qza\",\n",
    "        \"--output-path \"+proc+\"/intermediate/\"\n",
    "    ]))\n",
    "    \n",
    "    # Reformat classifications  \n",
    "    format_taxonomy(proc+\"/intermediate/taxonomy.tsv\", min_support)\n",
    "\n",
    "    # Export representative sequences\n",
    "    os.system(' '.join([\n",
    "        \"qiime tools export\",\n",
    "        \"--input-path \"+proc+\"/intermediate/\"+name+\".rep-seqs.qza\",\n",
    "        \"--output-path \"+proc+\"/intermediate/\"\n",
    "    ]))\n",
    "    \n",
    "    # Rename exported files and move to final directory\n",
    "    !mv $proc/intermediate/dna-sequences.fasta $fasta_final\n",
    "    !mv $proc/intermediate/feature-table.biom $count_final\n",
    "    !mv $proc/intermediate/taxonomy-fixed.tsv $tax_final\n",
    "    \n",
    "    # Reformat count table\n",
    "    tmp_tsv = re.sub(name+\".counts-final.biom\", \"tmp.tsv\", count_final)\n",
    "    tmp2_tsv = re.sub(name+\".counts-final.biom\", \"tmp2.tsv\", count_final)\n",
    "    count_tsv = re.sub(\".biom\", \".tsv\", count_final)\n",
    "    !biom convert -i $count_final -o $tmp_tsv --to-tsv # conver to .tsv\n",
    "    !tail -n +2 $tmp_tsv > $tmp2_tsv # remove header\n",
    "    !sed 's/\\#OTU ID/ASV/g' $tmp2_tsv > $count_tsv # replace OTU with ASV\n",
    "    !rm $tmp_tsv\n",
    "    !rm $tmp2_tsv\n",
    "    \n",
    "# Export tree\n",
    "os.system(' '.join([\n",
    "    \"qiime tools export\",\n",
    "    \"--input-path \"+project+\"/tree/\"+treename+\".tree-rooted.qza\",\n",
    "    \"--output-path \"+project+\"/tree/\"\n",
    "]))\n",
    "\n",
    "# Rename tree file\n",
    "tree_final = project+\"/tree/\"+treename+\".tree-final.nwk\"\n",
    "!mv $project/tree/\"tree.nwk\" $tree_final"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Done!"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "hide_input": true,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
