---
title: "WIP - SFA2 growth estimations exploration"
author: "Cassandra Wattenburger"
date: "11/16/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

```{r}
# Clear working directory, load in packages, generate package info
rm(list=ls())

library("tidyverse")
library("lmerTest")
library("emmeans")

sessionInfo()
```

# Import data

* growth estimates
* normalized abundance data for time series with estimated growth

```{r}
# Growth
growth <- readRDS("../data_intermediate/SFA2_growth_estimates_pap2.rds")

# Normalized abundance data
norm <- readRDS("../data_intermediate/SFA2_norm_grestimated.rds")

# Taxonomic information
tax <- readRDS("../data_intermediate/SFA2_taxonomy.rds")
```

# Overview

### Number of growing taxa

```{r}
# Number of taxa growing
num_grow <- growth %>% 
  group_by(DOC, Inoculant) %>% 
  summarize(num_grew = n()) %>% 
  ungroup()

num_grow

# treatment
num_grow %>% 
  group_by(DOC) %>% 
  summarize(mean_num = mean(num_grew), sd_num = sd(num_grew)) %>% 
  ungroup()

num_grow %>% 
  ggplot(aes(x=DOC, y=num_grew)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="Number of growth estimated taxa") +
  theme_test()
```

Statistics:

```{r}
# Linear model
num_lm <- lm(num_grew ~ DOC, data=num_grow)
hist(resid(num_lm))
plot(predict(num_lm), resid(num_lm)) # very heteroskedastic

# Welch's t-test
num_welch <- t.test(num_grew ~ DOC, data=num_grow, var.equal=FALSE)
num_welch
```

### Phylogenetic membership

```{r}
# Merge growth estimates with taxonomy
growth_tax <- inner_join(growth, tax) 

# Graph phylogenetic membership
growth_phylo_plot <- growth_tax %>% 
  group_by(Phylum, DOC, Inoculant) %>% 
  summarize(count = n()) %>% 
  ungroup() %>% 
  arrange(count) %>% 
  ggplot(aes(x=count, y=reorder(Phylum, count), color=DOC)) +
  geom_boxplot() +
  labs(title="Growth", y="", x="Number of ASVs") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=16),
        axis.text = element_text(size=14),
        panel.grid.major.y = element_line(color = "gray",
                                          size = 0.5,
                                          linetype = 2))
growth_phylo_plot
```

# Growth parameters

* specific growth rate (k)
* start day
* end day
* length of growth
* start abundance
* end abundance
* change in abundance

Other:
* 16S copy number
* genome size

# Sample-level analysis

Mean of all ASVs in each replicate.

```{r}
# Summarize parameters
growth_sample <- growth %>% 
  group_by(DOC, Innoculant) %>% 
  summarize(k = mean(k),
            start_day = mean(start_day),
            end_day = mean(end_day),
            length = mean(end_day - start_day),
            start_abund = mean(start_abund/n16S), # adjust by copy number
            end_abund = mean(end_abund/n16S), # adjust by copy number
            total_end_abund = sum(end_abund/n16S), # adjust by copy number
            mean_change_abund = mean(change_abund/n16S), # adjust by copy number
            total_change_abund = sum(change_abund/n16S), # adjust by copy number
            n16S = mean(n16S),
            genome_size = mean(genome_size)) %>%  
  ungroup()
```

### Plots

```{r}
# Mean k
k_plot <- growth_sample %>% 
  ggplot(aes(x=DOC, y=k)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="specific growth rate (k)") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=16),
        axis.text = element_text(size=14))
k_plot
  
# Mean start day
lag_plot <- growth_sample %>% 
  ggplot(aes(x=DOC, y=start_day)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="lag (days)") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=16),
        axis.text = element_text(size=14))
lag_plot

# Mean change abundance
change_abund_mean_plot <- growth_sample %>% 
  ggplot(aes(x=DOC, y=mean_change_abund)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="Mean change in abundance") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=14),
        axis.text = element_text(size=14))
change_abund_mean_plot

# Total change abundance
change_abund_total_plot <- growth_sample %>% 
  ggplot(aes(x=DOC, y=total_change_abund)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="Total change in abundance") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=14),
        axis.text = element_text(size=14))
change_abund_total_plot

# Length of growth
length_plot <- growth_sample %>%
  ggplot(aes(x=DOC, y=length)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="Length of growth (days)") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=14),
        axis.text = element_text(size=14))
length_plot

# rrn copies
rrn_plot <- growth_sample %>%
  ggplot(aes(x=DOC, y=n16S)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="rrn copy number") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=14),
        axis.text = element_text(size=14))
rrn_plot

# genome size
genome_plot <- growth_sample %>%
  ggplot(aes(x=DOC, y=genome_size)) +
  geom_point() +
  labs(x="DOC \"phenotype\"", y="genome size") +
  theme_test() +
  theme(title = element_text(size=18),
        axis.title = element_text(size=14),
        axis.text = element_text(size=14))
genome_plot
```

### Statistics

Specific growth rate:

```{r}
# Linear model
k_lm <- lm(k ~ DOC, data=growth_sample)
hist(resid(k_lm))
plot(predict(k_lm), resid(k_lm))
summary(k_lm)
```

Lag time:

```{r}
# Linear model
lag_lm <- lm(start_day ~ DOC, data=growth_sample)
hist(resid(lag_lm))
plot(predict(lag_lm), resid(lag_lm))
summary(lag_lm)
```

Length of growth:

```{r}
# Linear model
length_lm <- lm(length ~ DOC, data=growth_sample)
hist(resid(length_lm))
plot(predict(length_lm), resid(length_lm)) # heteroskedastic

# Use Welch's t-test
length_welch <- t.test(length ~ DOC, data=growth_sample, var.equal=FALSE)
length_welch
```

Mean change in abundance:

```{r}
# Linear model
change_lm <- lm(mean_change_abund ~ DOC, data=growth_sample)
hist(resid(change_lm))
plot(predict(change_lm), resid(change_lm))
summary(change_lm)
```

Total change in abundance:

```{r}
# Linear model
total_change_lm <- lm(total_change_abund ~ DOC, data=growth_sample)
hist(resid(total_change_lm))
plot(predict(total_change_lm), resid(total_change_lm)) # heteroskedastic

# Use Welch's t-test
change_welch <- t.test(total_change_abund ~ DOC, data=growth_sample, var.equal=FALSE)
change_welch
```

rrn copies:

```{r}
# Linear model
rrn_lm <- lm(n16S ~ DOC, data=growth_sample)
hist(resid(rrn_lm))
plot(predict(rrn_lm), resid(rrn_lm))
summary(rrn_lm)
```

Genome size:

```{r}
# Linear model
genome_lm <- lm(genome_size ~ DOC, data=growth_sample)
hist(resid(genome_lm))
plot(predict(genome_lm), resid(genome_lm)) # heteroskedastic, log transform makes worse
summary(genome_lm)
```

### Relationship to DOC concentration

Import DOC measurements

```{r}
# DOC measurements
doc <- readRDS("../data_DOC/SFA2/SFA2_doc.rds") %>% 
  select(everything(), Sample=sample_id) %>% 
  mutate(Sample = as.numeric(Sample))

# Metadata
meta <- read_tsv(file = "~/SFA/data_amplicon/SFA2_full/SFA2_full_metadata.tsv") %>% 
  select(Sample, Type, Inoculant, DOC, Replicate, Day)

# Merge and average technical replicates
doc_meta <- left_join(doc, meta) %>% 
  group_by(DOC, Inoculant) %>% 
  summarize(mean_doc = mean(mean)) %>% 
  ungroup() %>% 
  mutate(Inoculant = as.character(Inoculant))

# Join with growth estimates
doc_growth <- inner_join(doc_meta, growth_sample)
```

```{r}
# k
doc_growth %>% 
  ggplot(aes(x=mean_doc, y=k)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# start day
doc_growth %>% 
  ggplot(aes(x=mean_doc, y=start_day)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# start abundance
doc_growth %>% 
  ggplot(aes(x=mean_doc, y=start_abund)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# end abundance
doc_growth %>% 
  ggplot(aes(x=mean_doc, y=end_abund)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# mean change abundance
doc_growth %>% 
  ggplot(aes(x=mean_doc, y=mean_change_abund)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# total change abundance
doc_growth %>% 
  ggplot(aes(x=mean_doc, y=total_change_abund)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# genome size
doc_growth %>% 
  ggplot(aes(x=mean_doc, y=genome_size)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

# rrn
doc_growth %>% 
  ggplot(aes(x=mean_doc, y=n16S)) +
  geom_point(aes(color=DOC)) +
  geom_smooth(method="lm", linetype=2) +
  theme_test()

```

# ASV-level analysis

Mean of all replicates for each ASV.

```{r}
growth_asv <- growth %>% 
  group_by(DOC, ASV) %>% 
  summarize(k = mean(k),
            start_pt = mean(start_pt),
            end_pt = mean(end_pt),
            start_day = mean(start_day),
            end_day = mean(end_day),
            length = mean(end_day - start_day),
            start_abund = mean(start_abund/n16S), # adjust for copy number
            end_abund = mean(end_abund/n16S), # adjust for copy number
            change_abund = mean(change_abund/n16S), # adjust for copy number
            n16S = mean(n16S),
            genome_size = mean(genome_size)) %>% 
  ungroup() %>% 
  inner_join(tax)
```

### Change in abundance by phylum membership

```{r}
growth_asv %>% 
  ggplot(aes(y = Phylum, x=log(change_abund), color=Phylum)) +
  geom_point() +
  facet_wrap(~DOC, nrow=2) +
  theme_test()
```

### Distributions

Specific growth rates:

```{r}
# Distribution of growth_asv rates
growth_asv %>%
  ggplot(aes(x=k, color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="specific growth rate", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

Change in abundance:

```{r}
# Distribution of change in abundance during growth
growth_asv %>%
  ggplot(aes(x=log(change_abund), color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="ln change in abundance", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

Lag time:

```{r}
growth_asv %>%
  ggplot(aes(x=start_day, color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="lag time (days)", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

Length of growth:

```{r}
growth_asv %>%
  ggplot(aes(x=length, color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="length of growth (days)", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

rrn copy number:

```{r}
growth_asv %>%
  ggplot(aes(x=n16S, color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="rrn copies", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

Genome size:

```{r}
growth_asv %>%
  ggplot(aes(x=genome_size, color=DOC)) +
  geom_density(alpha=0.5) +
  labs(x="genome size (bp)", y="Kernel Density") +
  theme_test() +
  theme(#legend.position="none",
        title=element_text(size=12),
        axis.text=element_text(size=12),
        axis.title=element_text(size=12))
```

### Correlations

Change in abundace vs length of growth:

Prediction: ASvs that grew longer had more time to increase in abundance

```{r}
growth_asv %>% 
  ggplot(aes(x=length, y=log(change_abund))) +
  geom_point() +
  facet_wrap(~DOC) +
  geom_smooth(method="lm") +
  theme_test()
```

Stronger trend in high than low DOC. 

Change in abundace vs start of growth:

Prediction: ASVs that started growth earlier had more time to increase in abundance

```{r}
growth_asv %>% 
  ggplot(aes(x=start_day, y=log(change_abund))) +
  geom_point() +
  facet_wrap(~DOC) +
  geom_smooth(method="lm") +
  theme_test()
```

```{r}
growth_asv %>% 
  ggplot(aes(x=start_day, y=log(n16S))) +
  geom_point() +
  facet_wrap(~DOC) +
  geom_smooth(method="lm") +
  theme_test()
```


# Mixed model approach

Likely losing too much power by taking averages. I'm going to use all of the data available in a mixed-model approach.

### Plots

```{r}
# Growth rate
growth %>% ggplot(aes(x=DOC, y=k, color=Inoculant)) +
  geom_boxplot() +
  geom_jitter() +
  theme_test()

# Lag time
growth %>% ggplot(aes(x=DOC, y=start_day, color=Inoculant)) +
  geom_boxplot() +
  geom_jitter() +
  theme_test()

# Change in abundance
growth %>% ggplot(aes(x=DOC, y=change_abund, color=Inoculant)) +
  geom_boxplot() +
  geom_jitter() +
  theme_test()

# Change in abundance
growth %>% ggplot(aes(x=DOC, y=n16S, color=Inoculant)) +
  geom_boxplot() +
  geom_jitter() +
  theme_test()
```

It appears that a few very high change in abundance taxa in the low DOC community were driving the trend seen in the sample-level data. These ASVs are worth looking into further.

### Outliers

Isolate outlying ASVs:

Let's say anything above the maximum change in abundance observed in the high DOC treatment.

```{r, eval=FALSE}
# Isolate outliers
threshold <- growth %>% filter(DOC=="high") %>% group_by() %>% summarize(max = max(change_abund)) %>% as.numeric()
outliers <- growth %>% 
  filter(DOC=="low" & change_abund > threshold) 

# Who are they?
tax %>% filter(ASV %in% outliers$ASV)
```

```{r, eval=FALSE}
saveRDS(outliers, file="../data_intermediate/SFA2_growth_changeabundlow_outliers.rds")
```

Do these ASVs exist in the high DOC samples?

```{r}
out_asvs <- unique(outliers$ASV)
growth %>% filter(DOC=="high" & ASV %in% out_asvs)
```

Yes, some of them do. How do their change in abundances compare between treatments?

```{r}
growth %>% 
  filter(ASV %in% out_asvs) %>% 
  ggplot(aes(x=DOC, y=change_abund)) +
  geom_jitter() +
  geom_boxplot() +
  theme_test()
```

Statistics:

```{r}
# Linear model
out_chabund_lm <- lm(change_abund ~ DOC, filter(growth, ASV %in% out_asvs))
hist(resid(out_chabund_lm))  
plot(predict(out_chabund_lm), resid(out_chabund_lm)) # heteroskedastic

# Welch's t-test
out_chabund_welch  <- t.test(change_abund ~ DOC, filter(growth, ASV %in% out_asvs))
out_chabund_welch
```

### Statistics

Treating Inoculant as a random variable, like a blocking effect.

Growth rate:

```{r}
# Linear mixed effects
k_lmer <- lmer(log(k) ~ DOC + (1|Inoculant), data=growth)
hist(resid(k_lmer))
plot(predict(k_lmer), resid(k_lmer))
summary(k_lmer)
```

Change in abundance:

```{r}
# Linear mixed effects
chgabund_lmer <- lmer(log(change_abund) ~ DOC + (1|Inoculant), data=growth)
hist(resid(chgabund_lmer))
plot(predict(chgabund_lmer), resid(chgabund_lmer))

summary(chgabund_lmer)
```

# Window heatmap

I want to know if the time frame of growth might be a mechanism determining DOC.

Reformat data:

```{r}
# Mark each day where growth occurs for each taxon
norm_heat <- data.frame()
for (l in unique(growth$label)) {
  growth_sub <- filter(growth, label==l)
  start <- growth_sub$start_day
  end <- growth_sub$end_day
  norm_sub <- filter(norm, label==l)
  norm_sub2 <- mutate(norm_sub, window = if_else(Day <= end & Day >= start, 1, 0))
  norm_heat <- bind_rows(norm_heat, norm_sub2)
}

# Summarize each Inoculant
norm_heat <- norm_heat %>% 
  group_by(Inoculant, DOC, Day) %>% 
  summarize(value = sum(window)) %>% 
  ungroup()
```

Plot:

Here, value refers to the number of taxa that were actively growing at that time point.

```{r}
norm_heat %>% 
  ggplot(aes(x=Day, y=Inoculant)) +
  geom_tile(aes(fill=value)) +
  scale_fill_gradient(low="white", high="green") +
  facet_wrap(~DOC, nrow = 2) +
  theme_test()
```


