---
title: "Early vs late succession"
author: "Cassandra Wattenburger"
date: "3/28/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

```{r}
# Clear working directory, load in packages, generate package info
rm(list=ls())

library("tidyverse")
library("picante")

sessionInfo()
```

# Notes

# Import data

* growth and death estimates w/PAPRICA predictions
* normalized abundances
* taxonomy table
* phylogenetic tree

```{r}
# Metadata
meta <- readRDS("../data_intermediate/SFA2_metadata.rds")

# Growth
growth <- readRDS("../data_intermediate/SFA2_growth_estimates_pap.rds")

# Death
death <- readRDS("../data_intermediate/SFA2_death_estimates_pap.rds")

# Normalized abundance data for growth estimates
grnorm <- readRDS("../data_intermediate/SFA2_norm_grestimated.rds")

# Normalized abundance data for growth estimates
denorm <- readRDS("../data_intermediate/SFA2_norm_deestimated.rds")

# Normalized abundance data (all)
norm <- readRDS("../data_intermediate/SFA2_count_normalized.rds")

# Taxonomy
tax <- readRDS("../data_intermediate/SFA2_taxonomy.rds")

# Tree
tree <- read_tree("../data_amplicon/SFA2_full/tree/SFA2_full.tree-final.nwk")

```

# Distinguish early, mid, and late growers

Early = started on day 1
Mid = started between days 1 and 10
Late = started on or after day 10

```{r}
# Growth estimates
growth_succ <- growth %>% 
  mutate(Succession = if_else(start_day == 0, "early", "later")) %>% 
  mutate(Succession = as_factor(Succession))

growers <- select(growth_succ, label, Succession)
```

# Isolate death after growth

More interested in growth and then death.

```{r}
death_after <- death %>% 
  inner_join(growers)
```

# Proportions of early and later

```{r}
# Inoculant totals
growth_totals <- growth %>% 
  group_by(Inoculant) %>% 
  summarize(total = n()) %>% 
  ungroup()

# Succession totals
growth_succ_props <- growth_succ %>% 
  group_by(Inoculant, Succession) %>% 
  summarize(succ_total = n()) %>% 
  ungroup() %>% 
  full_join(growth_totals) %>% 
  mutate(proportion = succ_total/total)

# Visualize
growth_succ_props %>%
  ggplot(aes(x=Succession, y=proportion)) +
  geom_boxplot() +
  labs(y="Proportion of estimates") +
  geom_jitter(alpha=0.5) +
  theme_test()

# Summary
growth_succ_props %>% 
  group_by(Succession) %>% 
  summarize(prop_mean = mean(proportion), prop_sd = sd(proportion)) %>% 
  ungroup()
```

# Taxonomy of growers

```{r}
# By proportion
growth_succ_totals <- growth_succ %>% 
  group_by(Inoculant, Succession) %>% 
  summarize(total = n()) %>% 
  ungroup() 

growth_succ_prop <- growth_succ %>% 
  group_by(Inoculant, Succession, Phylum) %>% 
  summarize(phy_total = n()) %>%
  ungroup() %>% 
  full_join(growth_succ_totals) %>% 
  mutate(proportion = phy_total/total)

growth_succ_prop %>% 
  ggplot(aes(x=proportion, y=Phylum, color=Phylum, shape=Inoculant)) +
  geom_jitter(size=2, height=0.1) +
  facet_wrap(~Succession) +
  labs(x="Proportion of growth estimates") +
  theme_test() +
  theme(panel.grid.major.y = element_line(colour="gray"))
```

Seems as though Proteobacteria, Chloroflexi may have become more prominent in late succession. Bacteroidota and Myxococcota only present in early succession. Firmicutes, Actinobacteriota prominent in both early and late succession. Shifts could be happening at higher taxonomic resolution, however. Late seem to be made up of most speciose groups (more likely to pick up).

### Zoom in

Actinobacteriota:

```{r}
# By proportion, class level
growth_succ_prop_actino <- growth_succ %>% 
  filter(Phylum == "Actinobacteriota") %>% 
  group_by(Inoculant, Succession, Phylum, Class) %>% 
  summarize(class_total = n()) %>%
  ungroup() %>% 
  full_join(growth_succ_totals) %>% 
  mutate(proportion = class_total/total)

growth_succ_prop_actino %>% 
  ggplot(aes(x=proportion, y=Class, color=Class, shape=Inoculant)) +
  geom_jitter(size=2, height=0.1) +
  facet_wrap(~Succession) +
  labs(title="Actinobacteriota") +
  theme_test() +
  theme(panel.grid.major.y = element_line(colour="gray"))

# By proportion, genus level
growth_succ_prop_actino <- growth_succ %>% 
  filter(Phylum == "Actinobacteriota") %>% 
  group_by(Inoculant, Succession, Phylum, Genus) %>% 
  summarize(genus_total = n()) %>%
  ungroup() %>% 
  full_join(growth_succ_totals) %>% 
  mutate(proportion = genus_total/total)

growth_succ_prop_actino %>% 
  ggplot(aes(x=proportion, y=Genus, color=Genus, shape=Inoculant)) +
  geom_jitter(size=2, height=0.1) +
  facet_wrap(~Succession) +
  labs(title="Actinobacteriota") +
  theme_test() +
  theme(panel.grid.major.y = element_line(colour="gray"),
        legend.position = "none")
```

Firmcutes:

```{r}
# By proportion, class level
growth_succ_prop_firmi <- growth_succ %>% 
  filter(Phylum == "Firmicutes") %>% 
  group_by(Inoculant, Succession, Phylum, Class) %>% 
  summarize(class_total = n()) %>%
  ungroup() %>% 
  full_join(growth_succ_totals) %>% 
  mutate(proportion = class_total/total)

growth_succ_prop_firmi %>% 
  ggplot(aes(x=proportion, y=Class, color=Class, shape=Inoculant)) +
  geom_jitter(size=2, height=0.1) +
  facet_wrap(~Succession) +
  labs(title="Firmicutes") +
  theme_test() +
  theme(panel.grid.major.y = element_line(colour="gray"))

# By proportion, genus level
growth_succ_prop_firmi <- growth_succ %>% 
  filter(Phylum == "Firmicutes") %>% 
  group_by(Inoculant, Succession, Phylum, Genus) %>% 
  summarize(genus_total = n()) %>%
  ungroup() %>% 
  full_join(growth_succ_totals) %>% 
  mutate(proportion = genus_total/total)

growth_succ_prop_firmi %>% 
  ggplot(aes(x=proportion, y=Genus, color=Genus, shape=Inoculant)) +
  geom_jitter(size=2, height=0.1) +
  facet_wrap(~Succession) +
  labs(title="Firmicutes") +
  theme_test() +
  theme(panel.grid.major.y = element_line(colour="gray"),
        legend.position = "none")
```

Proteobacteria:

```{r}
# By proportion, class level
growth_succ_prop_proteo_class <- growth_succ %>% 
  filter(Phylum == "Proteobacteria") %>% 
  group_by(Inoculant, Succession, Phylum, Class) %>% 
  summarize(class_total = n()) %>%
  ungroup() %>% 
  full_join(growth_succ_totals) %>% 
  mutate(proportion = class_total/total)

growth_succ_prop_proteo_class %>% 
  ggplot(aes(x=proportion, y=Class, color=Class, shape=Inoculant)) +
  geom_jitter(size=2, height=0.1) +
  facet_wrap(~Succession) +
  labs(title="Proteobacteria") +
  theme_test() +
  theme(panel.grid.major.y = element_line(colour="gray"))

# By proportion, genus level
growth_succ_prop_proteo_genus <- growth_succ %>% 
  filter(Phylum == "Proteobacteria") %>% 
  group_by(Inoculant, Succession, Phylum, Genus) %>% 
  summarize(genus_total = n()) %>%
  ungroup() %>% 
  full_join(growth_succ_totals) %>% 
  mutate(proportion = genus_total/total)

growth_succ_prop_proteo_genus %>% 
  ggplot(aes(x=proportion, y=Genus, color=Genus, shape=Inoculant)) +
  geom_jitter(size=2, height=0.1) +
  facet_wrap(~Succession) +
  labs(title="Proteobacteria") +
  theme_test() +
  theme(panel.grid.major.y = element_line(colour="gray"),
        legend.position = "none")
```

Looks like the majority are Alphaproteobacteria, and they seem to be more dominant in later succession?

Statistics: 

Does Alphaproteobacteria have a higher relative abundance in late compared to early succession?

```{r}
# Try linear model first
alphapro_prop_lm <- lm(proportion ~ Succession, filter(growth_succ_prop_proteo_class, Class=="Alphaproteobacteria"))
hist(resid(alphapro_prop_lm)) # check normality of residuals
plot(predict(alphapro_prop_lm), resid(alphapro_prop_lm)) # check for heteroskedasticity
# Non-normal, use Kruskal-Wallis test
kruskal.test(proportion ~ Succession, filter(growth_succ_prop_proteo_class, Class=="Alphaproteobacteria"))
```

# Growth characteristics

### Overall

```{r}
# Average of each inoculant
growth_succ_overall <- growth_succ %>% 
  group_by(Inoculant, Succession) %>%
  summarize(k = mean(k),
            start_abund = mean(start_abund),
            end_abund = mean(end_abund),
            change_abund = mean(change_abund),
            length = mean(end_day - start_day)) %>% 
  ungroup()
  
# k 
growth_succ_overall %>% 
  ggplot(aes(x=Succession, y=k)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# start abundance
growth_succ_overall %>% 
  ggplot(aes(x=Succession, y=log(start_abund))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# end abundance
growth_succ_overall %>% 
  ggplot(aes(x=Succession, y=log(end_abund))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# change abundance
growth_succ_overall %>% 
  ggplot(aes(x=Succession, y=log(change_abund))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# length of growth
growth_succ_overall %>% 
  ggplot(aes(x=Succession, y=length)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()
```

Statistics

Specific growth rate (k):

```{r}
gr_k_lm <- lm(k ~ Succession, growth_succ_overall)
hist(resid(gr_k_lm))
plot(predict(gr_k_lm), resid(gr_k_lm))
summary(gr_k_lm)
```

Change in abundance:

```{r}
gr_change_lm <- lm(change_abund ~ Succession, growth_succ_overall)
hist(resid(gr_change_lm)) # not normal
plot(predict(gr_change_lm), resid(gr_change_lm))

# Kruskal Wallis
kruskal.test(change_abund ~ Succession, growth_succ_overall)
```

Length:

```{r}
gr_length_lm <- lm(length ~ Succession, growth_succ_overall)
hist(resid(gr_length_lm)) # non normal
plot(predict(gr_length_lm), resid(gr_length_lm))

# Use Kruskal Wallis test instead
kruskal.test(length ~ Succession, growth_succ_overall)
```

### Individual inoculants

```{r}
# k 
growth_succ %>% 
  ggplot(aes(x=Succession, y=k)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()

# start abundance
growth_succ %>% 
  ggplot(aes(x=Succession, y=log(start_abund))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()

# end abundance
growth_succ %>% 
  ggplot(aes(x=Succession, y=log(end_abund))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()

# change abundance
growth_succ %>% 
  ggplot(aes(x=Succession, y=log(change_abund))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()

# length of growth
growth_succ %>% 
  mutate(length=end_day-start_day) %>% 
  ggplot(aes(x=Succession, y=length)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()
```

# Death characteristics

### Overall

```{r}
# Average of each inoculant
death_after_overall <- death_after %>% 
  group_by(Inoculant, Succession) %>%
  summarize(k = mean(k),
            start_abund = mean(start_abund),
            end_abund = mean(end_abund),
            change_abund = mean(change_abund),
            length = mean(end_day - start_day)) %>% 
  ungroup()
  
# k 
death_after_overall %>% 
  ggplot(aes(x=Succession, y=k)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# start abundance
death_after_overall %>% 
  ggplot(aes(x=Succession, y=log(start_abund))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# end abundance
death_after_overall %>% 
  ggplot(aes(x=Succession, y=log(end_abund))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# change abundance
death_after_overall %>% 
  ggplot(aes(x=Succession, y=log(abs(change_abund)))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# length of growth
death_after_overall %>% 
  ggplot(aes(x=Succession, y=length)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()
```

Statistics:

Specific death rate (k):

```{r}
gr_k_lm <- lm(k ~ Succession, death_after_overall)
hist(resid(gr_k_lm)) # non normal
plot(predict(gr_k_lm), resid(gr_k_lm))

# Krukal Wallis test
kruskal.test(k ~ Succession, death_after_overall)
```

Change in abundance:

```{r}
de_change_lm <- lm(change_abund ~ Succession, death_after_overall)
hist(resid(de_change_lm)) # non-normal
plot(predict(de_change_lm), resid(de_change_lm)) # Heteroskedastic

# Kruskal Wallis
kruskal.test(change_abund ~ Succession, death_after_overall)
```

Length:

```{r}
de_length_lm <- lm(length ~ Succession, death_after_overall)
hist(resid(de_length_lm)) # non normal
plot(predict(de_length_lm), resid(de_length_lm))

# Use Kruskal Wallis test instead
kruskal.test(length ~ Succession, death_after_overall)
```

### Individual inoculants

```{r}
# death occurence (proportional)
death_after %>% 
  group_by(Inoculant, Succession) %>% 
  summarize(deaths = n()) %>% 
  ungroup() %>% 
  inner_join(growth_succ_totals) %>% 
  mutate(proportion = deaths/total) %>% 
  ggplot(aes(x=Succession, y=proportion)) +
  geom_jitter(width=0.1) +
  labs(y="Proportion of estimates that died after growth") +
  theme_test()

# k death
death_after %>% 
  ggplot(aes(x=Succession, y=k)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()

# start abundance
death_after %>% 
  ggplot(aes(x=Succession, y=log(start_abund))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()

# end abundance
death_after %>% 
  ggplot(aes(x=Succession, y=log(end_abund))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()

# change abundance
death_after %>% 
  ggplot(aes(x=Succession, y=log(abs(change_abund)))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()
```

Not enough data.

# Genomic characteristics

As predicted by PAPRICA.

### Overall

```{r}
# Average of each inoculant
genome_overall <- growth_succ %>% 
  group_by(Inoculant, Succession) %>%
  summarize(n16S = mean(change_abund),
            genome_size = mean(genome_size),
            phi = mean(phi),
            npaths_actual = mean(npaths_actual),
            nec_actual = mean(nec_actual)) %>% 
  ungroup()
  
# n16S
genome_overall %>% 
  ggplot(aes(x=Succession, y=n16S)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# Genome size
genome_overall %>% 
  ggplot(aes(x=Succession, y=genome_size)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# Genome plasticity
genome_overall %>% 
  ggplot(aes(x=Succession, y=phi)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# number of pathways
genome_overall %>% 
  ggplot(aes(x=Succession, y=npaths_actual)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()

# number of enzymes
genome_overall %>% 
  ggplot(aes(x=Succession, y=nec_actual)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  theme_test()
```

Statistics:

n16S:

```{r}
n16S_lm <- lm(log(n16S) ~ Succession, genome_overall)
hist(resid(n16S_lm)) # non-normal
plot(predict(n16S_lm), resid(n16S_lm))

# Kruskal Wallis
kruskal.test(log(n16S) ~ Succession, genome_overall)
```

Genome size:

```{r}
genome_lm <- lm(genome_size ~ Succession, genome_overall)
hist(resid(genome_lm))
plot(predict(genome_lm), resid(genome_lm))
summary(genome_lm)
```


### Individual inoculants

```{r}
# n16S
growth_succ %>% 
  ggplot(aes(x=Succession, y=n16S)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()

# Genome size
growth_succ %>% 
  ggplot(aes(x=Succession, y=genome_size)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()

# Genome plasticity
growth_succ %>% 
  ggplot(aes(x=Succession, y=phi)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()

# number of pathways
growth_succ %>% 
  ggplot(aes(x=Succession, y=npaths_actual)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()

# number of enzymes
growth_succ %>% 
  ggplot(aes(x=Succession, y=nec_actual)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  facet_wrap(~Inoculant) +
  theme_test()
```

# Dominance through time

* Using the normalized abundances for each taxon at each time point
* Creating proportions to control for unequal numbers of estimates

```{r}
# Label norm abundance data early/late
grnorm_succ <- inner_join(grnorm, growers, by="label")

# Total norm. abundance at each time point
grnorm_totals <- grnorm_succ  %>% 
  group_by(Inoculant, Day) %>%
  summarize(norm_abund_total = sum(norm_abund_avg)) %>% 
  ungroup()
  
# Calculate proportion of total abundance at each tp
grnorm_props <- grnorm_succ %>% 
  group_by(Inoculant, Day, Succession) %>%
  summarize(succ_total = sum(norm_abund_avg)) %>% 
  ungroup() %>% 
  inner_join(grnorm_totals) %>% 
  mutate(proportion = succ_total/norm_abund_total)

# Visualize
grnorm_props %>% 
  mutate(fade = if_else(Succession == "early", "yes", "no")) %>% 
  ggplot(aes(x=Day, y=proportion, color=Succession, alpha=fade)) +
  geom_point() +
  geom_line() +
  geom_vline(xintercept = 3.5, linetype=3) +
  geom_vline(xintercept = 9, linetype=3) +
  geom_vline(xintercept = 18, linetype=3) +
  geom_vline(xintercept = 24, linetype=3) +
  facet_wrap(~Inoculant) +
  labs(y="Proportion of total abundance") +
  theme_test()

# Calculate overall
grnorm_props_overall <- grnorm_props %>% 
  group_by(Day, Succession) %>% 
  summarize(prop_mean = mean(proportion)) %>% 
  ungroup() %>% 
  mutate(fade = if_else(Succession == "early", "yes", "no"))

# Visualize
grnorm_props_overall %>% 
  mutate(fade = if_else(Succession == "early", "yes", "no")) %>%
  ggplot(aes(x=Day, y=prop_mean, color=Succession, alpha=fade)) +
  geom_point() +
  geom_line() +
  geom_vline(xintercept = 3.5, linetype=3) +
  geom_vline(xintercept = 9, linetype=3) +
  geom_vline(xintercept = 18, linetype=3) +
  geom_vline(xintercept = 24, linetype=3) +
  labs(y="Proportion of total abundance (avg over inocs)") +
  theme_test()
```

# Phylogenetic conservation

Are groups that start growing "together" more phylogenetically conserved?

Prediction: conservation stronger at beginning growers than later (due to filter at beginning and common traits for rapid initial growth)

Prep data:

```{r}
# Create presence/absence matrix of growing taxa
growth_presabs <- growth_succ %>% 
  select(ASV, Inoculant, Succession) %>% 
  mutate(presence = 1) %>% # presence/absence
  pivot_wider(names_from="ASV", values_from="presence") %>% 
  mutate(across(-c("Inoculant", "Succession"), ~ if_else(is.na(.x), 0, .x))) %>% # replace coerced NAs with 0s
  rownames_to_column(var="Sample") # dummy sample names for reference
```

Calculate mean phylogenetic distance:

```{r}
# Calculate mpd for each inoculant group
mpd_inoc <- data.frame()

for (i in unique(growth_presabs$Inoculant)) {
  for (s in as.character(unique(growth_presabs[growth_presabs$Inoculant==i,]$Succession))) {
    # Subset
    data_sub <- filter(growth_presabs, Inoculant==i & Succession==s) %>% 
      select(-c(Sample, Inoculant, Succession)) %>% 
      as.matrix()
    # Prune tree
    tree_prune <- prune.sample(data_sub, tree)
    # Calculate MPD
    mpd_sub <- mpd(data_sub, cophenetic(tree_prune)) # broken here
    # Save result
    this_row <- bind_cols(Inoculant=i, Succession=s, mpd=mpd_sub)
    mpd_inoc <- bind_rows(mpd_inoc, this_row)
  }
}
```

Visualize:

```{r}
mpd_inoc %>% 
  mutate(Succession = as_factor(Succession)) %>% 
  ggplot(aes(x=Succession, y=mpd)) +
  geom_jitter(width=0.1) +
  labs(y="Mean pairwise phylogenetic distance", x="Start of growth stage") +
  theme_test()
```

# Batch dynamics

Whole communities:

```{r}
# Reformat
meta2 <- meta %>%
  select(SampleID, Sample, Inoculant, Day, Replicate)

norm_meta <- norm %>% 
  pivot_longer(-SampleID, values_to="norm_abund", names_to="ASV") %>% 
  inner_join(meta2)

# Average ASV abundance across replicates
norm_avg <- norm_meta %>% 
  group_by(Inoculant, Day, ASV) %>% 
  summarize(norm_abund = mean(norm_abund)) %>% 
  ungroup()

# Sum each day for each inoculant
norm_sum <- norm_avg %>% 
  group_by(Inoculant, Day) %>% 
  summarize(norm_abund = sum(norm_abund)) %>% 
  ungroup() %>% 
  mutate(Inoculant = as.character(Inoculant))

# Visualize
norm_sum %>% 
  ggplot(aes(x=Day, y=log(norm_abund), color=Inoculant)) +
  geom_point() +
  geom_line() +
  #facet_wrap(~Inoculant) +
  labs(title="Total community normalized abundance") +
  theme_test()
```

Growers only:

```{r}
growth_asvs <- growth$ASV %>% as.character()

grnorm_totals <- grnorm %>% 
  group_by(Inoculant, Day) %>% 
  summarize(total_abund = sum(norm_abund_avg)) %>% 
  ungroup()

grnorm_totals %>% 
  ggplot(aes(x=Day, y=log(total_abund), color=Inoculant)) +
  geom_point() +
  geom_line() + 
  labs(title="Growing taxa total normalized abundance") +
  theme_test()
```

Together:

```{r}
# Combine whole community and growing community normalized abundance totals
norm_wholegr <- norm_sum %>% 
  inner_join(grnorm_totals) %>% 
  rename(whole=norm_abund, growth=total_abund) %>% 
  pivot_longer(-c(Inoculant, Day), names_to="type", values_to="norm_abund")

# Visualize
norm_wholegr %>% 
  ggplot(aes(x=Day, y=log(norm_abund), alpha=type)) + 
  geom_point() +
  geom_line() +
  facet_wrap(~Inoculant) +
  theme_test()
```

Why do some time points have higher growth abundance than whole community abundance? That shouldn't be possible.

Most abundance is made up of taxa with growth estimates after day 3 or 4, indicating we did a pretty good job of capturing the growers/major players in the bottles.

But then why do betaNTI results for whole community vs growing taxa differ?


