---
title: "Growth and death dynamics during exponential and stationary phases"
author: "Cassandra Wattenburger"
date: "4/13/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

```{r}
# Clear working directory, load in packages, generate package info
rm(list=ls())

library("phyloseq")
library("vegan")
library("tidyverse")
library("lmerTest")
library("cowplot")

sessionInfo()

# I can't type "head" apparently so I'm going to make an alias
h <- function(x) {head(x)}
```

# Notes

* Copy of SFA2_expstatphases_growthdeath
* Plan to clean up

# Import data

* metadata
* growth estimates
* normalized abundances
* rarefied phyloseq

```{r}
# Metadata
meta <- readRDS("../data_intermediate/SFA2_metadata.rds")

# Growth estimates
growth <- readRDS("../data_intermediate/SFA2_growth_estimates_pap2.rds")

# Death estimates
death <- readRDS("../data_intermediate/SFA2_death_estimates_pap2.rds")

# All paprica estimates
pap_all <- readRDS("../data_intermediate/SFA2_paprica_all.rds")

# Normalized abundance data
norm <- readRDS("../data_intermediate/SFA2_normalized.rds")

# Rarefied phyloseq
physeq <- readRDS("../data_intermediate/SFA2_physeq_rare.rds")
```

# Total abundance

Defining as when the total normalized abundance of each inoculant evens out, indicating balanced growth and death.

## All taxa

```{r}
# Add metadata and reformat
meta <- meta %>%
  select(SampleID, Sample, Inoculant, Day, Replicate)

norm_meta <- norm %>%
  full_join(meta)

pap_slim <- pap_all %>% 
  select(ASV, n16S)

# Calculate total normalized abundance through time
norm_whole_total <- norm_meta %>% 
  inner_join(pap_slim) %>% # add 16S copy predictions
  mutate(norm_abund_corr = norm_abund / n16S) %>% # correct normalized abundance by 16S copies
  group_by(Inoculant, Day, ASV) %>% # average across replicates
  summarize(norm_abund_corr = mean(norm_abund_corr, na.rm=TRUE)) %>% 
  ungroup() %>%
  group_by(Inoculant, Day) %>%  # sum across ASVs
  summarize(total = sum(norm_abund_corr, na.rm=TRUE)) %>%
  ungroup() %>%
  filter(Inoculant %in% c(10, 12,19, 2, 41, 47)) %>% # remove controls
  select(Inoculant, Day, total) # tidy

# Calculate overall avg
norm_whole_avg <- norm_whole_total %>% 
  group_by(Day) %>% 
  summarize(total = mean(total)) %>% 
  ungroup() %>% 
  add_column(Inoculant="avg") %>% 
  select(Inoculant, Day, total)
```

Visualize:

```{r}
plot_whole_total <- norm_whole_total %>% 
  add_row(norm_whole_avg) %>% 
  mutate(fade = if_else(Inoculant!="avg", "avg", "inoc")) %>% 
  ggplot(aes(x=Day, y=log(total), group=Inoculant, alpha=fade)) +
  geom_point() +
  geom_line() + 
  geom_vline(xintercept=9, linetype=2) +
  labs(title="Summed normalized abundance of all taxa", y="ln summed normalized abundance") +
  theme_test() +
  theme(legend.position = "none")
plot_whole_total
```

## Growing taxa

```{r}
# Make a label for growing ASVs
growth_labels <- growth %>% 
  mutate(label = paste0(Inoculant, ASV)) %>%
  select(label, n16S)

# Calculate total normalized abundance through time
norm_growth_total <- norm_meta %>% 
  mutate(label = paste0(Inoculant, ASV)) %>% 
  inner_join(growth_labels) %>% # isolate growing taxa only
  mutate(norm_abund_corr = norm_abund / n16S) %>% # correct normalized abundance by 16S copies
  group_by(Inoculant, Day, ASV) %>% # average across replicates
  summarize(norm_abund_corr = mean(norm_abund_corr, na.rm=TRUE)) %>% 
  ungroup() %>%
  group_by(Inoculant, Day) %>% 
  summarize(total = sum(norm_abund_corr)) %>% # total normalized abundances
  ungroup() %>%
  filter(Inoculant %in% c(10, 12,19, 2, 41,47)) %>% # remove controls
  select(Inoculant, Day, total)

# Calculate overall avg
norm_growth_avg <- norm_growth_total %>% 
  group_by(Day) %>% 
  summarize(total = mean(total)) %>% 
  ungroup() %>% 
  add_column(Inoculant="avg") %>% 
  select(Inoculant, Day, total)
```

Visualize:

```{r}
plot_grow_total<- norm_growth_total %>% 
  add_row(norm_growth_avg) %>% 
  mutate(fade = if_else(Inoculant!="avg", "avg", "inoc")) %>% 
  ggplot(aes(x=Day, y=log(total), group=Inoculant, alpha=fade)) +
  geom_point() +
  geom_line() + 
  geom_vline(xintercept=9, linetype=2) +
  labs(title="Summed normalized abundance of growing taxa", y="ln summed normalized abundance") +
  coord_cartesian(xlim = c(0, 30)) +
  theme_test() +
  theme(legend.position = "none")
plot_grow_total 
```

## Contribution growing taxa

End of exponential and end of incubation:

```{r}
# norm_whole_total
# norm_growth_total

# Day 9
whole_day30 <- norm_whole_total %>% 
  filter(Day==9) %>% 
  select(Inoculant, wh_total=total)

norm_growth_total %>% 
  filter(Day==9) %>% 
  select(Inoculant, gr_total=total) %>% 
  inner_join(whole_day30) %>% 
  mutate(prop = gr_total/wh_total) %>% 
  summarize(prop_mean = mean(prop),
            prop_sd = sd(prop))

# Day 30
whole_day30 <- norm_whole_total %>% 
  filter(Day==30) %>% 
  select(Inoculant, wh_total=total)

norm_growth_total %>% 
  filter(Day==30) %>% 
  select(Inoculant, gr_total=total) %>% 
  inner_join(whole_day30) %>% 
  mutate(prop = gr_total/wh_total) %>% 
  summarize(prop_mean = mean(prop),
            prop_sd = sd(prop))
```

# Group ASVs

Group 1: start_day end_day <= 9 (grew during exponential phase only)
group 2: end_day  > 9 (grew during exponential and/or stationary phases)

```{r}
# Growth
growth_grp <- growth %>% 
  mutate(group = if_else(end_day <= 9, "grp1", "grp2"),
         group = as.factor(group))

# Death
death_grp <- death %>% 
  mutate(group = if_else(end_day <= 9, "grp1", "grp2"),
         group = as.factor(group))
```

## Group summaries

Summary of number ASVs per group:

```{r}
# Growth
growth_grp %>% 
  group_by(Inoculant, group) %>% 
  summarize(total = n()) %>% 
  ungroup() %>% 
  group_by(group) %>% 
  summarize(total_mean = mean(total), total_sd = sd(total)) %>% 
  ungroup()

# Death
death_grp %>% 
  group_by(Inoculant, group) %>% 
  summarize(total = n()) %>% 
  ungroup() %>% 
  group_by(group) %>% 
  summarize(total_mean = mean(total), total_sd = sd(total)) %>% 
  ungroup()
```

Proportion of total estimates:

```{r}
# Total growth estimates per inoculant
all_ests <- growth_grp %>% 
  group_by(Inoculant) %>% 
  summarize(both = n()) %>% 
  ungroup()

# Group 2
growth_grp %>% 
  group_by(Inoculant, group) %>% 
  summarize(total = n()) %>% 
  ungroup() %>% 
  pivot_wider(names_from=group, values_from=total) %>% 
  inner_join(all_ests) %>% 
  mutate(grp2_prop = grp2/both) %>% 
  summarize(grp2_prop_mean = mean(grp2_prop),
            grp2_prop_sd = sd(grp2_prop))
```



Summary of phylogenetic placements in groups:

* Phylum level

```{r}
growth_grp %>% 
  group_by(Inoculant, group, Phylum) %>% 
  summarize(total = n()) %>% 
  ungroup() %>% 
  group_by(group, Phylum) %>% 
  summarize(total_mean = mean(total, na.rm=TRUE), total_sd = sd(total, na.rm=TRUE)) %>% 
  ungroup() %>% 
  arrange(group, desc(total_mean))
```

```{r}
death_grp %>% 
  group_by(Inoculant, group, Phylum) %>% 
  summarize(total = n()) %>% 
  ungroup() %>% 
  group_by(group, Phylum) %>% 
  summarize(total_mean = mean(total, na.rm=TRUE), total_sd = sd(total, na.rm=TRUE)) %>% 
  ungroup() %>% 
  arrange(group, desc(total_mean))
```

Summary of k, lag, change abund for each group:

```{r}
growth_grp %>% 
  mutate(g = log(2)/abs(k)) %>% # calculate generation time
  group_by(group) %>% 
  summarize(g_mean = mean(g, na.rm=TRUE),
            g_df = sd(g, na.rm=TRUE),
            lag_mean = mean(start_day, na.rm=TRUE),
            lag_sd = sd(start_day, na.rm=TRUE),
            chgabund_mean = mean(change_abund_corr, na.rm=TRUE),
            chgabund_sd = sd(change_abund_corr, na.rm=TRUE)) %>% 
  ungroup()
```

```{r}
death_grp %>% 
  mutate(h = log(2)/abs(k)) %>% # calculate halving time
  group_by(group) %>% 
  summarize(h_mean = mean(h, na.rm=TRUE),
            h_df = sd(h, na.rm=TRUE),
            lag_mean = mean(start_day, na.rm=TRUE),
            lag_sd = sd(start_day, na.rm=TRUE),
            chgabund_mean = mean(change_abund_corr, na.rm=TRUE),
            chgabund_sd = sd(change_abund_corr, na.rm=TRUE)) %>% 
  ungroup()
```

## Growth group examples

```{r}
# 157, 129
# group 1
growth_grp1_ex <- growth_grp %>% 
  filter(ASV==growth[157,]$ASV & Inoculant==growth[157,]$Inoculant) # see growth estimation plots from script 03

# Prepare abudance table
norm_grp1_ex <- norm_meta %>%
  filter(ASV==growth[157,]$ASV & Inoculant==growth[157,]$Inoculant) %>% 
  group_by(Inoculant, Day) %>% 
  summarize(norm_abund_avg = mean(norm_abund, na.rm=TRUE)) %>% # average across technical replicates
  ungroup() %>% 
  arrange(Day) %>% 
  filter(norm_abund_avg > 0) # filter missing values
              
# Visualize
plot_grp1_ex <- norm_grp1_ex %>% 
  ggplot(aes(x=Day, y=log(norm_abund_avg))) +
  geom_point(shape=1, size=3, color="#6F7378") +
  geom_line(color="#6F7378") +
  geom_vline(xintercept=9, linetype=2) +
  geom_smooth(method="lm", data=norm_grp1_ex[growth_grp1_ex$start_pt:growth_grp1_ex$end_pt,], linetype=2, color="black") +
  labs(title=paste0("group 1 example - ", growth_grp1_ex$Phylum, ", ", growth_grp1_ex$Genus), x="Day", y="ln normalized abundance") +
  coord_cartesian(xlim = c(0, 30)) +
  theme_test()
plot_grp1_ex
```

```{r}
# group 2
growth_grp2_ex <- growth_grp %>% 
  filter(ASV==growth[79,]$ASV & Inoculant==growth[79,]$Inoculant) # see growth estimation plots from script 03

# Prepare abudnance table
norm_grp2_ex <- norm_meta %>%
  filter(ASV==growth[79,]$ASV & Inoculant==growth[79,]$Inoculant) %>% 
  group_by(Inoculant, Day) %>% 
  summarize(norm_abund_avg = mean(norm_abund, na.rm=TRUE)) %>% # average across technical replicates
  ungroup() %>% 
  arrange(Day) %>% 
  filter(norm_abund_avg > 0) # filter missing values
              
# Visualize
plot_grp2_ex <- norm_grp2_ex %>% 
  ggplot(aes(x=Day, y=log(norm_abund_avg))) +
  geom_point(shape=1, size=3, color="#6F7378") +
  geom_line(color="#6F7378") +
  geom_vline(xintercept=9, linetype=2) +
  geom_smooth(method="lm", data=norm_grp2_ex[growth_grp2_ex$start_pt:growth_grp2_ex$end_pt,], linetype=2, color="black") +
  labs(title=paste0("group 2 example - ", growth_grp2_ex$Phylum, ", ", growth_grp2_ex$Genus), x="Day", y="ln normalized abundance") +
  coord_cartesian(xlim = c(0, 30)) +
  theme_test()
plot_grp2_ex
```

All together:

```{r}
# Comgroupe examples into one df
norm_grp1_ex <- add_column(norm_grp1_ex, group="grp1")
norm_grp2_ex <- add_column(norm_grp2_ex, group="grp2")
norm_grps_ex <- bind_rows(norm_grp1_ex, norm_grp2_ex)

# Visualize
norm_grps_ex %>% 
  ggplot(aes(x=Day, y=log(norm_abund_avg))) +
  geom_point(size=3, aes(shape=group)) +
  geom_line(aes(group=group)) +
  geom_vline(xintercept=9, linetype=2) +
  # Growth estimates
  geom_smooth(method="lm", data=norm_grp1_ex[growth_grp1_ex$start_pt:growth_grp1_ex$end_pt,], linetype=2, alpha=0.2) +
  geom_smooth(method="lm", data=norm_grp2_ex[growth_grp2_ex$start_pt:growth_grp2_ex$end_pt,], linetype=2, alpha=0.2) +
  #labs(title=paste0("group 3 example -", growth_grp3_ex$Phylum, ", ", growth_grp3_ex$Genus), x="Day", y="ln normalized abundance") +
  coord_cartesian(xlim = c(0, 30)) +
  theme_test()
```

## Phylogenetic memberships

Relative abundances down to order level for each inoculant

By number of estimates.

Growth:

```{r}
# Phylum-level makeup of ASV estimates
growth_phyla <- growth_grp %>% 
  group_by(Phylum, ASV) %>% 
  summarize(total=n())

# total number of growth estimates
growth_ests <- nrow(growth_grp)

# Calculate % estimates of each phylum
growth_phyla %>%
  group_by(Phylum) %>% 
  summarize(total=n()) %>% 
  arrange(desc(total)) %>% 
  add_column(all = growth_ests) %>% 
  mutate(perc = (total/all)*100)
```

```{r}
# Total ASVs estimated per inoculant/group
gr_inocgrp_totals <- growth_grp %>% 
  group_by(Inoculant, group) %>% 
  summarize(total = n()) %>% 
  ungroup()

# Order level
gr_ord_relabund <- growth_grp %>% 
  group_by(Inoculant, group, Phylum, Order) %>% 
  summarize(order_total = n()) %>% 
  ungroup() %>% 
  inner_join(gr_inocgrp_totals) %>% 
  mutate(relabund = order_total/total)
gr_ord_relabund
```

Death: 

```{r}
# Phylum-level makeup of ASV estimates
death_phyla <- death_grp %>% 
  group_by(Phylum, ASV) %>% 
  summarize(total=n())

# total number of death estimates
death_ests <- nrow(death_grp)

# Calculate % estimates of each phylum
death_phyla %>%
  group_by(Phylum) %>% 
  summarize(total=n()) %>% 
  arrange(desc(total)) %>% 
  add_column(all = death_ests) %>% 
  mutate(perc = (total/all)*100)
```

Visualize:

```{r}
gr_ord_relabund %>% 
  ggplot(aes(x=group, y=order_total, color=Phylum)) +
  geom_jitter() +
  geom_vline(xintercept=1.5) +
  facet_wrap(~Inoculant) +
  theme_test()

gr_ord_relabund %>% 
  mutate(border = if_else(Phylum=="Actinobacteriota", "yes", "no")) %>% 
  ggplot(aes(x=group, y=relabund, fill=Order, color=Phylum)) +
  geom_col() +
  facet_wrap(~Inoculant) +
  theme_test()
```

## Group 3

Started growth after day 9.

```{r}
# Growth estimates
growth_grp3 <- growth %>% 
  mutate(group = if_else(start_day > 9, "grp3", "not"),
         group = as.factor(group)) %>% 
  filter(group=="grp3")

# Death estimates
norm_grp3_prep <- growth_grp3 %>% 
  inner_join(norm) %>%
  select(label, Sample, Inoculant, Day, ASV, norm_abund) %>% 
  group_by(label, Inoculant, Day, ASV) %>% # average technical reps
  summarize(norm_abund = mean(norm_abund)) %>% 
  ungroup() %>% 
  arrange(Inoculant, Day) %>% 
  filter(norm_abund > 0) %>% # remove drop outs
  mutate(norm_abund_ln = log(norm_abund))

# Taxonomy
tax_grp3 <- growth_grp3 %>% 
  inner_join(norm) %>%
  select(ASV, Phylum:Genus) %>% 
  unique()
```

Visualize:

```{r}
count <-  0
for (l in as.character(growth_grp3$label)) {
  # Subset time series
  count <- count+1
  growth_label <- filter(growth_grp3, label==l)
  norm_label <- filter(norm_grp3_prep, label==l) %>% 
    arrange(Day)
  # Title information
  asv <- growth_label$ASV
  tax_info <- filter(tax_grp3, ASV == asv)
  title <- paste0(tax_info$Phylum, ", ", tax_info$Genus)
  # Graph with estimate
  graph <- ggplot(norm_label, aes(x=Day, y=norm_abund_ln)) +
    geom_point(shape=1, size=3, color="#6F7378") +
    geom_line(color="#6F7378") +
    geom_smooth(method="lm", data=norm_label[growth_label$start_pt:growth_label$end_pt,], linetype=2, color="black") +
    labs(title=paste0(title), x="Day", y="ln Relational abundance") +
    theme_test() +
    theme(axis.title = element_blank(),
        axis.text = element_text(size=10))
  print(graph)
  assign(paste0("plot", count, "_grp3"), graph)
}
```

# Growth traits

* growth rates
* change in abundance
* n16S

Visualize:

```{r}
# k
growth_grp %>% 
  ggplot(aes(x=group, y=k)) +
  geom_boxplot() +
  geom_jitter(size=2, alpha=0.5) +
  theme_test() +
  labs(y="Growth rate (day-1)", x="") +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 15))

# change in abundance
growth_grp %>% 
  ggplot(aes(x=group, y=log(change_abund_corr))) +
  geom_boxplot() +
  geom_jitter(size=2, alpha=0.5) +
  theme_test() +
  labs(y="Normalized abundance change (ln)", x="") +
  theme(axis.text = element_text(size = 10),
        axis.title = element_text(size = 12))

# n16S
growth_grp %>% 
  ggplot(aes(x=group, y=log(n16S))) +
  geom_boxplot() +
  geom_jitter(size=2, alpha=0.5) +
  theme_test() +
  labs(y="16S copy number (ln)", x="") +
  theme(axis.text = element_text(size = 10),
        axis.title = element_text(size = 12))

# Below baked into definition, so not super informative but good to know
# Length
growth_grp %>% 
  mutate(length=end_day-start_day) %>% 
  ggplot(aes(x=group, y=length)) +
  geom_boxplot() +
  geom_jitter(size=2, alpha=0.5) +
  theme_test() +
  labs(y="length of growth", x="") +
  theme(axis.text = element_text(size = 10),
        axis.title = element_text(size = 12))
```

## Statistics

Linear mixed effects models with random slopes and intercepts for inoculant

```{r}
# k
k_lmer <- lmer(k ~ group + (1 + Inoculant | Inoculant), growth_grp)
hist(resid(k_lmer))
plot(predict(k_lmer), resid(k_lmer))
anova(k_lmer)

# change abund
chg_lmer <- lmer(log(change_abund_corr) ~ group + (1 + Inoculant | Inoculant), growth_grp)
hist(resid(chg_lmer))
plot(predict(chg_lmer), resid(chg_lmer))
anova(chg_lmer)

# 16S
n16s_lmer <- lmer(log(n16S) ~ group + (1 + Inoculant | Inoculant), growth_grp)
hist(resid(n16s_lmer))
plot(predict(n16s_lmer), resid(n16s_lmer)) # heteroskedastic
# Welch t-test
t.test(log(n16S) ~ group, data=growth_grp) 

# length
growth_grp <- mutate(growth_grp, length = end_day-start_day)
length_lmer <- lmer(log(length) ~ group + (1 + Inoculant | Inoculant), growth_grp)
hist(resid(length_lmer))
plot(predict(length_lmer), resid(length_lmer))
anova(length_lmer)
```

## Correlations

Length vs. gain in abudance: 

```{r}
# Average ASVs across inoculant (independence issue)
growth_grp_avg <- growth_grp %>% 
  group_by(ASV, group) %>% 
  summarize(k = mean(k),
            length = mean(length),
            change_abund_corr = mean(change_abund_corr)) %>%
  ungroup()

growth_grp_avg %>% 
  ggplot(aes(x=length, y=log(change_abund_corr))) + #, color=group
  geom_point() +
  geom_smooth(method="lm") + #, aes(group=group)
  theme_test()

growth_grp_avg %>% 
  ggplot(aes(x=length, y=log(change_abund_corr), color=group)) + 
  geom_point() +
  geom_smooth(method="lm", aes(group=group)) + 
  theme_test()
```

Pearson test:

```{r}
# All
cor.test(growth_grp_avg$length, log(growth_grp_avg$change_abund_corr), method="pearson")

# Group 2 only
cor.test(growth_grp_avg[growth_grp_avg$group=="grp2",]$length, log(growth_grp_avg[growth_grp_avg$group=="grp2",]$change_abund_corr), method="pearson")
```

Rate vs. gain in abudance: 

```{r}
growth_grp_avg %>% 
  ggplot(aes(x=k, y=log(change_abund_corr))) + #, color=group
  geom_point() +
  geom_smooth(method="lm") + #, aes(group=group)
  theme_test()

growth_grp_avg %>% 
  ggplot(aes(x=k, y=log(change_abund_corr), color=group)) + 
  geom_point() +
  geom_smooth(method="lm", aes(group=group)) + 
  theme_test()
```

Pearson test:

```{r}
# All
cor.test(growth_grp_avg$k, log(growth_grp_avg$change_abund_corr), method="pearson")

# Group 2 only
cor.test(growth_grp_avg[growth_grp_avg$group=="grp2",]$k, log(growth_grp_avg[growth_grp_avg$group=="grp2",]$change_abund_corr), method="pearson")
```

# Death Traits

Visualize:

```{r}
# k
death_grp %>% 
  ggplot(aes(x=group, y=k)) +
  geom_boxplot() +
  geom_jitter(size=2, alpha=0.5) +
  theme_test() +
  labs(y="Death rate (day-1)", x="") +
  theme(axis.text = element_text(size = 10),
        axis.title = element_text(size = 12))

# change in abundance
death_grp %>% 
  ggplot(aes(x=group, y=log(abs(change_abund_corr)))) +
  geom_boxplot() +
  geom_jitter(size=2, alpha=0.5) +
  theme_test() +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 15)) +
  labs(y="Normalized abundance change (ln abs. value)", x="")

# n16S
death_grp %>% 
  ggplot(aes(x=group, y=n16S)) +
  geom_boxplot() +
  geom_jitter(size=2, alpha=0.5) +
  theme_test() +
  theme(axis.text = element_text(size = 12),
        axis.title = element_text(size = 15)) +
  labs(y="16S copy numbers", x="")
```

### Statistics 

Linear mixed effects models with random slopes and intercepts for inoculant

```{r}
# k
kdeath_lmer <- lmer(log(abs(k)) ~ group + (1 + Inoculant | Inoculant), death_grp)
hist(resid(kdeath_lmer))
plot(predict(kdeath_lmer), resid(kdeath_lmer))
anova(kdeath_lmer)

# change abund
chgdeath_lmer <- lmer(log(abs(change_abund_corr)) ~ group + (1 + Inoculant | Inoculant), death_grp)
hist(resid(chgdeath_lmer))
plot(predict(chgdeath_lmer), resid(chgdeath_lmer))
anova(chgdeath_lmer)

# n16S
n16S_lmer <- lmer(n16S ~ group + (1 + Inoculant | Inoculant), death_grp)
hist(resid(chgdeath_lmer))
plot(predict(chgdeath_lmer), resid(chgdeath_lmer))
anova(chgdeath_lmer)
```

# Abundance gains/losses during phases

```{r}
# Prepare data
growth_slim <- growth %>% 
  select(Inoculant, ASV, slope, yint, k, g, start_day, end_day, n16S)

death_slim <- death %>% 
  select(Inoculant, ASV, slope, yint, k, h, start_day, end_day, n16S)
```

Impute based on linear estimate:

```{r}
# Impute values for start and end of growth/death based on linear estimate

# Growth
imputed_deltaNg <- growth_slim %>% 
  mutate(group = if_else(start_day < 9 & end_day <= 9, "group 1", "group 2"), # label group
         impute_expstart = if_else(group=="group 1", slope*start_day+yint, # impute ln abund
                                   if_else(group=="group 2" & start_day < 9,
                                           slope*start_day+yint, NULL)),
         impute_expend = if_else(group=="group 1", slope*end_day+yint,
                                 if_else(group=="group 2" & start_day < 9 & end_day >= 9,
                                         slope*9+yint, NULL)),
         impute_statstart = if_else(group=="group 2" & start_day <= 9, slope*9+yint, 
                                    if_else(group=="group 2" & start_day > 9,
                                            slope*start_day+yint, NULL)),
         impute_statend = if_else(group=="group 2", slope*end_day+yint, NULL),
         # Un-ln and correct for n16S
         expstart_corr = exp(impute_expstart)/n16S,
         expend_corr = exp(impute_expend)/n16S,
         statstart_corr = exp(impute_statstart)/n16S,
         statend_corr = exp(impute_statend)/n16S,
         exp_deltaN = expend_corr - expstart_corr,
         stat_deltaN = statend_corr - statstart_corr) %>% 
  select(Inoculant:group, exp_deltaN, stat_deltaN) %>% 
  group_by(Inoculant) %>% 
  summarize(exp = sum(exp_deltaN, na.rm = TRUE),
            stat = sum(stat_deltaN, na.rm=TRUE)) %>% 
  pivot_longer(-Inoculant, names_to = "phase", values_to = "deltaN") %>% 
  add_column(type="growth")

# Death
imputed_deltaNd <- death_slim %>% 
  mutate(group = if_else(start_day < 9 & end_day <= 9, "group 1", "group 2"), # label group
         impute_expstart = if_else(group=="group 1", slope*start_day+yint, # impute ln abund
                                   if_else(group=="group 2" & start_day < 9,
                                           slope*start_day+yint, NULL)),
         impute_expend = if_else(group=="group 1", slope*end_day+yint,
                                 if_else(group=="group 2" & start_day < 9 & end_day >= 9,
                                         slope*9+yint, NULL)),
         impute_statstart = if_else(group=="group 2" & start_day <= 9, slope*9+yint, 
                                    if_else(group=="group 2" & start_day > 9,
                                            slope*start_day+yint, NULL)),
         impute_statend = if_else(group=="group 2", slope*end_day+yint, NULL),
         # Un-ln and correct for n16S
         expstart_corr = exp(impute_expstart)/n16S,
         expend_corr = exp(impute_expend)/n16S,
         statstart_corr = exp(impute_statstart)/n16S,
         statend_corr = exp(impute_statend)/n16S,
         exp_deltaN = expend_corr - expstart_corr,
         stat_deltaN = statend_corr - statstart_corr) %>% 
  select(Inoculant:group, exp_deltaN, stat_deltaN) %>% 
  group_by(Inoculant) %>% 
  summarize(exp = abs(sum(exp_deltaN, na.rm = TRUE)),
            stat = abs(sum(stat_deltaN, na.rm=TRUE))) %>% 
  pivot_longer(-Inoculant, names_to = "phase", values_to = "deltaN") %>% 
  add_column(type="death")

# Combine growth and death
imputed_deltaN <- bind_rows(imputed_deltaNg, imputed_deltaNd)

imputed_deltaN %>% 
  group_by(phase, type) %>% 
  summarize(deltaN = mean(deltaN))
```

Visualize:

```{r}
imputed_deltaN %>% 
  ggplot(aes(x=phase, y=log(deltaN), color=type)) +
  #geom_point() +
  geom_boxplot() +
  theme_test()
```

# Dataframe for Dan

```{r}
# Prepare data
dan_df <- norm %>% # norm. abundance data
  filter(Type=="Growth" & norm_abund!=0) %>% # filter samples and remove non-present taxa
  select(Inoculant, Day, Replicate, ASV, norm_abund) %>%
  group_by(Inoculant, Day, ASV) %>% # average replicates
  summarize(norm_abund_avg = mean(norm_abund)) %>% 
  inner_join(growth_slim) %>% # add growth estimates
  mutate(norm_abund_corr = norm_abund_avg/n16S) %>%  # correct normalized abundance
  select(Inoculant, Day, ASV, start_day, end_day, g, norm_abund_corr) %>% 
  ungroup() %>% 
  mutate(group = if_else(start_day < 9 & end_day <= 9, "group 1", "group 2")) %>% # group
  mutate(phase = if_else(Day <= 9, "exponential", "stationary"), # label phase
         length = end_day - start_day, # calculate length of growth
         label = paste0(Inoculant, phase, "_", ASV)) %>% # unique label
  group_by(label, Inoculant, ASV, phase, g, start_day, length) %>% 
  summarize(norm_abund_corr = mean(norm_abund_corr)) %>% # average ASV abundance per phase
  ungroup()

# Calculate relative abundances
inoc_totals <- dan_df %>% 
  group_by(Inoculant, phase) %>% 
  summarize(norm_abund_total = sum(norm_abund_corr)) %>% 
  ungroup()

dan_df2 <- data.frame()
for (l in unique(dan_df$label)) {
  data_sub <- filter(dan_df, label==l)
  relabund <- data_sub$norm_abund_corr/filter(inoc_totals, Inoculant==data_sub$Inoculant & phase==data_sub$phase)$norm_abund_total
  this_row <- bind_cols(Inoculant=data_sub$Inoculant, ASV=data_sub$ASV,
                        phase=data_sub$phase, g=data_sub$g, lag=data_sub$start_day,
                        length=data_sub$length, 
                        mean_abund=data_sub$norm_abund_corr, relabund=relabund)
  dan_df2 <- bind_rows(dan_df2, this_row)
}

# Add taxonomy
tax <- read_rds("../data_intermediate/SFA2_taxonomy.rds")
dan_df2 <- left_join(dan_df2, tax)
```

Save dataframe:

```{r, eval=FALSE}
write_csv(dan_df2, "../SFA_growthphase.csv")
```

# Figures

## Community curve 

Total normalized abundance:

```{r}
plot1 <- norm_growth_total %>% 
  add_row(norm_growth_avg) %>% 
  mutate(fade = if_else(Inoculant!="avg", "avg", "inoc")) %>% 
  ggplot(aes(x=Day, y=log(total), group=Inoculant, alpha=fade)) +
  geom_point() +
  geom_line() + 
  geom_vline(xintercept=9, linetype=2) +
  coord_cartesian(xlim = c(0, 30)) +
  theme_test() +
  theme(axis.text = element_text(size=10),
        axis.title = element_blank())
plot1
```

Examples

```{r}
plot2 <- norm_grps_ex %>% 
  ggplot(aes(x=Day, y=log(norm_abund_avg))) +
  geom_point(aes(shape=group), size=2) +
  geom_line(aes(group=group)) +
  geom_vline(xintercept=9, linetype=2) +
  geom_smooth(method="lm", data=norm_grp1_ex[growth_grp1_ex$start_pt:growth_grp1_ex$end_pt,], linetype=2, color="#444444", alpha=0.3) +
  geom_smooth(method="lm", data=norm_grp2_ex[growth_grp2_ex$start_pt:growth_grp2_ex$end_pt,], linetype=2, color="#444444", alpha=0.3) +
  coord_cartesian(xlim = c(0, 30)) +
  theme_test() +
  theme(axis.text = element_text(size=10),
      axis.title = element_blank())
plot2
```

Combine:

```{r}
plot_comb12 <- plot_grid(plot1, plot2, nrow=2)
plot_comb12
```

```{r, eval=FALSE}
ggsave("../figures/fig_totalnorm_ex.svg", plot_comb12, height=180, width=180, units="mm", device="svg")
```

## Groups growth and death traits

```{r}
# growth k
plot3 <- growth_grp %>% 
  ggplot(aes(x=group, y=k)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  #labs(title="plot3") +
  theme_test() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size=10),
        axis.title = element_blank())
plot3

# growth change in abundance
plot4 <- growth_grp %>% 
  ggplot(aes(x=group, y=log(change_abund_corr))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  #labs(title="plot4") +
  theme_test() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size=10),
        axis.title = element_blank())
plot4

# death k
plot5 <- death_grp %>% 
  ggplot(aes(x=group, y=k)) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  #labs(title="plot5") +
  theme_test() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size=10),
        axis.title = element_blank())
plot5

# change in abundance
plot6 <- death_grp %>% 
  ggplot(aes(x=group, y=log(abs(change_abund_corr)))) +
  geom_boxplot() +
  geom_jitter(alpha=0.5) +
  #labs(title="plot6") +
  theme_test() +
  theme(axis.text = element_text(size = 10)) +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size=10),
        axis.title = element_blank())
plot6
```

Combine:

```{r}
plot_comb36 <- plot_grid(plot3, plot5, plot4, plot6, nrow=2)
plot_comb36
```

```{r, eval=FALSE}
ggsave("../figures/fig_growthdeathgroups.svg", plot_comb36, height=180, width=180, units="mm", device="svg")
```

## Supplemental: Whole vs growth total curve

```{r}
# Comgroupe whole and growing only data averages
norm_whole_avg <- add_column(type="whole", norm_whole_avg)
norm_growth_avg <- add_column(type="growth", norm_growth_avg)

norm_avg <- bind_rows(norm_whole_avg, norm_growth_avg)
```

```{r}
# Total averaged abundances by phase, compraing whole to growing
norm_avg %>% 
  mutate(phase = if_else(Day <= 9, "exp", "stat")) %>% 
  group_by(phase, type) %>% 
  summarize(total = mean(total))
```

```{r}
supp1 <- norm_avg %>% 
  mutate(type = as_factor(type),
         type = fct_relevel(type, c("whole", "growth"))) %>% 
  ggplot(aes(x=Day, y=log(total), shape=type)) +
  geom_point(size=3) +
  geom_line(aes(group=type, linetype=type)) +
  labs(y="Total relational abundnace (ln avg)") +
  theme_test()
supp1
```

```{r, eval=FALSE}
ggsave("../figures/supp_wholegrtotal.svg", supp1, height=180, width=180, units="mm", device="svg")
```

## Supplement group 3

```{r}
count <-  0
for (l in as.character(growth_grp3$label)) {
  # Subset time series
  count <- count+1
  growth_label <- filter(growth_grp3, label==l)
  norm_label <- filter(norm_grp3_prep, label==l) %>% 
    arrange(Day)
  # Title information
  asv <- growth_label$ASV
  tax_info <- filter(tax_grp3, ASV == asv)
  title <- paste0(tax_info$Phylum, ", ", tax_info$Genus)
  # Graph with estimate
  graph <- ggplot(norm_label, aes(x=Day, y=norm_abund_ln)) +
    geom_point(shape=1, size=3, color="#6F7378") +
    geom_line(color="#6F7378") +
    geom_smooth(method="lm", data=norm_label[growth_label$start_pt:growth_label$end_pt,], linetype=2, color="black") +
    scale_x_continuous(limits=c(0,30)) +
    theme_test() +
    theme(title = element_text(size=12),
        axis.title = element_blank(),
        axis.text = element_text(size=10))
  print(graph)
  assign(paste0("plot", count, "_grp3"), graph)
}
```

Combine:

```{r}
supp_grp3 <- plot_grid(plot1_grp3, plot2_grp3, plot3_grp3, plot4_grp3, plot5_grp3, plot6_grp3, nrow=2)
supp_grp3
```

```{r, eval=FALSE}
ggsave("../figures/supp_grp3.svg", supp_grp3, height=120, width=210, units="mm", device="svg")
```

## Length trait

```{r}
plot_lt <- growth_grp %>% 
  mutate(length=end_day-start_day) %>% 
  ggplot(aes(x=group, y=length)) +
  geom_boxplot() +
  geom_jitter(size=2, alpha=0.5) +
  theme_test() +
  theme(title = element_text(size=12),
        axis.title = element_blank(),
        axis.text = element_text(size=10),
        legend.position = "none")
plot_lt
```

```{r, eval=FALSE}
ggsave("../figures/supp_lengthtrait.svg", plot_lt, height=180, width=180, units="mm", device="svg")
```

## Length/k x deltaNg

Length:

```{r}
grp_colors <- c(grp1="#5b5b5b", grp2="black")

plot_length1 <- growth_grp_avg %>% 
  ggplot(aes(x=length, y=log(change_abund_corr))) +
  geom_point(aes(color=group, shape=group)) +
  geom_smooth(method="lm", linetype=2, color="black", alpha=0.5) +
  scale_color_manual(values=grp_colors) +
  theme_test() +
  theme(title = element_text(size=12),
        axis.title = element_blank(),
          axis.text = element_text(size=10),
        legend.position = "none")
plot_length1

plot_length2 <- growth_grp_avg %>% 
  ggplot(aes(x=length, y=log(change_abund_corr), color=group, shape=group)) + 
  geom_point() +
  geom_smooth(method="lm", linetype=2, alpha=0.5, aes(group=group)) + 
  scale_color_manual(values=grp_colors) +
  theme_test() +
  theme(title = element_text(size=12),
        axis.title = element_blank(),
        axis.text = element_text(size=10),
        legend.position = "none")
plot_length2
```

k: 

```{r}
plot_k1 <- growth_grp %>% 
  ggplot(aes(x=k, y=log(change_abund_corr))) + 
  geom_point(aes(color=group, shape=group)) +
  geom_smooth(method="lm", linetype=2, alpha=0.5, color="black") +
  scale_color_manual(values=grp_colors) +
  theme_test() +
  theme(title = element_text(size=12),
        axis.title = element_blank(),
        axis.text = element_text(size=10),
        legend.position = "none")
plot_k1

plot_k2 <- growth_grp %>% 
  ggplot(aes(x=k, y=log(change_abund_corr), color=group, shape=group)) + 
  geom_point() +
  geom_smooth(method="lm", linetype=2, alpha=0.5, aes(group=group)) + 
  scale_color_manual(values=grp_colors) +
  theme_test() +
  theme(title = element_text(size=12),
        axis.title = element_blank(),
        axis.text = element_text(size=10),
        legend.position = "none")
plot_k2
```

```{r}
# Combine
plot_lengthk <- plot_grid(plot_length1,  plot_k1, plot_length2, plot_k2, nrow=2)
plot_lengthk
```

```{r, eval=FALSE}
# save
ggsave("../figures/fig_lengthkxdeltang.svg", plot_lengthk, height=180, width=180, units="mm", device="svg")
```

## Abundance gain/loss by phase

```{r}
plotx <- imputed_deltaN %>% 
  mutate(phase2 = if_else(phase=="exp", "Exponential", "Stationary"),
         type2 = if_else(type=="death", "Death", "Growth")) %>% 
  ggplot(aes(x=phase2, y=log(deltaN))) +
  geom_boxplot() +
  theme_test() +
  facet_wrap(~type2) +
  theme_test() +
  theme(title = element_text(size=12),
        axis.title = element_blank(),
          axis.text = element_text(size=10))
plotx
```

```{r}
ggsave("../figures/fig_deltaexpstat.svg", plotx, height=120, width=210, units="mm", device="svg")
```

# SCRAP

Growth old

Calc. deltaNg per phase:

N0 = normalized abundance of taxon at start of time interval (not log transformed)
g = generation time of taxon (days)
t = days in time frame
n = t/g, number of generations
Nt = N0*2^n, biomass at end of time frame
deltaN = Nt - N0, change in biomass

Full equation:
deltaN = N0*2^((end-start)/g) - N0

```{r}
# Exponential phase

# Group 1
exp_gain1 <- filter(growth_norm, end_day <= 9) %>% # grew only during exponential
  filter(start_day==Day) %>% # use abundance at start of growth (N0)
  mutate(deltaN = norm_abund_corr*2^((end_day-start_day)/g) - norm_abund_corr) %>% # gained during growth
  group_by(Inoculant) %>% 
  summarize(deltaN = sum(deltaN)) %>% 
  ungroup()

# Group 2
exp_gain2 <- filter(growth_norm, end_day >= 9 & start_day < 9) %>% # grew after exponential too
  filter(start_day==Day) %>% # starting abundance at start day
  mutate(deltaN = norm_abund_corr*2^((9-start_day)/g) - norm_abund_corr) %>% # gained prior to day 9
  group_by(Inoculant) %>% 
  summarize(deltaN = sum(deltaN)) %>% 
  ungroup()

# Stationary phase

# Group 2
stat_gain1 <- filter(growth_norm, end_day >= 9 & start_day <= 9) %>% # started growing in exponential
  filter(Day==9) %>% # starting abundnace will be abundance on day 9
  mutate(deltaN = norm_abund_corr*2^((end_day-9)/g) - norm_abund_corr) %>% # gained after day 9
  group_by(Inoculant) %>% 
  summarize(deltaN = sum(deltaN)) %>% 
  ungroup()

# Group 3
stat_gain2 <- filter(growth_norm, end_day >= 9 & start_day >= 9) %>% # started growing in stationary
  filter(Day==start_day) %>% # abundance at start of growth
  mutate(deltaN = norm_abund_corr*2^((end_day-start_day)/g) - norm_abund_corr) %>%
  group_by(Inoculant) %>% 
  summarize(deltaN = sum(deltaN)) %>% 
  ungroup() %>% 
  add_row(Inoculant="41", deltaN=0) %>% # add missing incoulants (no observed estimates)
  add_row(Inoculant="19", deltaN=0)

# Combine
# SUMMING CORRECT?
expstat_gain <- bind_cols(exp_gain1, exp_gain2) %>% 
  bind_cols(stat_gain1) %>%
  bind_cols(stat_gain2) %>% head()
  select(Inoculant=`Inoculant...1`, deltaN_exp1=`deltaN...2`, deltaN_exp2=`deltaN...4`, deltaN_stat=`deltaN...6`) %>% 
  mutate(deltaN_exp = deltaN_exp1 + deltaN_exp2)

expstat_gain %>% 
  summarize(exp_mean = mean(deltaN_exp),
            exp_sd = sd(deltaN_exp),
            stat_mean = mean(deltaN_stat),
            stat_sd = sd(deltaN_stat))
```

Death old

```{r}
# Prepare data
death_slim <- death %>% 
  select(Inoculant, ASV, slope, yint, k, h, start_day, end_day, n16S) # WHERE IS H?

death_norm <- norm %>% 
  filter(Type=="Growth" & norm_abund!=0) %>% 
  select(Inoculant, Day, Replicate, ASV, norm_abund) %>%
  group_by(Inoculant, Day, ASV) %>% # average replicates
  summarize(norm_abund_avg = mean(norm_abund)) %>% 
  inner_join(death_slim) %>%
  mutate(norm_abund_corr = norm_abund_avg/n16S) %>% 
  select(Inoculant, Day, ASV, start_day, end_day, slope, yint, h, norm_abund_corr, n16S) %>% 
  group_by(Inoculant, ASV, Day, slope, yint, h, start_day, end_day, n16S) %>% 
  summarize(norm_abund_corr = mean(norm_abund_corr)) %>% # average replicate values
  ungroup() %>% 
  mutate(group = if_else(start_day < 9 & end_day <= 9, "group 1", "group 2"))
```

Calc. deltaNd per phase:

```{r}
# Exponential phase

# Group 1
exp_loss1 <- filter(death_norm, end_day <= 9) %>% # grew only during exponential
  filter(start_day==Day) %>% # use abundance at start of death (N0)
  mutate(deltaN = norm_abund_corr*2^((end_day-start_day)/h) - norm_abund_corr) %>% # lossed during death
  group_by(Inoculant) %>% 
  summarize(deltaN = sum(deltaN)) %>% 
  ungroup()

# Group 2
exp_loss2 <- filter(death_norm, end_day >= 9 & start_day <= 9) %>% # grew after exponential too
  filter(start_day==Day) %>% # starting abundance at start day
  mutate(deltaN = norm_abund_corr*2^((9-start_day)/h) - norm_abund_corr) %>% # lossed prior to day 9 
  group_by(Inoculant) %>% 
  summarize(deltaN = sum(deltaN)) %>% 
  ungroup() %>% 
  add_row(Inoculant="41", deltaN=0) # inlcude inoculants with no estimates here

# Stationary phase

# Group 2
stat_loss1 <- filter(death_norm, end_day >= 9 & start_day <= 9) %>% # started growing in exponential
  filter(Day==9) %>% # starting abundnace will be abundance on day 9
  mutate(deltaN = norm_abund_corr*2^((end_day-9)/h) - norm_abund_corr) %>% # lossed after day 9
  group_by(Inoculant) %>% 
  summarize(deltaN = sum(deltaN)) %>% 
  ungroup() %>% 
  add_row(Inoculant="41", deltaN=0) # inlcude inoculants with no estimates here

# Group 3
stat_loss2 <- filter(death_norm, end_day >= 9 & start_day >= 9) %>% # started growing in stationary
  filter(Day==start_day) %>% # abundance at start of death
  mutate(deltaN = norm_abund_corr*2^((end_day-start_day)/h) - norm_abund_corr) %>%
  group_by(Inoculant) %>% 
  summarize(deltaN = sum(deltaN)) %>% 
  ungroup() %>% 
  add_row(Inoculant="41", deltaN=0) %>% # add missing incoulants (no observed estiamtes)
  add_row(Inoculant="19", deltaN=0)

# Combine
expstat_loss <- bind_cols(exp_loss1, exp_loss2) %>% 
  bind_cols(stat_loss1) %>%
  bind_cols(stat_loss2) %>% 
  select(Inoculant=`Inoculant...1`, deltaN_exp1=`deltaN...2`, deltaN_exp2=`deltaN...4`, deltaN_stat=`deltaN...6`) %>% 
  mutate(deltaN_exp = deltaN_exp1 + deltaN_exp2)

expstat_loss %>% 
  summarize(exp_mean = mean(deltaN_exp),
            exp_sd = sd(deltaN_exp),
            stat_mean = mean(deltaN_stat),
            stat_sd = sd(deltaN_stat))
```

Visualize:

```{r}
# Combine gains and losses
expstat_gain2 <- expstat_gain %>% 
  select(Inoculant, deltaN_exp, deltaN_stat) %>% 
  pivot_longer(cols=!Inoculant, names_to="phase", values_to="deltaN") %>% 
  mutate(phase = gsub("deltaN_", "", phase),
         type="growth")

expstat_loss2 <- expstat_loss %>% 
  select(Inoculant, deltaN_exp, deltaN_stat) %>% 
  pivot_longer(cols=!Inoculant, names_to="phase", values_to="deltaN") %>% 
  mutate(phase = gsub("deltaN_", "", phase),
         type="death")

expstat_gainloss <- bind_rows(expstat_gain2, expstat_loss2)

# Plot
expstat_gainloss %>% 
  ggplot(aes(x=phase, y=log(deltaN), color=type)) +
  geom_point() +
  geom_boxplot() +
  theme_test()
```

Group 1 : Group 2

How does the ratio of biomass in each group change with time?

Impute 0s

```{r}
growth_grp_slim <- growth_grp %>% 
  select(label, Inoculant, ASV, slope, yint, start_day, end_day, n16S, group)

# Mark for imputation
norm_impute <- norm %>% 
  filter(Type=="Growth") %>% 
  select(Inoculant, Replicate, Day, ASV, norm_abund) %>% 
  group_by(Inoculant, Day, ASV) %>%  # average replicates
  summarize(norm_abund = mean(norm_abund)) %>% 
  ungroup() %>% 
  inner_join(pap_slim) %>% 
  inner_join(growth_grp_slim) %>% 
  mutate(impute = if_else((start_day <= Day & Day <= end_day) & norm_abund==0, "linear_model", # imputation method depends on if during growth or not
                  if_else((start_day > Day | Day > end_day) & norm_abund==0, "average", "none")))

# Impute 0s via linear model
norm_imputed <- norm_impute %>% 
  mutate(norm_impute_ln = case_when(impute=="linear_model" ~ slope*Day+yint,
                                          impute=="average" ~ 0, # placeholder, need to do this in a loop because it has a lot of conditionals
                                          impute=="none" ~ log(norm_abund)))

# Impute via averaging nearest neighbors
days <- data.frame(Day=na.omit(unique(meta$Day))) %>% 
  add_column(rownum = 1:nrow(.))


testing <- head(unique(norm_imputed$label))

for (l in testing) {
  data_label <- filter(norm_imputed, label==l)
  data_avg <- filter(data_label, impute=="average")
  for (d in data_avg$Day) {
    if (d==0) {
      norm_abund <- 0
      while (norm_abund==0) {
        day_after <- days[days[days$Day==d,]$rownum + 1,]$Day # BOOKMARK
        x <- filter(data_label, Day==day_after)
        norm_abund <- x$norm_abund
      next
      }
    }
    else {
      break
    }
  }
}









# ye olde way:
norm_grp_ratio <- norm %>% 
  filter(Type=="Growth" & norm_abund!=0) %>% 
  select(Inoculant, Day, ASV, norm_abund) %>% 
  inner_join(pap_slim) %>% 
  inner_join(growth_grp_slim) %>% 
  mutate(norm_abund_corr = norm_abund/n16S) %>% # correct multiple copy numbers
  group_by(Inoculant, Day, group) %>% 
  summarize(total = sum(norm_abund_corr)) %>% 
  ungroup() %>% 
  pivot_wider(names_from = group, values_from = total) %>% ################# NAs introduced in grp1 col here (imputation or removal needed)
  filter(complete.cases(grp1, grp2)) %>% # remove days where no grp1 or grp2 representatives
  mutate(ratio = grp1/grp2)
```

Visualize:

```{r}
norm_grp_ratio %>% 
  ggplot(aes(x=Day, y=log(ratio), color=Inoculant)) +
  geom_smooth(method="lm", linetype=2, aes(growth=Inoculant)) +
  geom_point() +
  geom_line(alpha=0.5, aes(group=Inoculant)) +
  geom_vline(xintercept = 9, linetype=2) +
  theme_test()
```

Impute 0s

Identify problem 0s:

```{r}
# Check for problematic 0s for taxa on start of growth
impute_start <- growth_norm %>% 
  mutate(label = paste0(Inoculant, "_", ASV),
         impute = if_else(Day==start_day & norm_abund_corr==0, "impute", # mark for imputation if 0 at start of growth
                                   if_else(Day!=start_day, "NA", "no impute"))) %>% 
  filter(impute=="impute")
impute_start
# none

# Check for 0s on day 9 (shift between exp and stationary)
impute_day9 <- growth_norm %>%
  mutate(label = paste0(Inoculant, "_", ASV),
         impute = if_else(Day==9 & norm_abund_corr==0, "impute", # mark for imputation if 0 at start
                                   if_else(Day!=start_day, "NA", "no impute"))) %>%
  filter(impute=="impute" & group=="group 2")
impute_day9
# none
```

Impute day 9 for group 2:

Using slope of linear model to solve y (ln abundance) at x=9 (day)

Group 2 only.

```{r}
growth_impute <- growth_norm %>% 
  filter(group == "group 2" & Day==9) %>%
  mutate(impute_lnabund = slope*9 + yint, # estimate ln abundance
         impute_abund = exp(impute_lnabund), # untransform
         impute_abund_corr = impute_abund/n16S) %>% # correct for 16S copy number 
  full_join(growth_norm) %>% 
  mutate(norm_abund_corr = if_else(Day==9 & norm_abund_corr==0, impute_abund_corr, norm_abund_corr)) %>% # impute
  select(Inoculant, ASV, Day, start_day, end_day, g, norm_abund_corr)

head(growth_impute)
```

Impute 0s

Identify problem 0s:

```{r}
# Check for problematic 0s for taxa on start of death
impute_start <- death_norm %>% 
  mutate(label = paste0(Inoculant, "_", ASV),
         impute = if_else(Day==start_day & norm_abund_corr==0, "impute", # mark for imputation if 0 at start
                                   if_else(Day!=start_day, "NA", "no impute"))) %>% 
  filter(impute=="impute")
impute_start
# none

# Check for 0s on day 9 (shift between exp and stationary)
impute_day9 <- death_norm %>%
  mutate(label = paste0(Inoculant, "_", ASV),
         impute = if_else(Day==9 & norm_abund_corr==0, "impute", # mark for imputation if 0 at start
                                   if_else(Day!=start_day, "NA", "no impute"))) %>%
  filter(impute=="impute" & group=="group 2")
impute_day9
```

Impute day 9 for group 2:

Using slope of linear model to solve y (ln abundance) at x=9 (day)

Group 2 only.

```{r}
death_impute <- death_norm %>% 
  filter(group == "group 2" & Day==9) %>%
  mutate(impute_lnabund = slope*9 + yint,
         impute_abund = exp(impute_lnabund),
         impute_abund_corr = impute_abund/n16S) %>% 
  full_join(death_norm) %>% 
  mutate(norm_abund_corr = if_else(Day==9 & norm_abund_corr==0, impute_abund_corr, norm_abund_corr)) %>% # impute
  select(Inoculant, ASV, Day, start_day, end_day, h, norm_abund_corr)
```
